// drizzle-orm@1.0.0-rc.4/effect-mysql2/codecs.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __mysql_core_codecs_ts = require("../mysql-core/codecs.cjs");

//#region src/effect-mysql2/codecs.ts
const effectMysql2Codecs = (0, __mysql_core_codecs_ts.refineGenericMySqlCodecs)({
	bigint: { normalize: BigInt },
	"bigint:number": { normalize: Number },
	"bigint:string": { normalize: String },
	boolean: { normalize: (value) => value === 1 },
	"decimal:number": { normalize: Number },
	"decimal:bigint": { normalize: BigInt },
	float: { normalize: __mysql_core_codecs_ts.floatFromDouble },
	date: {
		cast: __mysql_core_codecs_ts.castToText,
		normalize: (value) => new Date(value)
	},
	"date:string": { cast: __mysql_core_codecs_ts.castToText },
	datetime: {
		cast: __mysql_core_codecs_ts.castToText,
		normalize: (value) => /* @__PURE__ */ new Date(value.replace(" ", "T") + "Z")
	},
	"datetime:string": { cast: __mysql_core_codecs_ts.castToText },
	timestamp: {
		cast: __mysql_core_codecs_ts.castToText,
		normalize: (value) => /* @__PURE__ */ new Date(value + "+0000")
	},
	"timestamp:string": { cast: __mysql_core_codecs_ts.castToText },
	binary: { normalize: (value) => value.toString() },
	varbinary: { normalize: (value) => value.toString() },
	blob: { normalize: (value) => value.toString("utf8") },
	tinyblob: { normalize: (value) => value.toString("utf8") },
	mediumblob: { normalize: (value) => value.toString("utf8") },
	longblob: { normalize: (value) => value.toString("utf8") }
});

//#endregion
exports.effectMysql2Codecs = effectMysql2Codecs;
