// drizzle-orm@1.0.0-rc.4/effect-mysql2/codecs.js
import { castToText, floatFromDouble, refineGenericMySqlCodecs } from "../mysql-core/codecs.js";

//#region src/effect-mysql2/codecs.ts
const effectMysql2Codecs = refineGenericMySqlCodecs({
	bigint: { normalize: BigInt },
	"bigint:number": { normalize: Number },
	"bigint:string": { normalize: String },
	boolean: { normalize: (value) => value === 1 },
	"decimal:number": { normalize: Number },
	"decimal:bigint": { normalize: BigInt },
	float: { normalize: floatFromDouble },
	date: {
		cast: castToText,
		normalize: (value) => new Date(value)
	},
	"date:string": { cast: castToText },
	datetime: {
		cast: castToText,
		normalize: (value) => /* @__PURE__ */ new Date(value.replace(" ", "T") + "Z")
	},
	"datetime:string": { cast: castToText },
	timestamp: {
		cast: castToText,
		normalize: (value) => /* @__PURE__ */ new Date(value + "+0000")
	},
	"timestamp:string": { cast: castToText },
	binary: { normalize: (value) => value.toString() },
	varbinary: { normalize: (value) => value.toString() },
	blob: { normalize: (value) => value.toString("utf8") },
	tinyblob: { normalize: (value) => value.toString("utf8") },
	mediumblob: { normalize: (value) => value.toString("utf8") },
	longblob: { normalize: (value) => value.toString("utf8") }
});

//#endregion
export { effectMysql2Codecs };
