// drizzle-orm@1.0.0-rc.4/effect-mysql2/driver.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
const require_effect_mysql2_codecs = require('./codecs.cjs');
const require_effect_mysql2_session = require('./session.cjs');
let __entity_ts = require("../entity.cjs");
let __utils_ts = require("../utils.cjs");
let __mysql_core_dialect_ts = require("../mysql-core/dialect.cjs");
let effect_Effect = require("effect/Effect");
effect_Effect = require_runtime.__toESM(effect_Effect);
let __cache_core_cache_effect_ts = require("../cache/core/cache-effect.cjs");
let __effect_core_defaults_ts = require("../effect-core/defaults.cjs");
let __effect_core_index_ts = require("../effect-core/index.cjs");
let _effect_sql_mysql2_MysqlClient = require("@effect/sql-mysql2/MysqlClient");
let __mysql_core_effect_db_ts = require("../mysql-core/effect/db.cjs");

//#region src/effect-mysql2/driver.ts
var EffectMysql2Database = class extends __mysql_core_effect_db_ts.MySqlEffectDatabase {
	static [__entity_ts.entityKind] = "EffectMysql2Database";
};
/**
* Creates an EffectMysql2Database instance.
*
* Requires `MysqlClient`, `EffectLogger`, and `EffectCache` services to be provided.
* Use `DefaultServices` to provide default (no-op) logger and cache implementations.
*
* @example
* ```ts
* // With default services (no logging, no caching)
* const db = yield* MySqlDrizzle.make({ relations }).pipe(
*   Effect.provide(MySqlDrizzle.DefaultServices),
* );
*
* // With Effect-based logging
* const db = yield* MySqlDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layer),
*   Effect.provide(MySqlDrizzle.DefaultServices),
* );
*
* // With custom Drizzle logger
* const db = yield* MySqlDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layerFromDrizzle(myLogger)),
*   Effect.provide(MySqlDrizzle.DefaultServices),
* );
* ```
*/
const make = effect_Effect.fn("MySqlDrizzle.make")(function* (config = {}) {
	const client = yield* _effect_sql_mysql2_MysqlClient.MysqlClient;
	const cache = yield* __cache_core_cache_effect_ts.EffectCache;
	const logger = yield* __effect_core_index_ts.EffectLogger;
	const dialect = new __mysql_core_dialect_ts.MySqlDialect({
		useJitMappers: (0, __utils_ts.jitCompatCheck)(config.jit),
		codecs: config.codecs ?? require_effect_mysql2_codecs.effectMysql2Codecs,
		paginationToBigint: true
	});
	const relations = config.relations ?? {};
	const db = new EffectMysql2Database(dialect, new require_effect_mysql2_session.EffectMysql2Session(client, dialect, relations, {
		logger,
		cache
	}), relations);
	db.$client = client;
	db.$cache = cache;
	if (db.$cache) db.$cache["invalidate"] = cache.onMutate;
	return db;
});
/**
* Convenience function that creates an EffectMysql2Database with `DefaultServices` already provided.
*/
const makeWithDefaults = (config = {}) => make(config).pipe(effect_Effect.provide(__effect_core_defaults_ts.DefaultServices));

//#endregion
Object.defineProperty(exports, 'DefaultServices', {
  enumerable: true,
  get: function () {
    return __effect_core_defaults_ts.DefaultServices;
  }
});
exports.EffectMysql2Database = EffectMysql2Database;
exports.make = make;
exports.makeWithDefaults = makeWithDefaults;
