// drizzle-orm@1.0.0-rc.4/effect-mysql2/driver.js
import { effectMysql2Codecs } from "./codecs.js";
import { EffectMysql2Session } from "./session.js";
import { entityKind } from "../entity.js";
import { jitCompatCheck } from "../utils.js";
import { MySqlDialect } from "../mysql-core/dialect.js";
import * as Effect from "effect/Effect";
import { EffectCache } from "../cache/core/cache-effect.js";
import { DefaultServices, DefaultServices as DefaultServices$1 } from "../effect-core/defaults.js";
import { EffectLogger } from "../effect-core/index.js";
import { MysqlClient } from "@effect/sql-mysql2/MysqlClient";
import { MySqlEffectDatabase } from "../mysql-core/effect/db.js";

//#region src/effect-mysql2/driver.ts
var EffectMysql2Database = class extends MySqlEffectDatabase {
	static [entityKind] = "EffectMysql2Database";
};
/**
* Creates an EffectMysql2Database instance.
*
* Requires `MysqlClient`, `EffectLogger`, and `EffectCache` services to be provided.
* Use `DefaultServices` to provide default (no-op) logger and cache implementations.
*
* @example
* ```ts
* // With default services (no logging, no caching)
* const db = yield* MySqlDrizzle.make({ relations }).pipe(
*   Effect.provide(MySqlDrizzle.DefaultServices),
* );
*
* // With Effect-based logging
* const db = yield* MySqlDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layer),
*   Effect.provide(MySqlDrizzle.DefaultServices),
* );
*
* // With custom Drizzle logger
* const db = yield* MySqlDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layerFromDrizzle(myLogger)),
*   Effect.provide(MySqlDrizzle.DefaultServices),
* );
* ```
*/
const make = Effect.fn("MySqlDrizzle.make")(function* (config = {}) {
	const client = yield* MysqlClient;
	const cache = yield* EffectCache;
	const logger = yield* EffectLogger;
	const dialect = new MySqlDialect({
		useJitMappers: jitCompatCheck(config.jit),
		codecs: config.codecs ?? effectMysql2Codecs,
		paginationToBigint: true
	});
	const relations = config.relations ?? {};
	const db = new EffectMysql2Database(dialect, new EffectMysql2Session(client, dialect, relations, {
		logger,
		cache
	}), relations);
	db.$client = client;
	db.$cache = cache;
	if (db.$cache) db.$cache["invalidate"] = cache.onMutate;
	return db;
});
/**
* Convenience function that creates an EffectMysql2Database with `DefaultServices` already provided.
*/
const makeWithDefaults = (config = {}) => make(config).pipe(Effect.provide(DefaultServices$1));

//#endregion
export { DefaultServices, EffectMysql2Database, make, makeWithDefaults };
