// drizzle-orm@1.0.0-rc.4/effect-mysql2/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let effect_Effect = require("effect/Effect");
effect_Effect = require_runtime.__toESM(effect_Effect);
let __mysql_core_effect_session_ts = require("../mysql-core/effect/session.cjs");

//#region src/effect-mysql2/session.ts
var EffectMysql2Session = class extends __mysql_core_effect_session_ts.MySqlEffectSession {
	static [__entity_ts.entityKind] = "EffectMysql2Session";
	constructor(client, dialect, relations, options) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const executor = (params) => {
			const q = this.client.unsafe(query.sql, params);
			if (mode === "arrays") return q.values;
			if (mode === "objects") return q.withoutTransform;
			if (!mapper) return q.raw;
			return q.raw.pipe(effect_Effect.map(({ insertId, affectedRows }) => ({
				insertId,
				affectedRows
			})));
		};
		return new __mysql_core_effect_session_ts.MySqlEffectPreparedQuery(executor, query, mapper, mode, this.options.logger, this.options.cache, queryMetadata, cacheConfig);
	}
	transaction(transaction) {
		const { dialect, relations } = this;
		return this.client.withTransaction(effect_Effect.gen({ self: this }, function* () {
			return yield* transaction(new EffectMysql2Transaction(dialect, this, relations));
		}));
	}
};
var EffectMysql2Transaction = class extends __mysql_core_effect_session_ts.MySqlEffectTransaction {
	static [__entity_ts.entityKind] = "EffectMysql2Transaction";
	transaction(transaction) {
		return this.session.transaction(transaction);
	}
};

//#endregion
exports.EffectMysql2Session = EffectMysql2Session;
exports.EffectMysql2Transaction = EffectMysql2Transaction;
