// drizzle-orm@1.0.0-rc.4/effect-mysql2/session.js
import { entityKind } from "../entity.js";
import * as Effect from "effect/Effect";
import { MySqlEffectPreparedQuery, MySqlEffectSession, MySqlEffectTransaction } from "../mysql-core/effect/session.js";

//#region src/effect-mysql2/session.ts
var EffectMysql2Session = class extends MySqlEffectSession {
	static [entityKind] = "EffectMysql2Session";
	constructor(client, dialect, relations, options) {
		super(dialect);
		this.client = client;
		this.relations = relations;
		this.options = options;
	}
	prepareQuery(query, mode, mapper, queryMetadata, cacheConfig) {
		const executor = (params) => {
			const q = this.client.unsafe(query.sql, params);
			if (mode === "arrays") return q.values;
			if (mode === "objects") return q.withoutTransform;
			if (!mapper) return q.raw;
			return q.raw.pipe(Effect.map(({ insertId, affectedRows }) => ({
				insertId,
				affectedRows
			})));
		};
		return new MySqlEffectPreparedQuery(executor, query, mapper, mode, this.options.logger, this.options.cache, queryMetadata, cacheConfig);
	}
	transaction(transaction) {
		const { dialect, relations } = this;
		return this.client.withTransaction(Effect.gen({ self: this }, function* () {
			return yield* transaction(new EffectMysql2Transaction(dialect, this, relations));
		}));
	}
};
var EffectMysql2Transaction = class extends MySqlEffectTransaction {
	static [entityKind] = "EffectMysql2Transaction";
	transaction(transaction) {
		return this.session.transaction(transaction);
	}
};

//#endregion
export { EffectMysql2Session, EffectMysql2Transaction };
