// @vite-plugin-vercel/schemas@1.1.1/dist/index.js
import { z } from "zod/v4";
//#region src/config.ts
/**
* Schema definition for `.vercel/output/config.json`
* @see {@link https://vercel.com/docs/build-output-api/v3#build-output-configuration}
*/
const HasOrMissing = z.array(z.union([
	z.object({
		type: z.literal("host"),
		value: z.string()
	}).strict(),
	z.object({
		type: z.literal("header"),
		key: z.string(),
		value: z.string().optional()
	}).strict(),
	z.object({
		type: z.literal("cookie"),
		key: z.string(),
		value: z.string().optional()
	}).strict(),
	z.object({
		type: z.literal("query"),
		key: z.string(),
		value: z.string().optional()
	}).strict()
])).optional();
const vercelOutputConfigSchema = z.object({
	version: z.literal(3),
	routes: z.array(z.union([z.object({
		src: z.string(),
		dest: z.string().optional(),
		headers: z.record(z.string(), z.string()).optional(),
		methods: z.array(z.string()).optional(),
		status: z.number().int().positive().optional(),
		continue: z.boolean().optional(),
		check: z.boolean().optional(),
		missing: HasOrMissing,
		has: HasOrMissing,
		locale: z.object({
			redirect: z.record(z.string(), z.string()).optional(),
			cookie: z.string().optional()
		}).strict().optional(),
		middlewarePath: z.string().optional(),
		transforms: z.array(z.object({
			type: z.enum([
				"request.headers",
				"request.query",
				"response.headers"
			]),
			op: z.enum([
				"append",
				"set",
				"delete"
			]),
			target: z.object({ key: z.union([z.string(), z.object({
				eq: z.union([z.string(), z.number()]).optional(),
				neq: z.string().optional(),
				inc: z.array(z.string()).optional(),
				ninc: z.array(z.string()).optional(),
				pre: z.string().optional(),
				suf: z.string().optional(),
				gt: z.number().optional(),
				gte: z.number().optional(),
				lt: z.number().optional(),
				lte: z.number().optional()
			}).strict().refine((data) => Object.values(data).some((v) => v !== void 0), { message: "At least one field must be provided" })]) }).strict(),
			args: z.union([z.string(), z.array(z.string())]).optional()
		}).strict()).optional()
	}).strict(), z.object({
		handle: z.union([
			z.literal("rewrite"),
			z.literal("filesystem"),
			z.literal("resource"),
			z.literal("miss"),
			z.literal("hit"),
			z.literal("error")
		]),
		src: z.string().optional(),
		dest: z.string().optional(),
		status: z.number().optional()
	}).strict()])).optional(),
	images: z.object({
		sizes: z.tuple([z.number().int().positive(), z.number().int().positive()]),
		domains: z.array(z.string()).nonempty().optional(),
		minimumCacheTTL: z.number().int().positive().optional(),
		formats: z.union([z.literal("image/avif"), z.literal("image/webp")]).array().nonempty().optional(),
		dangerouslyAllowSVG: z.boolean().optional(),
		contentSecurityPolicy: z.string().optional()
	}).strict().optional(),
	wildcard: z.array(z.object({
		domain: z.string(),
		value: z.string()
	}).strict()).optional(),
	overrides: z.record(z.string(), z.object({
		path: z.string().optional(),
		contentType: z.string().optional()
	}).strict()).optional(),
	cache: z.array(z.string()).optional()
}).strict();
//#endregion
//#region src/exports.ts
const vercelEndpointExports = z.object({
	edge: z.boolean().optional(),
	headers: z.record(z.string(), z.string()).optional(),
	streaming: z.boolean().optional(),
	isr: z.object({ expiration: z.number().or(z.literal(false)) }).optional()
});
//#endregion
//#region src/prerender-config.ts
/**
* Schema definition for `.vercel/output/config.json`
* @see {@link https://vercel.com/docs/build-output-api/v3#build-output-configuration}
*/
const vercelOutputPrerenderConfigSchema = z.object({
	expiration: z.union([z.number().int().positive(), z.literal(false)]),
	group: z.number().int().optional(),
	bypassToken: z.string().optional(),
	fallback: z.string().optional(),
	allowQuery: z.array(z.string()).optional()
}).strict();
//#endregion
//#region src/vc-config.ts
/**
* Schema definition for `.vercel/output/functions/<name>.func/.vc-config.json`
* @see {@link https://vercel.com/docs/build-output-api/v3/primitives#serverless-function-configuration}
*/
/**
* Vercel Queues consumer trigger configuration for Serverless Functions.
*
* @experimental
* @see {@link https://vercel.com/docs/queues}
*/
const vercelOutputQueueTriggerSchema = z.object({
	type: z.literal("queue/v2beta"),
	topic: z.string(),
	consumer: z.string().optional(),
	maxDeliveries: z.number().int().positive().optional(),
	retryAfterSeconds: z.number().int().nonnegative().optional(),
	initialDelaySeconds: z.number().int().nonnegative().optional(),
	maxConcurrency: z.number().int().positive().optional()
}).passthrough();
const vercelOutputEdgeVcConfigSchema = z.object({
	runtime: z.literal("edge"),
	entrypoint: z.string(),
	envVarsInUse: z.array(z.string()).optional()
}).strict();
const vercelOutputServerlessVcConfigSchema = z.object({
	runtime: z.string(),
	handler: z.string(),
	memory: z.number().int().min(128).max(3008).optional(),
	maxDuration: z.number().int().positive().optional(),
	environment: z.record(z.string(), z.string()).optional(),
	regions: z.array(z.string()).optional(),
	supportsWrapper: z.boolean().optional(),
	supportsResponseStreaming: z.boolean().optional(),
	experimentalTriggers: z.array(vercelOutputQueueTriggerSchema).optional()
}).strict();
const vercelOutputServerlessNodeVcConfigSchema = vercelOutputServerlessVcConfigSchema.extend({
	launcherType: z.literal("Nodejs"),
	shouldAddHelpers: z.boolean().optional(),
	shouldAddSourcemapSupport: z.boolean().optional(),
	awsLambdaHandler: z.string().optional()
}).strict();
const vercelOutputVcConfigSchema = z.union([
	vercelOutputEdgeVcConfigSchema,
	vercelOutputServerlessVcConfigSchema,
	vercelOutputServerlessNodeVcConfigSchema
]);
//#endregion
export { vercelEndpointExports, vercelOutputConfigSchema, vercelOutputEdgeVcConfigSchema, vercelOutputPrerenderConfigSchema, vercelOutputQueueTriggerSchema, vercelOutputServerlessNodeVcConfigSchema, vercelOutputServerlessVcConfigSchema, vercelOutputVcConfigSchema };
