// drizzle-orm@1.0.0-rc.4/d1/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let __errors_ts = require("../errors.cjs");
let __logger_ts = require("../logger.cjs");
let __sqlite_core_async_session_ts = require("../sqlite-core/async/session.cjs");
let __cache_core_index_ts = require("../cache/core/index.cjs");

//#region src/d1/session.ts
var SQLiteD1Session = class extends __sqlite_core_async_session_ts.SQLiteAsyncSession {
	static [__entity_ts.entityKind] = "SQLiteD1Session";
	logger;
	cache;
	constructor(client, dialect, relations, options = {}) {
		super(dialect, "async");
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
		this.cache = options.cache ?? new __cache_core_index_ts.NoopCache();
	}
	prepareQuery(query, mode, _prepare, executeMethod, mapper, queryMetadata, cacheConfig) {
		let stmt;
		try {
			stmt = this.client.prepare(query.sql);
		} catch (e) {
			throw new __errors_ts.DrizzleQueryError(query.sql, query.params, e);
		}
		return new D1PreparedQuery(stmt, "async", executeMethod, {
			all: (params) => {
				if (mode === "arrays") return stmt.bind(...params).raw();
				return stmt.bind(...params).all().then(({ results }) => results);
			},
			get: (params) => {
				if (mode === "arrays") return stmt.bind(...params).raw().then((rows) => rows[0]);
				return stmt.bind(...params).first();
			},
			run: (params) => {
				return stmt.bind(...params).run();
			},
			values: (params) => {
				return stmt.bind(...params).raw();
			}
		}, query, mapper, mode, this.logger, this.cache, queryMetadata, cacheConfig);
	}
	async batch(queries) {
		const preparedQueries = [];
		const builtQueries = [];
		for (const query of queries) {
			const preparedQuery = query._prepare();
			const builtQuery = preparedQuery.getQuery();
			preparedQueries.push(preparedQuery);
			if (builtQuery.params.length > 0) builtQueries.push(preparedQuery.stmt.bind(...builtQuery.params));
			else {
				const builtQuery = preparedQuery.getQuery();
				builtQueries.push(this.client.prepare(builtQuery.sql).bind(...builtQuery.params));
			}
		}
		return (await this.client.batch(builtQueries)).map((result, i) => {
			const { executeMethod, mapper, mode } = preparedQueries[i];
			if (executeMethod === "run") return result;
			let values = result.results;
			if (executeMethod === "values") return d1ToRawMapping(values);
			if (executeMethod === "get") {
				if (!values[0]) return;
				if (!mapper) return mode === "arrays" ? d1ToRawMapping([values[0]])[0] : values[0];
				return mapper(mode === "arrays" ? d1ToRawMapping([values[0]]) : [values[0]])[0];
			}
			values = mode === "arrays" ? d1ToRawMapping(values) : values;
			if (!mapper) return values;
			return mapper(values);
		});
	}
	async transaction(transaction, config) {
		const tx = new D1Transaction("async", this.dialect, this, this.relations, void 0, true);
		await this.run(__sql_sql_ts.sql.raw(`begin${config?.behavior ? " " + config.behavior : ""}`));
		try {
			const result = await transaction(tx);
			await this.run(__sql_sql_ts.sql`commit`);
			return result;
		} catch (err) {
			await this.run(__sql_sql_ts.sql`rollback`);
			throw err;
		}
	}
};
var D1Transaction = class D1Transaction extends __sqlite_core_async_session_ts.SQLiteAsyncTransaction {
	static [__entity_ts.entityKind] = "D1Transaction";
	async transaction(transaction) {
		const savepointName = `sp${this.nestedIndex}`;
		const tx = new D1Transaction("async", this.dialect, this.session, this._.relations, this.nestedIndex + 1, this.forbidJsonb);
		await this.session.run(__sql_sql_ts.sql.raw(`savepoint ${savepointName}`));
		try {
			const result = await transaction(tx);
			await this.session.run(__sql_sql_ts.sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (err) {
			await this.session.run(__sql_sql_ts.sql.raw(`rollback to savepoint ${savepointName}`));
			throw err;
		}
	}
};
/**
* This function was taken from the D1 implementation: https://github.com/cloudflare/workerd/blob/4aae9f4c7ae30a59a88ca868c4aff88bda85c956/src/cloudflare/internal/d1-api.ts#L287
* It may cause issues with duplicated column names in join queries, which should be fixed on the D1 side.
* @param results
* @returns
*/
function d1ToRawMapping(results) {
	const rows = [];
	for (const row of results) {
		const entry = Object.keys(row).map((k) => row[k]);
		rows.push(entry);
	}
	return rows;
}
var D1PreparedQuery = class extends __sqlite_core_async_session_ts.SQLiteAsyncPreparedQuery {
	static [__entity_ts.entityKind] = "D1PreparedQuery";
	/** @internal */
	fields;
	/** @internal */
	stmt;
	constructor(stmt, resultKind, executeMethod = "all", executors, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(resultKind, executeMethod, executors, query, mapper, mode, logger, cache, queryMetadata, cacheConfig);
		this.stmt = stmt;
	}
};

//#endregion
exports.D1PreparedQuery = D1PreparedQuery;
exports.D1Transaction = D1Transaction;
exports.SQLiteD1Session = SQLiteD1Session;
