// drizzle-orm@1.0.0-rc.4/mysql-core/columns/custom.js
import { MySqlColumn, MySqlColumnBuilder } from "./common.js";
import { resolveMySqlTypeAlias } from "../codecs.js";
import { entityKind } from "../../entity.js";
import { getColumnNameAndConfig } from "../../utils.js";

//#region src/mysql-core/columns/custom.ts
var MySqlCustomColumnBuilder = class extends MySqlColumnBuilder {
	static [entityKind] = "MySqlCustomColumnBuilder";
	constructor(name, fieldConfig, customTypeParams) {
		super(name, "custom", "MySqlCustomColumn");
		this.config.fieldConfig = fieldConfig;
		this.config.customTypeParams = customTypeParams;
	}
	/** @internal */
	build(table) {
		return new MySqlCustomColumn(table, this.config);
	}
};
var MySqlCustomColumn = class extends MySqlColumn {
	static [entityKind] = "MySqlCustomColumn";
	/** @internal */
	codec;
	sqlName;
	mapFromJsonValue;
	jsonSelectIdentifier;
	constructor(table, config) {
		super(table, config);
		this.sqlName = config.customTypeParams.dataType(config.fieldConfig);
		this.mapToDriverValue = config.customTypeParams.toDriver ?? this.mapToDriverValue;
		this.mapFromDriverValue = config.customTypeParams.fromDriver ?? this.mapFromDriverValue;
		this.mapFromJsonValue = config.customTypeParams.fromJson;
		this.jsonSelectIdentifier = config.customTypeParams.forJsonSelect;
		const cfgCodec = typeof config.customTypeParams.codec === "string" || typeof config.customTypeParams.codec === "undefined" ? config.customTypeParams.codec : config.customTypeParams.codec(config.fieldConfig);
		this.codec = typeof cfgCodec === "string" ? resolveMySqlTypeAlias(cfgCodec) : void 0;
	}
	getSQLType() {
		return this.sqlName;
	}
};
/**
* Custom mysql database data type generator
*/
function customType(customTypeParams) {
	return (a, b) => {
		const { name, config } = getColumnNameAndConfig(a, b);
		return new MySqlCustomColumnBuilder(name, config, customTypeParams);
	};
}

//#endregion
export { MySqlCustomColumn, MySqlCustomColumnBuilder, customType };
