// drizzle-orm@1.0.0-rc.4/mysql-core/columns/date.js
import { MySqlColumn, MySqlColumnBuilder } from "./common.js";
import { entityKind } from "../../entity.js";
import { getColumnNameAndConfig } from "../../utils.js";

//#region src/mysql-core/columns/date.ts
var MySqlDateBuilder = class extends MySqlColumnBuilder {
	static [entityKind] = "MySqlDateBuilder";
	constructor(name) {
		super(name, "object date", "MySqlDate");
	}
	/** @internal */
	build(table) {
		return new MySqlDate(table, this.config);
	}
};
var MySqlDate = class extends MySqlColumn {
	static [entityKind] = "MySqlDate";
	/** @internal */
	codec = "date";
	constructor(table, config) {
		super(table, config);
	}
	getSQLType() {
		return `date`;
	}
};
var MySqlDateStringBuilder = class extends MySqlColumnBuilder {
	static [entityKind] = "MySqlDateStringBuilder";
	constructor(name) {
		super(name, "string date", "MySqlDateString");
	}
	/** @internal */
	build(table) {
		return new MySqlDateString(table, this.config);
	}
};
var MySqlDateString = class extends MySqlColumn {
	static [entityKind] = "MySqlDateString";
	/** @internal */
	codec = "date:string";
	constructor(table, config) {
		super(table, config);
	}
	getSQLType() {
		return `date`;
	}
};
function date(a, b) {
	const { name, config } = getColumnNameAndConfig(a, b);
	if (config?.mode === "string") return new MySqlDateStringBuilder(name);
	return new MySqlDateBuilder(name);
}

//#endregion
export { MySqlDate, MySqlDateBuilder, MySqlDateString, MySqlDateStringBuilder, date };
