// drizzle-orm@1.0.0-rc.4/mysql-core/columns/timestamp.js
import { MySqlDateBaseColumn, MySqlDateColumnBaseBuilder } from "./date.common.js";
import { entityKind } from "../../entity.js";
import { getColumnNameAndConfig } from "../../utils.js";

//#region src/mysql-core/columns/timestamp.ts
var MySqlTimestampBuilder = class extends MySqlDateColumnBaseBuilder {
	static [entityKind] = "MySqlTimestampBuilder";
	constructor(name, config) {
		super(name, "object date", "MySqlTimestamp");
		this.config.fsp = config?.fsp;
	}
	/** @internal */
	build(table) {
		return new MySqlTimestamp(table, this.config);
	}
};
var MySqlTimestamp = class extends MySqlDateBaseColumn {
	static [entityKind] = "MySqlTimestamp";
	/** @internal */
	codec = "timestamp";
	fsp = this.config.fsp;
	getSQLType() {
		return `timestamp${this.fsp === void 0 ? "" : `(${this.fsp})`}`;
	}
	mapToDriverValue = (value) => {
		if (typeof value === "string") return value;
		return value.toISOString().slice(0, -1).replace("T", " ");
	};
};
var MySqlTimestampStringBuilder = class extends MySqlDateColumnBaseBuilder {
	static [entityKind] = "MySqlTimestampStringBuilder";
	constructor(name, config) {
		super(name, "string timestamp", "MySqlTimestampString");
		this.config.fsp = config?.fsp;
	}
	/** @internal */
	build(table) {
		return new MySqlTimestampString(table, this.config);
	}
};
var MySqlTimestampString = class extends MySqlDateBaseColumn {
	static [entityKind] = "MySqlTimestampString";
	/** @internal */
	codec = "timestamp:string";
	fsp = this.config.fsp;
	getSQLType() {
		return `timestamp${this.fsp === void 0 ? "" : `(${this.fsp})`}`;
	}
	mapToDriverValue = (value) => {
		if (typeof value === "string") return value;
		return value.toISOString().slice(0, -1).replace("T", " ");
	};
};
function timestamp(a, b = {}) {
	const { name, config } = getColumnNameAndConfig(a, b);
	if (config?.mode === "string") return new MySqlTimestampStringBuilder(name, config);
	return new MySqlTimestampBuilder(name, config);
}

//#endregion
export { MySqlTimestamp, MySqlTimestampBuilder, MySqlTimestampString, MySqlTimestampStringBuilder, timestamp };
