// drizzle-orm@1.0.0-rc.4/effect-pglite/codecs.js
import { effectPgCodecs } from "../effect-postgres/codecs.js";
import { base64ToUint8Array } from "../utils.js";
import { refineCodecs } from "../codecs.js";
import { makePgArray, parsePgArray } from "../pg-core/array.js";
import { arrayCompatNormalize, castToText, castToTextArr, genericPgCodecs, parseGeometryTuple, parseGeometryXY, parsePgArrayAndNormalize } from "../pg-core/codecs.js";

//#region src/effect-pglite/codecs.ts
const effectPgliteCodecs = refineCodecs(effectPgCodecs, {
	bigint: {
		cast: castToText,
		castArray: castToTextArr,
		normalize: BigInt,
		normalizeArray: arrayCompatNormalize(BigInt)
	},
	"bigint:string": {
		cast: castToText,
		castArray: castToTextArr
	},
	"bigint:number": {
		cast: castToText,
		castArray: castToTextArr
	},
	bigserial: {
		normalize: BigInt,
		normalizeArray: arrayCompatNormalize(BigInt),
		cast: castToText,
		castArray: castToTextArr
	},
	"bigserial:number": {
		cast: castToText,
		castArray: castToTextArr
	},
	bytea: {
		normalizeInJson: typeof Buffer === "undefined" ? base64ToUint8Array : genericPgCodecs.bytea?.normalizeInJson,
		normalizeArrayInJson: typeof Buffer === "undefined" ? arrayCompatNormalize(base64ToUint8Array) : genericPgCodecs.bytea?.normalizeArrayInJson,
		normalize: typeof Buffer === "undefined" ? void 0 : (v) => Buffer.from(v),
		normalizeArray: typeof Buffer === "undefined" ? void 0 : arrayCompatNormalize((v) => Buffer.from(v))
	},
	"geometry(point)": {
		normalizeArray: parsePgArrayAndNormalize(parseGeometryXY),
		castParam: (name) => `${name}::geometry`,
		castArrayParam: (name, _column, dimensions) => `${name}::geometry${"[]".repeat(dimensions)}`,
		normalizeParamArray: makePgArray
	},
	"geometry(point):tuple": {
		normalizeArray: parsePgArrayAndNormalize(parseGeometryTuple),
		castParam: (name) => `${name}::geometry`,
		castArrayParam: (name, _column, dimensions) => `${name}::geometry${"[]".repeat(dimensions)}`,
		normalizeParamArray: makePgArray
	},
	bit: { normalizeArray: void 0 },
	halfvec: {
		castParam: (name) => `${name}::halfvec`,
		castArrayParam: (name, _column, dimensions) => `${name}::halfvec${"[]".repeat(dimensions)}`,
		normalizeParamArray: makePgArray
	},
	vector: {
		castParam: (name) => `${name}::vector`,
		castArrayParam: (name, _column, dimensions) => `${name}::vector${"[]".repeat(dimensions)}`,
		normalizeParamArray: makePgArray
	},
	sparsevec: {
		normalizeArray: parsePgArray,
		castParam: (name) => `${name}::sparsevec`,
		castArrayParam: (name, _column, dimensions) => `${name}::sparsevec${"[]".repeat(dimensions)}`,
		normalizeParamArray: makePgArray
	}
});

//#endregion
export { effectPgliteCodecs };
