// drizzle-orm@1.0.0-rc.4/effect-pglite/driver.js
import { effectPgliteCodecs } from "./codecs.js";
import { EffectPgSession } from "./session.js";
import { entityKind } from "../entity.js";
import { jitCompatCheck } from "../utils.js";
import { PgDialect } from "../pg-core/dialect.js";
import * as Effect from "effect/Effect";
import { EffectCache } from "../cache/core/cache-effect.js";
import { DefaultServices, DefaultServices as DefaultServices$1 } from "../effect-core/defaults.js";
import { EffectLogger } from "../effect-core/index.js";
import { PgEffectDatabase } from "../pg-core/effect/db.js";
import { PgliteClient } from "@effect/sql-pglite/PgliteClient";

//#region src/effect-pglite/driver.ts
var EffectPgDatabase = class extends PgEffectDatabase {
	static [entityKind] = "EffectPgDatabase";
};
/**
* Creates an EffectPgDatabase instance.
*
* Requires `PgliteClient`, `EffectLogger`, and `EffectCache` services to be provided.
* Use `DefaultServices` to provide default (no-op) logger and cache implementations.
*
* @example
* ```ts
* // With default services (no logging, no caching)
* const db = yield* PgDrizzle.make({ relations }).pipe(
*   Effect.provide(PgDrizzle.DefaultServices),
* );
*
* // With Effect-based logging
* const db = yield* PgDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layer),
*   Effect.provide(PgDrizzle.DefaultServices),
* );
*
* // With custom Drizzle logger
* const db = yield* PgDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layerFromDrizzle(myLogger)),
*   Effect.provide(PgDrizzle.DefaultServices),
* );
* ```
*/
const make = Effect.fn("PgDrizzle.make")(function* (config = {}) {
	const client = yield* PgliteClient;
	const cache = yield* EffectCache;
	const logger = yield* EffectLogger;
	const dialect = new PgDialect({
		useJitMappers: jitCompatCheck(config.jit),
		codecs: config.codecs ?? effectPgliteCodecs
	});
	const relations = config.relations ?? {};
	const db = new EffectPgDatabase(dialect, new EffectPgSession(client, dialect, relations, {
		logger,
		cache
	}), relations);
	db.$client = client;
	db.$cache = cache;
	if (db.$cache) db.$cache["invalidate"] = cache.onMutate;
	return db;
});
/**
* Convenience function that creates an EffectPgDatabase with `DefaultServices` already provided.
*/
const makeWithDefaults = (config = {}) => make(config).pipe(Effect.provide(DefaultServices$1));

//#endregion
export { DefaultServices, EffectPgDatabase, make, makeWithDefaults };
