// drizzle-orm@1.0.0-rc.4/effect-d1/driver.js
import { EffectSQLiteD1Session } from "./session.js";
import { entityKind } from "../entity.js";
import { SQLiteDialect } from "../sqlite-core/dialect.js";
import * as Effect from "effect/Effect";
import { EffectCache } from "../cache/core/cache-effect.js";
import { DefaultServices, DefaultServices as DefaultServices$1 } from "../effect-core/defaults.js";
import { EffectLogger } from "../effect-core/index.js";
import { SQLiteEffectDatabase } from "../sqlite-core/effect/db.js";
import { D1Client } from "@effect/sql-d1/D1Client";

//#region src/effect-d1/driver.ts
var EffectSQLiteD1Database = class extends SQLiteEffectDatabase {
	static [entityKind] = "EffectSQLiteD1Database";
};
/**
* Creates an EffectSQLiteD1Database instance.
*
* Requires `D1Client`, `EffectLogger`, and `EffectCache` services to be provided.
* Use `DefaultServices` to provide default (no-op) logger and cache implementations.
*
* @example
* ```ts
* // With default services (no logging, no caching)
* const db = yield* SQLiteD1Drizzle.make({ relations }).pipe(
*   Effect.provide(SQLiteD1Drizzle.DefaultServices),
* );
*
* // With Effect-based logging
* const db = yield* SQLiteD1Drizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layer),
*   Effect.provide(SQLiteD1Drizzle.DefaultServices),
* );
*
* // With custom Drizzle logger
* const db = yield* SQLiteD1Drizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layerFromDrizzle(myLogger)),
*   Effect.provide(SQLiteD1Drizzle.DefaultServices),
* );
* ```
*/
const make = Effect.fn("SQLiteD1Drizzle.make")(function* (config) {
	const client = yield* D1Client;
	const cache = yield* EffectCache;
	const logger = yield* EffectLogger;
	const dialect = new SQLiteDialect();
	const relations = config.relations ?? {};
	const db = new EffectSQLiteD1Database(dialect, new EffectSQLiteD1Session(client, dialect, relations, {
		logger,
		cache
	}), relations, true);
	db.$client = client;
	db.$cache = cache;
	if (db.$cache) db.$cache["invalidate"] = cache.onMutate;
	return db;
});
/**
* Convenience function that creates an EffectSQLiteD1Database with `DefaultServices` already provided.
*/
const makeWithDefaults = (config) => make(config).pipe(Effect.provide(DefaultServices$1));

//#endregion
export { DefaultServices, EffectSQLiteD1Database, make, makeWithDefaults };
