// drizzle-orm@1.0.0-rc.4/bun-sqlite/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");
let __errors_ts = require("../errors.cjs");
let __logger_ts = require("../logger.cjs");
let __sqlite_core_async_session_ts = require("../sqlite-core/async/session.cjs");

//#region src/bun-sqlite/session.ts
var SQLiteBunSession = class extends __sqlite_core_async_session_ts.SQLiteAsyncSession {
	static [__entity_ts.entityKind] = "SQLiteBunSession";
	logger;
	constructor(client, dialect, relations, options = {}) {
		super(dialect, "sync");
		this.client = client;
		this.relations = relations;
		this.options = options;
		this.logger = options.logger ?? new __logger_ts.NoopLogger();
	}
	exec(query) {
		this.client.exec(query);
	}
	prepareQuery(query, mode, _prepare, executeMethod, mapper, queryMetadata) {
		let stmt;
		try {
			stmt = this.client.query(query.sql);
		} catch (e) {
			throw new __errors_ts.DrizzleQueryError(query.sql, query.params, e);
		}
		return new __sqlite_core_async_session_ts.SQLiteAsyncPreparedQuery("sync", executeMethod, {
			all: (params) => {
				if (mode === "arrays") return stmt.values(...params);
				return stmt.all(...params);
			},
			get: (params) => {
				if (mode === "arrays") return stmt.values(...params)[0];
				return stmt.get(...params);
			},
			run: (params) => {
				return stmt.run(...params);
			},
			values: (params) => {
				return stmt.values(...params);
			}
		}, query, mapper, mode, this.logger, void 0, queryMetadata, void 0);
	}
	transaction(transaction, config = {}) {
		const tx = new SQLiteBunTransaction("sync", this.dialect, this, this.relations);
		let result;
		this.client.transaction(() => {
			result = transaction(tx);
		})[config.behavior ?? "deferred"]();
		return result;
	}
};
var SQLiteBunTransaction = class SQLiteBunTransaction extends __sqlite_core_async_session_ts.SQLiteAsyncTransaction {
	static [__entity_ts.entityKind] = "SQLiteBunTransaction";
	transaction(transaction) {
		const savepointName = `sp${this.nestedIndex}`;
		const tx = new SQLiteBunTransaction("sync", this.dialect, this.session, this._.relations, this.nestedIndex + 1);
		this.session.run(__sql_sql_ts.sql.raw(`savepoint ${savepointName}`));
		try {
			const result = transaction(tx);
			this.session.run(__sql_sql_ts.sql.raw(`release savepoint ${savepointName}`));
			return result;
		} catch (err) {
			this.session.run(__sql_sql_ts.sql.raw(`rollback to savepoint ${savepointName}`));
			throw err;
		}
	}
};

//#endregion
exports.SQLiteBunSession = SQLiteBunSession;
exports.SQLiteBunTransaction = SQLiteBunTransaction;
