// vite-plugin-vercel@11.1.1/dist/api-DQ6GwdAx.js
//#region src/utils/assert.ts
function assert(condition, errorMessage) {
	if (condition) return;
	throw new Error(`[vite-plugin-vercel] ${errorMessage}`);
}
//#endregion
//#region src/api.ts
function createAPI(outfiles, pluginConfig) {
	return {
		/**
		* @internal
		*/
		getOutFiles() {
			return outfiles;
		},
		get config() {
			pluginConfig.config ??= {};
			return pluginConfig.config;
		},
		get defaultMaxDuration() {
			return pluginConfig.defaultMaxDuration;
		},
		set defaultMaxDuration(value) {
			pluginConfig.defaultMaxDuration = value;
		},
		get expiration() {
			return pluginConfig.expiration;
		},
		set expiration(value) {
			pluginConfig.expiration = value;
		},
		get rewrites() {
			pluginConfig.rewrites ??= [];
			return pluginConfig.rewrites;
		},
		get headers() {
			pluginConfig.headers ??= [];
			return pluginConfig.headers;
		},
		get redirects() {
			pluginConfig.redirects ??= [];
			return pluginConfig.redirects;
		},
		get cleanUrls() {
			return pluginConfig.cleanUrls;
		},
		set cleanUrls(value) {
			pluginConfig.cleanUrls = value;
		},
		get trailingSlash() {
			return pluginConfig.trailingSlash;
		},
		set trailingSlash(value) {
			pluginConfig.trailingSlash = value;
		},
		get defaultSupportsResponseStreaming() {
			return pluginConfig.defaultSupportsResponseStreaming;
		},
		set defaultSupportsResponseStreaming(value) {
			pluginConfig.defaultSupportsResponseStreaming = value;
		}
	};
}
function getVercelAPI(pluginContextOrServer) {
	const vpv = ("environment" in pluginContextOrServer ? pluginContextOrServer.environment.config : pluginContextOrServer.config).plugins.find((p) => p.name === "vite-plugin-vercel:api");
	assert(vpv, "Could not find vite-plugin-vercel:api plugin");
	assert(vpv.api, "Missing `api`. Make sure vite-plugin-vercel is up-to-date");
	return vpv.api("environment" in pluginContextOrServer ? pluginContextOrServer : void 0);
}
//#endregion
export { getVercelAPI as n, assert as r, createAPI as t };
