// vite-plugin-vercel@11.1.1/dist/index.js
import { t as pathRelativeTo } from "./path-5p9XRHnD.js";
import path from "node:path";
import { vercelEndpointExports } from "@vite-plugin-vercel/schemas";
import { fromNextFs } from "convert-route/next-fs";
import { toRou3 } from "convert-route/rou3";
import glob from "fast-glob";
import { generateCode, loadFile } from "magicast";
import { normalizePath } from "vite";
//#region src/index.ts
/**
* Scans the filesystem for entry points.
* @beta
*/
async function getVercelEntries(dir, { destination = dir, tryParseExports = true }) {
	const normalizedDir = normalizePath(dir);
	destination = normalizePath(destination);
	const apiEntries = glob.sync(`${path.posix.resolve(normalizedDir)}/**/*.?(m)[jt]s?(x)`).filter((filepath) => !path.basename(filepath).startsWith("_"));
	const entryPoints = [];
	for (const filePath of apiEntries) {
		const outFilePath = pathRelativeTo(filePath, normalizedDir);
		const parsed = path.posix.parse(outFilePath);
		let xports;
		if (tryParseExports) xports = await extractExports(filePath);
		const key = path.posix.join(destination, parsed.dir, parsed.name);
		const entry = {
			id: filePath,
			vercel: {}
		};
		if (xports?.edge) entry.vercel.edge = xports.edge;
		if (xports?.isr) entry.vercel.isr = xports.isr;
		if (xports?.headers) entry.vercel.headers = xports.headers;
		if (xports?.streaming) entry.vercel.streaming = xports.streaming;
		entry.route = entryToRou3(key);
		if (key.includes("[...")) entry.vercel.enforce = "post";
		entryPoints.push(entry);
	}
	return entryPoints;
}
function entryToRou3(key) {
	return toRou3(fromNextFs(path.posix.resolve("/", key)))[0].replace(/\/index$/, "/");
}
async function extractExports(filepath) {
	try {
		const mod = await loadFile(filepath);
		const subject = {
			edge: evalExport(mod.exports.edge),
			headers: evalExport(mod.exports.headers),
			streaming: evalExport(mod.exports.streaming),
			isr: evalExport(mod.exports.isr)
		};
		return vercelEndpointExports.parse(subject);
	} catch (e) {
		console.warn(`Warning: failed to read exports of '${filepath}'`, e);
		return null;
	}
}
function isPrimitive(test) {
	return test !== Object(test);
}
function _eval(code) {
	return new Function(`{ return function(){ return ${code} } };`).call(null).call(null);
}
function evalExport(exp) {
	if (!exp) return;
	return _eval(isPrimitive(exp) ? exp : generateCode(exp).code);
}
//#endregion
export { extractExports, getVercelEntries };
