// vite-plugin-vercel@11.1.1/dist/vite.js
import { t as pathRelativeTo$1 } from "./path-5p9XRHnD.js";
import { n as getVercelAPI, r as assert, t as createAPI } from "./api-DQ6GwdAx.js";
import { builtinModules } from "node:module";
import path from "node:path";
import { vercelOutputConfigSchema, vercelOutputPrerenderConfigSchema, vercelOutputVcConfigSchema } from "@vite-plugin-vercel/schemas";
import { fromRou3 } from "convert-route/rou3";
import { BuildEnvironment, createRunnableDevEnvironment, mergeConfig, normalizePath } from "vite";
import { catchAll, devServer } from "@universal-deploy/vite";
import { addEntry, getAllEntries } from "@universal-deploy/store";
import fs, { existsSync, readFileSync, writeFileSync } from "node:fs";
import { copyFile, cp, rmdir, unlink } from "node:fs/promises";
import { findRoot } from "@manypkg/find-root";
import { nodeFileTrace, resolve } from "@vercel/nft";
import { transform } from "oxc-transform";
import { build } from "rolldown";
import { cpus } from "node:os";
import { externals } from "nf3/plugin";
import pLimit from "p-limit";
import { getNodeVersion } from "@vercel/build-utils";
import { toPathToRegexpV6 } from "convert-route/path-to-regexp-v6";
import { getTransformedRoutes, mergeRoutes, normalizeRoutes } from "@vercel/routing-utils";
//#region src/utils/buildEnvs.ts
function getBuildEnvNames(pluginConfig) {
	return {
		client: pluginConfig.viteEnvNames?.client ?? "vercel_client",
		edge: pluginConfig.viteEnvNames?.edge ?? "vercel_edge",
		node: pluginConfig.viteEnvNames?.node ?? "vercel_node"
	};
}
//#endregion
//#region src/utils/dedupeRoutes.ts
/**
* When multiple entries point to the same module, we can deploy them as a single function.
* Create a separate function only when specific configuration is provided (`isr`, `headers`, `edge` or `streaming`).
*/
function dedupeRoutes() {
	const entriesToKeep = [];
	const entriesGroupedByModuleId = groupBy(getAllEntries(), (e) => e.id);
	for (const entries of entriesGroupedByModuleId.values()) {
		let groupedEntry;
		for (const entry of entries) if (entry.vercel && Object.keys(entry.vercel).length > 0) {
			if (!Array.isArray(entry.route)) entry.route = [entry.route];
			entriesToKeep.push(entry);
		} else if (!groupedEntry) {
			groupedEntry = structuredClone(entry);
			if (!Array.isArray(groupedEntry.route)) groupedEntry.route = [groupedEntry.route];
			entriesToKeep.push(groupedEntry);
		} else groupedEntry.route.push(...[entry.route].flat());
	}
	return entriesToKeep;
}
/**
* Sorts routes by specificity.
*
* Specificity rules:
* 1. Segment type priority at each position: Static > Dynamic/Param > Catch-all.
* 2. Total route length: Longer routes are generally more specific.
* 3. Prefix match exception: A shorter static route is more specific than a longer route
*    that only adds a catch-all segment to the same prefix.
* 4. Alphabetical order: Final tie-breaker for static segments at the same position.
*
* @param routes - Array of routes to sort.
* @returns A new array of routes sorted from most specific to least specific.
*/
function sortRoutes(routes) {
	return [...routes].sort((a, b) => {
		const lenA = a.pathname.length;
		const lenB = b.pathname.length;
		const commonLen = Math.min(lenA, lenB);
		for (let i = 0; i < commonLen; i++) {
			const segA = a.pathname[i];
			const segB = b.pathname[i];
			if (segA.value && !segB.value) return -1;
			if (!segA.value && segB.value) return 1;
			if (segA.value && segB.value) {
				if (segA.value !== segB.value) {
					if (lenA !== lenB) return lenB - lenA;
					return segA.value.localeCompare(segB.value);
				}
			}
			if (!segA.optional && segB.optional) return -1;
			if (segA.optional && !segB.optional) return 1;
			if (segA.catchAll && segB.catchAll) {
				if (!segA.catchAll.greedy && segB.catchAll.greedy) return -1;
				if (segA.catchAll.greedy && !segB.catchAll.greedy) return 1;
			}
			if (segA.catchAll && !segB.catchAll) return 1;
			if (!segA.catchAll && segB.catchAll) return -1;
		}
		if (lenA !== lenB) {
			if (lenA > lenB) return a.pathname[lenB].catchAll ? 1 : -1;
			return b.pathname[lenA].catchAll ? -1 : 1;
		}
		return 0;
	});
}
function groupBy(list, fn, selector) {
	return Array.from(list).reduce((acc, curr) => {
		const key = fn(curr);
		if (!acc.has(key)) acc.set(key, []);
		if (selector) acc.get(key).push(...selector(curr));
		else acc.get(key).push(curr);
		return acc;
	}, /* @__PURE__ */ new Map());
}
//#endregion
//#region src/utils/extension.ts
function removeExtension$1(subject) {
	return subject.replace(/\.[^/.]+$/, "");
}
//#endregion
//#region src/utils/destination.ts
function parseViteId(id) {
	const [path, query] = id.split("?");
	return {
		path: decodeURIComponent(path),
		query
	};
}
function extractBestPath({ path, query }) {
	if (!query) return path;
	return query.split("&").map(decodeURIComponent).find((p) => p.includes("/")) || path;
}
function shortenPath(path, maxSegments = 2) {
	return removeExtension$1(path.split(/[\\/]/).filter(Boolean).slice(-maxSegments).join("_"));
}
function mapPath(path, maxSegments = 2) {
	const nodeModuleMatch = path.match(/node_modules[\\/](.+?)([\\/].*)?$/);
	if (nodeModuleMatch) return `${nodeModuleMatch[1].split(/[\\/]/)[0]}_${removeExtension$1(nodeModuleMatch[2]?.split(/[\\/]/).pop() || "index")}`;
	return shortenPath(path, maxSegments);
}
function hashString(str, length = 6) {
	let hash = 5381;
	for (let i = 0; i < str.length; i++) hash = hash * 33 ^ str.charCodeAt(i);
	hash = hash >>> 0;
	return hash.toString(36).slice(0, length);
}
function uniqueViteName(id, entry, maxSegments = 2) {
	return removeExtension$1(`${mapPath(extractBestPath(parseViteId(id)), maxSegments)}_${hashString(JSON.stringify(entry))}`.replace(/[^a-zA-Z0-9\-_[\]/]/g, "-"));
}
function entryDestinationDefault(root, entry) {
	return uniqueViteName(pathRelativeTo$1(entry.id, root), entry);
}
function entryDestination(root, entry, postfix) {
	return `${path.posix.join("functions/", entryDestinationDefault(root, entry))}${postfix}`;
}
//#endregion
//#region src/plugins/api.ts
function apiPlugin(pluginConfig) {
	const envNames = getBuildEnvNames(pluginConfig);
	const outfiles = [];
	return {
		name: "vite-plugin-vercel:api",
		api() {
			return createAPI(outfiles, pluginConfig);
		},
		applyToEnvironment({ name }) {
			return name === envNames.edge || name === envNames.node;
		},
		writeBundle(_opts, bundle) {
			const root = this.environment.config.root ?? process.cwd();
			const entryMapByDestination = new Map(dedupeRoutes().map((e) => [entryDestination(root, e, ".func/index"), e]));
			for (const [key, value] of Object.entries(bundle)) if (value.type === "chunk" && entryMapByDestination.has(removeExtension$1(key))) outfiles.push({
				type: "chunk",
				root: this.environment.config.root,
				outdir: this.environment.config.build.outDir,
				filepath: key,
				relatedEntry: entryMapByDestination.get(removeExtension$1(key))
			});
			else if (value.type === "asset" && key.startsWith("functions/") || key === "config.json") outfiles.push({
				type: "asset",
				root: this.environment.config.root,
				outdir: this.environment.config.build.outDir,
				filepath: key
			});
		},
		sharedDuringBuild: true
	};
}
//#endregion
//#region src/utils/edge.ts
const edgeConditions = [
	"edge-light",
	"worker",
	"browser",
	"module",
	"import",
	"default"
];
//#endregion
//#region src/utils/external.ts
const _external = [
	"async_hooks",
	"events",
	"buffer",
	"assert",
	"util"
];
const edgeExternal = [..._external, ..._external.map((e) => `node:${e}`)];
//#endregion
//#region src/plugins/bundle/basic.ts
const builtIns = new Set(builtinModules.flatMap((m) => [m, `node:${m}`]));
const edgeWasmPlugin = {
	name: "edge-wasm-vercel",
	resolveId: {
		filter: { id: [/\.wasm\?module$/] },
		handler(id) {
			return {
				id: id.replace(/\.wasm\?module$/, ".wasm"),
				external: true
			};
		}
	}
};
const dynamicNativeImportPlugin = {
	name: "edge-dynamic-import-native",
	resolveDynamicImport(specifier) {
		if (typeof specifier === "string" && builtIns.has(specifier)) return {
			id: specifier,
			external: true
		};
	}
};
const reactEdgePlugin$1 = {
	name: "react-edge-plugin",
	resolveId: {
		filter: { id: [/react-dom\/server/] },
		async handler(_id, importer, options) {
			return this.resolve("react-dom/server.edge", importer, options);
		}
	}
};
function basicBundlePlugin(pluginConfig) {
	const envNames = getBuildEnvNames(pluginConfig);
	const bundledAssets = /* @__PURE__ */ new Map();
	const bundledChunks = [];
	return [{
		name: "vite-plugin-vercel:bundle",
		enforce: "post",
		apply: "build",
		generateBundle(_opts, bundle) {
			for (const b of Object.values(bundle)) {
				const outFile = joinAbsolute(this.environment, this.environment.config.build.outDir, b.fileName);
				if (b.type === "asset") {
					const originalFileNames = b.originalFileNames.map((relativePath) => path.resolve(this.environment.config.root, relativePath));
					const asset = {
						env: this.environment.name,
						root: this.environment.config.root,
						outDir: this.environment.config.build.outDir,
						outFile,
						fileName: b.fileName
					};
					for (const originalFileName of originalFileNames) bundledAssets.set(originalFileName, asset);
				} else if (this.environment.name === envNames.edge || this.environment.name === envNames.node) bundledChunks.push(outFile);
			}
		},
		closeBundle: {
			order: "post",
			async handler() {
				if (!isVercelLastBuildStep(this.environment, pluginConfig)) return;
				this.environment.logger.info("Creating Vercel bundles...");
				const outfiles = getVercelAPI(this).getOutFiles();
				const filesToKeepP = [];
				for (const outfile of outfiles) if (outfile.type === "chunk") filesToKeepP.push(bundle$1(this, bundledAssets, outfile));
				await cleanup$1((await Promise.all(filesToKeepP)).flat(), bundledChunks);
			}
		},
		sharedDuringBuild: true
	}];
}
function getAbsoluteOutFile(outfile) {
	const source = joinAbsolutePosix(outfile.root, outfile.outdir, outfile.filepath);
	return {
		source,
		destination: source.replace(outfile.outdir, outfile.outdir)
	};
}
async function bundle$1(pluginContext, bundledAssets, outfile) {
	const output = [];
	const { source, destination } = getAbsoluteOutFile(outfile);
	const isEdge = Boolean(outfile.relatedEntry.vercel?.edge);
	const { environment } = pluginContext;
	const buildOptions = {};
	buildOptions.output = {
		format: "esm",
		comments: { legal: false },
		codeSplitting: false
	};
	buildOptions.checks = { pluginTimings: false };
	buildOptions.input = [source];
	buildOptions.treeshake = true;
	buildOptions.resolve ??= {};
	if (isEdge) {
		buildOptions.platform = "browser";
		buildOptions.external = edgeExternal;
		buildOptions.resolve.conditionNames = edgeConditions;
		buildOptions.transform = { define: { "process.env.NODE_ENV": JSON.stringify("production") } };
		buildOptions.output.file = destination.replace(/\.mjs$/, ".js");
		buildOptions.plugins = [
			edgeWasmPlugin,
			dynamicNativeImportPlugin,
			reactEdgePlugin$1
		];
	} else {
		buildOptions.platform = "node";
		buildOptions.output.file = destination.replace(/\.js$/, ".mjs");
	}
	try {
		await build(buildOptions);
		output.push(buildOptions.output.file);
	} catch (e) {
		throw new Error(`Error while bundling ${destination}`, { cause: e });
	}
	let base = environment.config.root;
	try {
		base = (await findRoot(environment.config.root)).rootDir;
	} catch (_e) {}
	const resolvedEntryP = pluginContext.resolve(outfile.relatedEntry.id);
	const entryFilePath = existsSync(outfile.relatedEntry.id) ? outfile.relatedEntry.id : (await resolvedEntryP)?.id && existsSync((await resolvedEntryP).id) ? (await resolvedEntryP).id : null;
	if (entryFilePath === null) return [];
	const { fileList, reasons } = await nodeFileTrace([entryFilePath], {
		base,
		processCwd: environment.config.root,
		mixedModules: true,
		ignore: [
			"**/node_modules/react{,-dom,-dom-server-turbopack}/**/*.development.js",
			"**/*.d.ts",
			"**/*.map",
			"**/node_modules/webpack5/**/*"
		],
		async readFile(filepath) {
			const code = readFileSync(filepath, "utf-8");
			if (filepath.endsWith(".ts") || filepath.endsWith(".tsx")) return (await transform(filepath, code, { target: "es2022" })).code;
			return code;
		}
	});
	for (const file of fileList) if (reasons.has(file) && reasons.get(file)?.type.includes("asset") && !file.endsWith(".js") && !file.endsWith(".cjs") && !file.endsWith(".mjs") && !file.endsWith("package.json")) {
		const absolutePath = path.join(base, file);
		const assetRenamedByVite = bundledAssets.get(absolutePath);
		const basename = path.basename(assetRenamedByVite ? assetRenamedByVite.fileName : absolutePath);
		if (assetRenamedByVite) writeFileSync(destination, readFileSync(destination, "utf-8").replaceAll(`/${assetRenamedByVite.fileName}`, `./${path.basename(assetRenamedByVite.fileName)}`));
		const to = path.join(path.dirname(destination), basename);
		output.push(to);
		await copyFile(absolutePath, to);
	}
	return output;
}
function joinAbsolute(env_or_p0, p1, ...p) {
	if (path.isAbsolute(p1)) return path.join(p1, ...p);
	return path.join(typeof env_or_p0 === "string" ? env_or_p0 : env_or_p0.config.root, p1, ...p);
}
function joinAbsolutePosix(env_or_p0, p1, ...p) {
	if (path.isAbsolute(p1)) return path.posix.join(p1, ...p);
	return path.posix.join(typeof env_or_p0 === "string" ? env_or_p0 : env_or_p0.config.root, p1, ...p);
}
function isVercelLastBuildStep(env, pluginConfig) {
	const envNames = getBuildEnvNames(pluginConfig);
	return (typeof env !== "string" ? env.name : env) === envNames.node;
}
async function cleanup$1(filesToKeep, bundledChunks) {
	const toKeep = new Set(filesToKeep);
	const removedFiles = [];
	await Promise.all(bundledChunks.map(async (file) => {
		if (!toKeep.has(file)) try {
			await unlink(file);
			removedFiles.push(file);
		} catch {}
	}));
	const dirsToRemove = /* @__PURE__ */ new Set();
	for (const file of removedFiles) {
		let dir = path.dirname(file);
		while (dir && dir !== "." && dir !== "/" && !toKeep.has(dir)) {
			dirsToRemove.add(dir);
			dir = path.dirname(dir);
		}
	}
	const sortedDirs = Array.from(dirsToRemove).sort((a, b) => b.length - a.length);
	for (const dir of sortedDirs) try {
		await rmdir(dir);
	} catch {}
}
//#endregion
//#region src/plugins/bundle/nf3.ts
function nf3BundlePlugin(pluginConfig) {
	console.warn("[vite-plugin-vercel] bundleStrategy 'nf3' is highly experimental. Use at your own risk; support is not provided for related issues.");
	const envNames = getBuildEnvNames(pluginConfig);
	const externalsPlugin = externals({});
	delete externalsPlugin.buildEnd;
	let buildOutput;
	return [{
		...externalsPlugin,
		applyToEnvironment(env) {
			return env.config.consumer !== "client";
		}
	}, {
		name: "vite-plugin-vercel:isolate-functions",
		apply: "build",
		applyToEnvironment(env) {
			return env.config.consumer !== "client";
		},
		async writeBundle(_, output) {
			const isEdge = this.environment.name === envNames.edge;
			const config = this.environment.config;
			const outDir = normalizePath(path.isAbsolute(config.build.outDir) ? config.build.outDir : path.join(config.root, config.build.outDir));
			const entries = Object.entries(output).filter((e) => "isEntry" in e[1] && e[1].isEntry).map((e) => ({
				name: e[1].name,
				fileName: e[1].fileName,
				outPath: path.join(outDir, e[1].fileName)
			}));
			if (entries.length === 0) return;
			const outPaths = entries.map((entry) => entry.outPath);
			const input = Object.fromEntries(outPaths.map((e) => [removeExtension(pathRelativeTo(e, outDir)), e]));
			const limit = pLimit(Math.max(1, Math.ceil(cpus().length / 2)));
			const nonVitePlugins = this.environment.config.plugins.filter((p) => {
				return !p.name.startsWith("vite:") && p.name !== "alias" && p.name !== "commonjs" && p.name !== "nitro:externals";
			}).map((x) => {
				const { buildStart, buildEnd, writeBundle, generateBundle, ...rest } = x;
				return rest;
			});
			const localOutput = (await Promise.all(Object.values(input).map((entryPath) => limit(async () => {
				const outDir = path.dirname(entryPath);
				return { output: (await bundle({
					plugins: nonVitePlugins,
					isEdge,
					input: { index: entryPath },
					outDir,
					externals: {
						conditions: this.environment.config.resolve.conditions,
						rootDir: this.environment.config.root,
						trace: {
							outDir: path.dirname(entryPath),
							nft: { async resolve(id, parent, job, cjsResolve) {
								return resolve(id.replace(/\.wasm\?module$/, ".wasm"), parent, job, cjsResolve);
							} }
						}
					}
				})).output.map((o) => ({
					...o,
					fileName: path.join(outDir, o.fileName)
				})) };
			})))).flatMap((r) => r.output);
			buildOutput = buildOutput ? [...buildOutput, ...localOutput] : localOutput;
			cleanup(output, buildOutput, this.environment.config.build.outDir);
		}
	}];
}
function bundle(options) {
	assert(options.input, "No input specified");
	return build({
		platform: options.isEdge ? "browser" : "node",
		external: options.isEdge ? edgeExternal : [],
		write: true,
		plugins: [...options.plugins, externals(options.externals)],
		input: options.input,
		resolve: { conditionNames: options.externals.conditions },
		output: {
			entryFileNames: options.isEdge ? "[name].js" : "[name].mjs",
			sanitizeFileName: false,
			dir: options.outDir,
			hoistTransitiveImports: false
		},
		checks: { pluginTimings: false }
	});
}
function cleanup(viteOutput, rolldownOutput, outputDir) {
	const viteFilenames = Object.values(viteOutput).map((o) => path.join(outputDir, o.fileName));
	const rolldownFilenames = new Set(rolldownOutput.map((o) => o.fileName));
	const rolldownModuleIds = new Set(rolldownOutput.flatMap((o) => o.type === "chunk" ? o.moduleIds : []));
	const filesToDelete = viteFilenames.filter((id) => id.startsWith(outputDir) && rolldownModuleIds.has(id) && !rolldownFilenames.has(id));
	const parentDirs = /* @__PURE__ */ new Set();
	for (const file of filesToDelete) try {
		fs.unlinkSync(file);
		const mapFile = `${file}.map`;
		if (fs.existsSync(mapFile)) fs.unlinkSync(mapFile);
		parentDirs.add(path.dirname(file));
	} catch {}
	for (const dir of parentDirs) try {
		if (fs.readdirSync(dir).length === 0) fs.rmdirSync(dir);
	} catch {}
}
function pathRelativeTo(filePath, rel) {
	return normalizePath(path.relative(normalizePath(path.resolve(rel)), path.resolve(filePath)));
}
function removeExtension(subject) {
	return subject.replace(/\.[^/.]+$/, "");
}
//#endregion
//#region src/plugins/clean-outdir.ts
const outDir$1 = path.posix.join(process.cwd(), ".vercel/output");
function vercelCleanupPlugin(pluginConfig) {
	let alreadyRun = false;
	return {
		apply: "build",
		name: "vite-plugin-vercel:cleanup",
		enforce: "pre",
		buildStart: {
			order: "pre",
			sequential: true,
			handler() {
				if (alreadyRun) return;
				alreadyRun = true;
				cleanOutputDirectory(pluginConfig.outDir ?? outDir$1);
			}
		},
		sharedDuringBuild: true
	};
}
function cleanOutputDirectory(outdir) {
	fs.rmSync(outdir, {
		recursive: true,
		force: true
	});
	fs.mkdirSync(outdir, { recursive: true });
}
//#endregion
//#region src/build.ts
function getVcConfig(pluginConfig, filename, options) {
	return vercelOutputVcConfigSchema.parse(options.edge ? {
		runtime: "edge",
		entrypoint: filename
	} : {
		runtime: options.nodeVersion.runtime,
		handler: filename,
		maxDuration: pluginConfig.defaultMaxDuration,
		launcherType: "Nodejs",
		shouldAddHelpers: false,
		supportsResponseStreaming: options.streaming ?? pluginConfig.defaultSupportsResponseStreaming ?? true,
		experimentalTriggers: options.experimentalTriggers
	});
}
//#endregion
//#region src/utils/request.ts
function getOriginalRequest(request) {
	const xOriginalPath = request.headers.get("x-original-path");
	let newUrl = null;
	let newRequest = request;
	if (typeof xOriginalPath === "string") {
		const requestUrl = new URL(request.url);
		const originalUrl = new URL(xOriginalPath, requestUrl);
		if (!originalUrl.search) originalUrl.search = requestUrl.search;
		newUrl = originalUrl.toString();
	}
	if (newUrl && request.url !== newUrl) {
		const requestInit = {
			method: request.method,
			headers: request.headers,
			body: request.body,
			mode: request.mode,
			credentials: request.credentials,
			cache: request.cache,
			redirect: request.redirect,
			referrer: request.referrer,
			integrity: request.integrity
		};
		if (request.body) requestInit.duplex = "half";
		newRequest = new Request(newUrl, requestInit);
	}
	return newRequest;
}
//#endregion
//#region src/plugins/loader.ts
const re_DUMMY = new RegExp(`__DUMMY__$`);
const re_edge = /[?&]vercel_edge\b/;
const re_node = /[?&]vercel_node\b/;
const queueConsumerEdgeError = "Vercel queue consumers must be serverless functions, not edge functions.";
function validateQueueConsumersAreServerless(entries) {
	for (const entry of entries) if (entry.vercel?.edge === true && (entry.vercel.experimentalTriggers?.length ?? 0) > 0) throw new Error(queueConsumerEdgeError);
}
function loaderPlugin(pluginConfig) {
	const envNames = getBuildEnvNames(pluginConfig);
	let root;
	return [
		{
			name: "vite-plugin-vercel:dummy",
			enforce: "pre",
			resolveId: {
				filter: { id: re_DUMMY },
				handler(id) {
					return id;
				}
			},
			load: {
				filter: { id: re_DUMMY },
				handler() {
					return "console.log('');export default {};";
				}
			}
		},
		{
			name: "vite-plugin-vercel:load-edge",
			apply: "build",
			resolveId: {
				filter: { id: re_edge },
				async handler(id, importer, opts) {
					const resolved = await this.resolve(id.replace(re_edge, ""), importer, opts);
					if (!resolved) return null;
					return `${resolved.id}?vercel_edge`;
				}
			},
			load: {
				filter: { id: re_edge },
				async handler(id) {
					const mod = id.replace(re_edge, "");
					return `import mod from ${JSON.stringify(mod)};

${getOriginalRequest.toString()}

const fn = (r) => {
  return mod.fetch(getOriginalRequest(r));
};

export default fn`;
				}
			}
		},
		{
			name: "vite-plugin-vercel:load-node",
			apply: "build",
			resolveId: {
				filter: { id: re_node },
				async handler(id, importer, opts) {
					const resolved = await this.resolve(id.replace(re_node, ""), importer, opts);
					if (!resolved) return null;
					return `${resolved.id}?vercel_node`;
				}
			},
			load: {
				filter: { id: re_node },
				async handler(id) {
					const mod = id.replace(re_node, "");
					return `import mod from ${JSON.stringify(mod)};

${getOriginalRequest.toString()}

if (mod?.fetch) {
  const ori = mod.fetch;
  mod.fetch = (r) => {
    return ori(getOriginalRequest(r));
  }
}

const def = mod?.server?.nodeHandler ?? mod;

export default def;`;
				}
			}
		},
		{
			name: "vite-plugin-vercel:build-functions",
			apply: "build",
			applyToEnvironment(env) {
				return env.name === envNames.node || env.name === envNames.edge;
			},
			config: {
				order: "post",
				handler(config) {
					root = config.root;
				}
			},
			configEnvironment: {
				order: "post",
				handler(name) {
					const isEdge = name === envNames.edge;
					if (name === envNames.node || isEdge) {
						const entries = dedupeRoutes();
						validateQueueConsumersAreServerless(entries);
						return { build: { rollupOptions: {
							input: Object.fromEntries(entries.filter((e) => (e.vercel?.edge ?? false) === isEdge).map((e) => [entryDestination(root ?? process.cwd(), e, ".func/index"), isEdge ? `${e.id}?vercel_edge` : `${e.id}?vercel_node`])),
							output: {
								hoistTransitiveImports: false,
								entryFileNames: "[name].js"
							}
						} } };
					}
				}
			},
			async buildStart() {
				const isEdge = this.environment.name === envNames.edge;
				const nodeVersion = await getNodeVersion(process.cwd());
				const entries = dedupeRoutes();
				for (const entry of entries.filter((e) => (e.vercel?.edge ?? false) === isEdge)) {
					const hasExperimentalTriggers = (entry.vercel?.experimentalTriggers?.length ?? 0) > 0;
					this.emitFile({
						type: "asset",
						fileName: entryDestination(root ?? process.cwd(), entry, ".func/.vc-config.json"),
						source: JSON.stringify(getVcConfig(pluginConfig, isEdge ? "index.js" : "index.mjs", {
							nodeVersion,
							edge: isEdge,
							streaming: entry.vercel?.streaming,
							experimentalTriggers: entry.vercel?.experimentalTriggers
						}), void 0, 2)
					});
					if (entry.vercel?.isr) this.emitFile({
						type: "asset",
						fileName: entryDestination(root ?? process.cwd(), entry, ".prerender-config.json"),
						source: JSON.stringify(vercelOutputPrerenderConfigSchema.parse(entry.vercel.isr), void 0, 2)
					});
					pluginConfig.rewrites ??= [];
					if (hasExperimentalTriggers) continue;
					for (const ir of sortRoutes([entry.route].flat().map((p) => fromRou3(p)))) {
						const source = toPathToRegexpV6(ir);
						pluginConfig.rewrites.push({
							enforce: entry.vercel?.enforce,
							source,
							destination: `/${entryDestinationDefault(root ?? process.cwd(), entry)}`
						});
						if (entry.vercel?.headers) {
							pluginConfig.headers ??= [];
							pluginConfig.headers.push({
								source,
								headers: Object.entries(entry.vercel.headers).map(([key, value]) => ({
									key,
									value
								}))
							});
						}
					}
				}
			},
			sharedDuringBuild: true
		}
	];
}
//#endregion
//#region src/plugins/react-edge.ts
function reactEdgePlugin(pluginConfig) {
	const envNames = getBuildEnvNames(pluginConfig);
	return {
		name: "vite-plugin-vercel:react-edge",
		applyToEnvironment(env) {
			return env.name === envNames.edge;
		},
		resolveId: {
			order: "pre",
			filter: { id: [/^react-dom\/server$/] },
			handler(_id, importer, opts) {
				return this.resolve("react-dom/server.edge", importer, opts);
			}
		}
	};
}
//#endregion
//#region src/config.ts
function reorderEnforce(arr) {
	return [
		...arr.filter((r) => r.enforce === "pre"),
		...arr.filter((r) => !r.enforce),
		...arr.filter((r) => r.enforce === "post")
	];
}
function getConfig(pluginConfig) {
	const _rewrites = [...pluginConfig.rewrites ?? []];
	const { routes, error } = getTransformedRoutes({
		cleanUrls: pluginConfig.cleanUrls ?? true,
		trailingSlash: pluginConfig.trailingSlash,
		rewrites: reorderEnforce(_rewrites.map((r) => {
			if (r.source === "/:_1*") r.source = "(.*)";
			return r;
		})),
		redirects: pluginConfig.redirects ? reorderEnforce(pluginConfig.redirects) : void 0,
		headers: [{
			source: "/(.*)",
			headers: [{
				key: "x-original-path",
				value: "/$1"
			}]
		}, ...pluginConfig.headers ?? []]
	});
	if (error) throw error;
	if (pluginConfig.config?.routes && pluginConfig.config.routes.length > 0 && !pluginConfig.config.routes.every((r) => "continue" in r && r.continue)) console.warn("Did you forget to add `\"continue\": true` to your routes? See https://vercel.com/docs/build-output-api/v3/configuration#source-route\nIf not, it is discouraged to use `routes` config to override routes. Prefer using `rewrites` and `redirects`.");
	let userRoutes = [];
	let buildRoutes = [];
	if (pluginConfig.config?.routes) {
		const norm = normalizeRoutes(pluginConfig.config.routes);
		if (norm.error) throw norm.error;
		userRoutes = norm.routes ?? [];
	}
	if (routes) {
		const norm = normalizeRoutes(routes);
		if (norm.error) throw norm.error;
		buildRoutes = norm.routes ?? [];
	}
	const cleanRoutes = mergeRoutes({
		userRoutes,
		builds: [{
			use: "@vercel/node",
			entrypoint: "index.mjs",
			routes: buildRoutes
		}]
	});
	return vercelOutputConfigSchema.parse({
		version: 3,
		...pluginConfig.config,
		routes: cleanRoutes,
		overrides: { ...pluginConfig.config?.overrides }
	});
}
//#endregion
//#region src/utils/const.ts
const virtualEntry = "virtual:vite-plugin-vercel:entry";
//#endregion
//#region src/plugins/setupEnvs.ts
const outDir = path.posix.join(process.cwd(), ".vercel/output");
const DUMMY = "__DUMMY__";
let injected = false;
function setupEnvs(pluginConfig) {
	const envNames = getBuildEnvNames(pluginConfig);
	return [
		{
			name: "vite-plugin-vercel:setup-envs",
			buildApp: {
				order: "post",
				async handler(builder) {
					const vercelClientEnv = builder.environments[envNames.client];
					if (!vercelClientEnv.isBuilt) try {
						await builder.build(vercelClientEnv);
					} catch (e) {
						if (!(e instanceof Error && e.message.includes(`Could not resolve entry module "index.html"`))) throw e;
					}
					await copyClientOutput(vercelClientEnv);
					if (envNames.edge !== false && !builder.environments[envNames.edge].isBuilt) await builder.build(builder.environments[envNames.edge]);
					if (!builder.environments[envNames.node].isBuilt) await builder.build(builder.environments[envNames.node]);
				}
			},
			config: {
				order: "post",
				handler() {
					if (!injected) {
						injected = true;
						if (pluginConfig.entries) pluginConfig.entries.forEach((entry) => {
							addEntry(entry);
						});
					}
					const outDirOverride = pluginConfig.outDir ? { build: { outDir: pluginConfig.outDir } } : {};
					const environments = {};
					if (envNames.client) environments[envNames.client] = {
						build: {
							outDir: path.join(pluginConfig.outDir ?? outDir, "static"),
							copyPublicDir: true,
							rollupOptions: { input: getDummyInput() },
							emptyOutDir: false
						},
						consumer: "client"
					};
					if (envNames.edge) environments[envNames.edge] = createVercelEnvironmentOptions(outDirOverride);
					if (envNames.node) environments[envNames.node] = createVercelEnvironmentOptions(outDirOverride);
					return {
						environments,
						builder: {}
					};
				}
			},
			sharedDuringBuild: true
		},
		{
			name: "vite-plugin-vercel:setup-envs:vercel_edge",
			applyToEnvironment(env) {
				return env.name === envNames.edge;
			},
			configEnvironment(name, config, env) {
				if (name !== envNames.edge) return;
				return {
					resolve: {
						external: edgeExternal,
						...!(env.command === "serve") ? { conditions: edgeConditions } : {}
					},
					build: {
						target: "es2022",
						rollupOptions: {
							input: {},
							treeshake: true
						},
						rolldownOptions: {}
					},
					optimizeDeps: {
						...config.optimizeDeps,
						...this.meta.rolldownVersion ? { rolldownOptions: {
							target: "es2022",
							format: "esm"
						} } : { esbuildOptions: {
							target: "es2022",
							format: "esm"
						} }
					}
				};
			},
			generateBundle: {
				order: "post",
				async handler(_opts, bundle) {
					cleanupDummy(bundle);
				}
			},
			sharedDuringBuild: true
		},
		{
			name: "vite-plugin-vercel:setup-envs:vercel_node",
			applyToEnvironment(env) {
				return env.name === envNames.node;
			},
			configEnvironment(name, config) {
				if (name !== envNames.node) return;
				return { optimizeDeps: { ...config.optimizeDeps } };
			},
			generateBundle: {
				order: "post",
				async handler(_opts, bundle) {
					cleanupDummy(bundle);
					this.emitFile({
						type: "asset",
						fileName: "config.json",
						source: JSON.stringify(getConfig(pluginConfig), void 0, 2)
					});
				}
			},
			sharedDuringBuild: true
		},
		{
			name: "vite-plugin-vercel:setup-envs:vercel_client",
			applyToEnvironment(env) {
				return env.name === envNames.client;
			},
			generateBundle: { async handler(_opts, bundle) {
				cleanupDummy(bundle);
			} },
			sharedDuringBuild: true
		}
	];
}
function createVercelEnvironmentOptions(overrides) {
	return mergeConfig({
		dev: { async createEnvironment(name, config) {
			return createRunnableDevEnvironment(name, config);
		} },
		build: {
			createEnvironment(name, config) {
				return new BuildEnvironment(name, config);
			},
			outDir,
			copyPublicDir: false,
			rollupOptions: {
				input: getDummyInput(),
				output: {
					sanitizeFileName: (filename) => {
						return filename.replace("\0", "_");
					},
					sourcemap: false
				}
			},
			target: "es2022",
			emptyOutDir: false,
			emitAssets: true
		},
		consumer: "server",
		keepProcessEnv: true
	}, overrides ?? {});
}
function getDummyInput() {
	return { [DUMMY]: `${virtualEntry}:${DUMMY}` };
}
function cleanupDummy(bundle) {
	const dummy = Object.keys(bundle).find((key) => key.includes("_DUMMY_"));
	if (dummy) delete bundle[dummy];
}
async function copyClientOutput(vercelClientEnv) {
	const topLevelConfig = vercelClientEnv.getTopLevelConfig();
	const srcDir = path.resolve(topLevelConfig.root, topLevelConfig.environments.client.build.outDir);
	const destDir = path.resolve(topLevelConfig.root, vercelClientEnv.config.build.outDir);
	if (srcDir === destDir) return;
	try {
		await cp(srcDir, destDir, {
			recursive: true,
			force: true,
			dereference: true
		});
	} catch (e) {
		if (!(e instanceof Error && "code" in e && e.code === "ENOENT")) throw e;
	}
}
//#endregion
//#region src/plugins/index.ts
function vercel(pluginConfig = {}) {
	return [
		reactEdgePlugin(pluginConfig),
		vercelCleanupPlugin(pluginConfig),
		apiPlugin(pluginConfig),
		...setupEnvs(pluginConfig),
		...loaderPlugin(pluginConfig),
		...pluginConfig?.bundleStrategy === "nf3" ? nf3BundlePlugin(pluginConfig) : basicBundlePlugin(pluginConfig),
		catchAll(),
		devServer()
	];
}
//#endregion
export { vercel as default, vercel };
