// drizzle-orm@1.0.0-rc.4/mysql-core/async/query.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_mysql_core_query_builders_query = require('../query-builders/query.cjs');
let __entity_ts = require("../../entity.cjs");
let __utils_ts = require("../../utils.cjs");
let __query_promise_ts = require("../../query-promise.cjs");

//#region src/mysql-core/async/query.ts
var MySqlAsyncRelationalQuery = class extends require_mysql_core_query_builders_query.MySqlRelationalQuery {
	static [__entity_ts.entityKind] = "MySqlAsyncRelationalQueryV2";
	prepare() {
		const { query, builtQuery } = this._toSQL();
		const mapper = this.dialect.mapperGenerators.relationalRows({
			isFirst: this.mode === "first",
			parseJson: false,
			parseJsonIfString: false,
			rootJsonMappers: false,
			arrayModeRoot: true,
			selection: query.selection
		});
		return this.session.prepareQuery(builtQuery, "arrays", mapper);
	}
	execute() {
		return this.prepare().execute();
	}
};
(0, __utils_ts.applyMixins)(MySqlAsyncRelationalQuery, [__query_promise_ts.QueryPromise]);

//#endregion
exports.MySqlAsyncRelationalQuery = MySqlAsyncRelationalQuery;
