// drizzle-orm@1.0.0-rc.4/mysql-core/async/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_mysql_core_async_db = require('./db.cjs');
const require_mysql_core_session = require('../session.cjs');
let __entity_ts = require("../../entity.cjs");
let __utils_ts = require("../../utils.cjs");
let __sql_sql_ts = require("../../sql/sql.cjs");
let __errors_ts = require("../../errors.cjs");
let __migrator_utils_ts = require("../../migrator.utils.cjs");
let __cache_core_cache_ts = require("../../cache/core/cache.cjs");
let __up_migrations_mysql_ts = require("../../up-migrations/mysql.cjs");

//#region src/mysql-core/async/session.ts
var MySqlAsyncPreparedQuery = class extends require_mysql_core_session.MySqlBasePreparedQuery {
	static [__entity_ts.entityKind] = "MySqlAsyncPreparedQuery";
	/** @internal */
	mapper;
	fastPath;
	constructor(executor, _iterator, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(query);
		this.executor = executor;
		this._iterator = _iterator;
		this.mode = mode;
		this.logger = logger;
		this.cache = cache;
		this.queryMetadata = queryMetadata;
		this.cacheConfig = cacheConfig;
		this.mapper = mapper;
		if (cache && cache.strategy() === "all" && cacheConfig === void 0) this.cacheConfig = {
			enabled: true,
			autoInvalidate: true
		};
		if (!this.cacheConfig?.enabled) this.cacheConfig = void 0;
		this.fastPath = cacheConfig === void 0 && (cache === void 0 || (0, __entity_ts.is)(cache, __cache_core_cache_ts.NoopCache));
	}
	/** @internal */
	async queryWithCache(queryString, params, query) {
		const cacheStrat = this.cache !== void 0 && !(0, __entity_ts.is)(this.cache, __cache_core_cache_ts.NoopCache) ? await (0, __cache_core_cache_ts.strategyFor)(queryString, params, this.queryMetadata, this.cacheConfig) : { type: "skip" };
		if (cacheStrat.type === "skip") return query().catch((e) => {
			throw new __errors_ts.DrizzleQueryError(queryString, params, e);
		});
		const cache = this.cache;
		if (cacheStrat.type === "invalidate") return Promise.all([query(), cache.onMutate({ tables: cacheStrat.tables })]).then((res) => res[0]).catch((e) => {
			throw new __errors_ts.DrizzleQueryError(queryString, params, e);
		});
		if (cacheStrat.type === "try") {
			const { tables, key, isTag, autoInvalidate, config } = cacheStrat;
			const fromCache = await cache.get(key, tables, isTag, autoInvalidate);
			if (fromCache === void 0) {
				const result = await query().catch((e) => {
					throw new __errors_ts.DrizzleQueryError(queryString, params, e);
				});
				await cache.put(key, result, autoInvalidate ? tables : [], isTag, config);
				return result;
			}
			return fromCache;
		}
		(0, __utils_ts.assertUnreachable)(cacheStrat);
	}
	async execute(placeholderValues = {}) {
		const { query, logger, executor, mapper, fastPath } = this;
		const sql = query._sql ? query._sql.join(" ") : query.sql;
		const params = query.params.length === 0 ? query.params : (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues);
		logger.logQuery(sql, params);
		const res = fastPath ? executor(params).catch((e) => {
			throw new __errors_ts.DrizzleQueryError(sql, params, e);
		}) : this.queryWithCache(sql, params, () => executor(params));
		if (!mapper) return res;
		return res.then((rows) => mapper(rows));
	}
	async *iterator(placeholderValues = {}) {
		const { query, logger, executor, _iterator, mapper, fastPath } = this;
		const sql = query._sql ? query._sql.join(" ") : query.sql;
		const params = query.params.length === 0 ? query.params : (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues);
		logger.logQuery(sql, params);
		if (_iterator) try {
			if (mapper) {
				for await (const row of _iterator(params)) yield mapper([row])[0];
				return;
			}
			for await (const row of _iterator(params)) yield row;
			return;
		} catch (e) {
			throw new __errors_ts.DrizzleQueryError(sql, params, e);
		}
		const rows = await (fastPath ? executor(params).catch((e) => {
			throw new __errors_ts.DrizzleQueryError(sql, params, e);
		}) : this.queryWithCache(sql, params, () => executor(params)));
		if (mapper) {
			for (const row of rows) yield mapper([row])[0];
			return;
		}
		for (const row of rows) yield row;
	}
};
var MySqlAsyncSession = class extends require_mysql_core_session.MySqlSession {
	static [__entity_ts.entityKind] = "MySqlAsyncSession";
	execute(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "raw").execute();
	}
	arrays(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays").execute();
	}
	objects(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "objects").execute();
	}
};
var MySqlAsyncTransaction = class extends require_mysql_core_async_db.MySqlAsyncDatabase {
	static [__entity_ts.entityKind] = "MySqlAsyncTransaction";
	constructor(dialect, session, relations, nestedIndex) {
		super(dialect, session, relations);
		this.relations = relations;
		this.nestedIndex = nestedIndex;
	}
	rollback() {
		throw new __errors_ts.TransactionRollbackError();
	}
};
async function migrate(migrations, db, config) {
	const migrationsTable = config.migrationsTable ?? "__drizzle_migrations";
	const { newDb } = await (0, __up_migrations_mysql_ts.upgradeIfNeeded)(migrationsTable, db.session, migrations);
	if (newDb) {
		const migrationTableCreate = __sql_sql_ts.sql`
			CREATE TABLE IF NOT EXISTS ${__sql_sql_ts.sql.identifier(migrationsTable)} (
				id SERIAL PRIMARY KEY,
				hash TEXT NOT NULL,
				created_at BIGINT,
				name TEXT,
				applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
			)
		`;
		await db.session.execute(migrationTableCreate);
	}
	const dbMigrations = await db.session.objects(__sql_sql_ts.sql`select id, hash, created_at, name from ${__sql_sql_ts.sql.identifier(migrationsTable)}`);
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return { exitCode: "databaseMigrations" };
		if (migrations.length > 1) return { exitCode: "localMigrations" };
		const [migration] = migrations;
		if (!migration) return;
		await db.session.execute(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		return;
	}
	const migrationsToRun = (0, __migrator_utils_ts.getMigrationsToRun)({
		localMigrations: migrations,
		dbMigrations
	});
	await db.transaction(async (tx) => {
		for (const migration of migrationsToRun) {
			for (const stmt of migration.sql) await tx.execute(__sql_sql_ts.sql.raw(stmt));
			await tx.execute(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		}
	});
}

//#endregion
exports.MySqlAsyncPreparedQuery = MySqlAsyncPreparedQuery;
exports.MySqlAsyncSession = MySqlAsyncSession;
exports.MySqlAsyncTransaction = MySqlAsyncTransaction;
exports.migrate = migrate;
