// drizzle-orm@1.0.0-rc.4/mysql-core/async/session.js
import { MySqlAsyncDatabase } from "./db.js";
import { MySqlBasePreparedQuery, MySqlSession } from "../session.js";
import { entityKind, is } from "../../entity.js";
import { assertUnreachable } from "../../utils.js";
import { fillPlaceholders, sql } from "../../sql/sql.js";
import { DrizzleQueryError, TransactionRollbackError } from "../../errors.js";
import { getMigrationsToRun } from "../../migrator.utils.js";
import { NoopCache, strategyFor } from "../../cache/core/cache.js";
import { upgradeIfNeeded } from "../../up-migrations/mysql.js";

//#region src/mysql-core/async/session.ts
var MySqlAsyncPreparedQuery = class extends MySqlBasePreparedQuery {
	static [entityKind] = "MySqlAsyncPreparedQuery";
	/** @internal */
	mapper;
	fastPath;
	constructor(executor, _iterator, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(query);
		this.executor = executor;
		this._iterator = _iterator;
		this.mode = mode;
		this.logger = logger;
		this.cache = cache;
		this.queryMetadata = queryMetadata;
		this.cacheConfig = cacheConfig;
		this.mapper = mapper;
		if (cache && cache.strategy() === "all" && cacheConfig === void 0) this.cacheConfig = {
			enabled: true,
			autoInvalidate: true
		};
		if (!this.cacheConfig?.enabled) this.cacheConfig = void 0;
		this.fastPath = cacheConfig === void 0 && (cache === void 0 || is(cache, NoopCache));
	}
	/** @internal */
	async queryWithCache(queryString, params, query) {
		const cacheStrat = this.cache !== void 0 && !is(this.cache, NoopCache) ? await strategyFor(queryString, params, this.queryMetadata, this.cacheConfig) : { type: "skip" };
		if (cacheStrat.type === "skip") return query().catch((e) => {
			throw new DrizzleQueryError(queryString, params, e);
		});
		const cache = this.cache;
		if (cacheStrat.type === "invalidate") return Promise.all([query(), cache.onMutate({ tables: cacheStrat.tables })]).then((res) => res[0]).catch((e) => {
			throw new DrizzleQueryError(queryString, params, e);
		});
		if (cacheStrat.type === "try") {
			const { tables, key, isTag, autoInvalidate, config } = cacheStrat;
			const fromCache = await cache.get(key, tables, isTag, autoInvalidate);
			if (fromCache === void 0) {
				const result = await query().catch((e) => {
					throw new DrizzleQueryError(queryString, params, e);
				});
				await cache.put(key, result, autoInvalidate ? tables : [], isTag, config);
				return result;
			}
			return fromCache;
		}
		assertUnreachable(cacheStrat);
	}
	async execute(placeholderValues = {}) {
		const { query, logger, executor, mapper, fastPath } = this;
		const sql = query._sql ? query._sql.join(" ") : query.sql;
		const params = query.params.length === 0 ? query.params : fillPlaceholders(query.params, placeholderValues);
		logger.logQuery(sql, params);
		const res = fastPath ? executor(params).catch((e) => {
			throw new DrizzleQueryError(sql, params, e);
		}) : this.queryWithCache(sql, params, () => executor(params));
		if (!mapper) return res;
		return res.then((rows) => mapper(rows));
	}
	async *iterator(placeholderValues = {}) {
		const { query, logger, executor, _iterator, mapper, fastPath } = this;
		const sql = query._sql ? query._sql.join(" ") : query.sql;
		const params = query.params.length === 0 ? query.params : fillPlaceholders(query.params, placeholderValues);
		logger.logQuery(sql, params);
		if (_iterator) try {
			if (mapper) {
				for await (const row of _iterator(params)) yield mapper([row])[0];
				return;
			}
			for await (const row of _iterator(params)) yield row;
			return;
		} catch (e) {
			throw new DrizzleQueryError(sql, params, e);
		}
		const rows = await (fastPath ? executor(params).catch((e) => {
			throw new DrizzleQueryError(sql, params, e);
		}) : this.queryWithCache(sql, params, () => executor(params)));
		if (mapper) {
			for (const row of rows) yield mapper([row])[0];
			return;
		}
		for (const row of rows) yield row;
	}
};
var MySqlAsyncSession = class extends MySqlSession {
	static [entityKind] = "MySqlAsyncSession";
	execute(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "raw").execute();
	}
	arrays(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays").execute();
	}
	objects(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "objects").execute();
	}
};
var MySqlAsyncTransaction = class extends MySqlAsyncDatabase {
	static [entityKind] = "MySqlAsyncTransaction";
	constructor(dialect, session, relations, nestedIndex) {
		super(dialect, session, relations);
		this.relations = relations;
		this.nestedIndex = nestedIndex;
	}
	rollback() {
		throw new TransactionRollbackError();
	}
};
async function migrate(migrations, db, config) {
	const migrationsTable = config.migrationsTable ?? "__drizzle_migrations";
	const { newDb } = await upgradeIfNeeded(migrationsTable, db.session, migrations);
	if (newDb) {
		const migrationTableCreate = sql`
			CREATE TABLE IF NOT EXISTS ${sql.identifier(migrationsTable)} (
				id SERIAL PRIMARY KEY,
				hash TEXT NOT NULL,
				created_at BIGINT,
				name TEXT,
				applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
			)
		`;
		await db.session.execute(migrationTableCreate);
	}
	const dbMigrations = await db.session.objects(sql`select id, hash, created_at, name from ${sql.identifier(migrationsTable)}`);
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return { exitCode: "databaseMigrations" };
		if (migrations.length > 1) return { exitCode: "localMigrations" };
		const [migration] = migrations;
		if (!migration) return;
		await db.session.execute(sql`insert into ${sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		return;
	}
	const migrationsToRun = getMigrationsToRun({
		localMigrations: migrations,
		dbMigrations
	});
	await db.transaction(async (tx) => {
		for (const migration of migrationsToRun) {
			for (const stmt of migration.sql) await tx.execute(sql.raw(stmt));
			await tx.execute(sql`insert into ${sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		}
	});
}

//#endregion
export { MySqlAsyncPreparedQuery, MySqlAsyncSession, MySqlAsyncTransaction, migrate };
