// drizzle-orm@1.0.0-rc.4/mysql-core/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __entity_ts = require("../entity.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");

//#region src/mysql-core/session.ts
var MySqlBasePreparedQuery = class {
	static [__entity_ts.entityKind] = "MySqlBasePreparedQuery";
	constructor(query) {
		this.query = query;
	}
	getQuery() {
		return this.query;
	}
};
var MySqlSession = class {
	static [__entity_ts.entityKind] = "MySqlSession";
	constructor(dialect) {
		this.dialect = dialect;
	}
	getSetTransactionSQL(config) {
		const parts = [];
		if (config.isolationLevel) parts.push(`isolation level ${config.isolationLevel}`);
		return parts.length ? __sql_sql_ts.sql`set transaction ${__sql_sql_ts.sql.raw(parts.join(" "))}` : void 0;
	}
	getStartTransactionSQL(config) {
		const parts = [];
		if (config.withConsistentSnapshot) parts.push("with consistent snapshot");
		if (config.accessMode) parts.push(config.accessMode);
		return parts.length ? __sql_sql_ts.sql`start transaction ${__sql_sql_ts.sql.raw(parts.join(" "))}` : void 0;
	}
};

//#endregion
exports.MySqlBasePreparedQuery = MySqlBasePreparedQuery;
exports.MySqlSession = MySqlSession;
