// drizzle-orm@1.0.0-rc.4/sqlite-core/effect/query.js
import { entityKind } from "../../entity.js";
import { SQLiteRelationalQuery } from "../query-builders/query.js";
import { applyEffectWrapper } from "../../effect-core/query-effect.js";

//#region src/sqlite-core/effect/query.ts
var SQLiteEffectRelationalQuery = class extends SQLiteRelationalQuery {
	static [entityKind] = "SQLiteEffectRelationalQueryV2";
	/** @internal */
	_prepare(prepare = false) {
		const { query, builtQuery } = this._toSQL();
		const mapper = this.dialect.mapperGenerators.relationalRows({
			isFirst: this.mode === "first",
			parseJson: true,
			parseJsonIfString: false,
			rootJsonMappers: false,
			selection: query.selection,
			arrayModeRoot: true
		});
		return this.session.prepareQuery(builtQuery, "arrays", prepare, "all", mapper);
	}
	prepare() {
		return this._prepare(true);
	}
	execute(placeholderValues) {
		return this._prepare().execute(placeholderValues);
	}
};
applyEffectWrapper(SQLiteEffectRelationalQuery);

//#endregion
export { SQLiteEffectRelationalQuery };
