// drizzle-orm@1.0.0-rc.4/sqlite-core/effect/session.js
import { SQLiteEffectDatabase } from "./db.js";
import { entityKind, is } from "../../entity.js";
import { assertUnreachable } from "../../utils.js";
import { fillPlaceholders, sql } from "../../sql/sql.js";
import { getMigrationsToRun } from "../../migrator.utils.js";
import * as Effect from "effect/Effect";
import { EffectCache } from "../../cache/core/cache-effect.js";
import { NoopCache, strategyFor } from "../../cache/core/cache.js";
import { SQLitePreparedQuery, SQLiteSession } from "../session.js";
import * as Cause from "effect/Cause";
import { EffectDrizzleQueryError, EffectTransactionRollbackError, MigratorInitError } from "../../effect-core/errors.js";
import { upgradeIfNeeded } from "../../up-migrations/effect-sqlite.js";

//#region src/sqlite-core/effect/session.ts
var SQLiteEffectPreparedQuery = class extends SQLitePreparedQuery {
	static [entityKind] = "SQLiteEffectPreparedQuery";
	constructor(executeMethod = "all", executors, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(executeMethod, query, mapper, mode);
		this.executors = executors;
		this.logger = logger;
		this.cache = cache;
		this.queryMetadata = queryMetadata;
		this.cacheConfig = cacheConfig;
		if (cache && cache.strategy() === "all" && cacheConfig === void 0) this.cacheConfig = {
			enabled: true,
			autoInvalidate: true
		};
		if (!this.cacheConfig?.enabled) this.cacheConfig = void 0;
	}
	run(placeholderValues = {}) {
		return Effect.gen({ self: this }, function* () {
			const { query, logger, executors } = this;
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : fillPlaceholders(query.params, placeholderValues);
			yield* logger.logQuery(sql, params);
			return yield* this.queryWithCache(sql, params, "run", Effect.suspend(() => executors.run(params)));
		});
	}
	all(placeholderValues = {}) {
		return Effect.gen({ self: this }, function* () {
			const { query, logger, executors, mapper } = this;
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : fillPlaceholders(query.params, placeholderValues);
			yield* logger.logQuery(sql, params);
			const rows = yield* this.queryWithCache(sql, params, "all", Effect.suspend(() => executors.all(params)));
			return mapper ? mapper(rows) : rows;
		});
	}
	get(placeholderValues = {}) {
		return Effect.gen({ self: this }, function* () {
			const { query, logger, executors, mapper } = this;
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : fillPlaceholders(query.params, placeholderValues);
			yield* logger.logQuery(sql, params);
			const row = yield* this.queryWithCache(sql, params, "get", Effect.suspend(() => executors.get(params)));
			if (!row) return;
			if (!mapper) return row;
			return mapper([row])[0];
		});
	}
	values(placeholderValues = {}) {
		return Effect.gen({ self: this }, function* () {
			const { query, logger, executors } = this;
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : fillPlaceholders(query.params, placeholderValues);
			yield* logger.logQuery(sql, params);
			return yield* this.queryWithCache(sql, params, "values", Effect.suspend(() => executors.values(params)));
		});
	}
	execute(placeholderValues) {
		return this[this.executeMethod](placeholderValues);
	}
	/** @internal */
	queryWithCache(queryString, params, executeMethod, query) {
		return Effect.gen({ self: this }, function* () {
			const { cacheConfig, queryMetadata } = this;
			const cache = yield* EffectCache;
			const cacheStrat = cache && !(cache.cache && is(cache.cache, NoopCache)) ? yield* Effect.tryPromise(() => strategyFor(queryString, params, queryMetadata, cacheConfig)) : { type: "skip" };
			if (cacheStrat.type === "skip") return yield* query;
			if (cacheStrat.type === "invalidate") {
				const result = yield* query;
				yield* cache.onMutate({ tables: cacheStrat.tables });
				return result;
			}
			if (cacheStrat.type === "try") {
				const { tables, key: _key, isTag, autoInvalidate, config } = cacheStrat;
				const key = `${executeMethod}_${_key}`;
				const fromCache = yield* cache.get(key, tables, isTag, autoInvalidate);
				if (typeof fromCache !== "undefined") return fromCache;
				const result = yield* query;
				yield* cache.put(key, result, autoInvalidate ? tables : [], isTag, config);
				return result;
			}
			assertUnreachable(cacheStrat);
		}).pipe(Effect.provideService(EffectCache, this.cache), Effect.catch((e) => {
			return Effect.fail(new EffectDrizzleQueryError({
				query: queryString,
				params,
				cause: Cause.fail(e)
			}));
		}));
	}
};
var SQLiteEffectSession = class extends SQLiteSession {
	static [entityKind] = "SQLiteEffectSession";
	constructor(dialect) {
		super(dialect);
	}
	run(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "raw", false, "run").run();
	}
	objects(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "objects", false, "all").all();
	}
	object(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "objects", false, "get").get();
	}
	arrays(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays", false, "all").all();
	}
	array(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays", false, "get").get();
	}
};
var SQLiteEffectTransaction = class extends SQLiteEffectDatabase {
	static [entityKind] = "SQLiteEffectTransaction";
	constructor(dialect, session, relations, nestedIndex = 0, forbidJsonb) {
		super(dialect, session, relations, forbidJsonb);
		this.nestedIndex = nestedIndex;
	}
	rollback() {
		return new EffectTransactionRollbackError();
	}
};
const migrate = Effect.fn("migrate")(function* (migrations, session, config) {
	const migrationsTable = config === void 0 ? "__drizzle_migrations" : typeof config === "string" ? "__drizzle_migrations" : config.migrationsTable ?? "__drizzle_migrations";
	const { newDb } = yield* upgradeIfNeeded(migrationsTable, session, migrations);
	if (newDb) {
		const migrationTableCreate = sql`
			CREATE TABLE IF NOT EXISTS ${sql.identifier(migrationsTable)} (
				id INTEGER PRIMARY KEY,
				hash text NOT NULL,
				created_at numeric,
				name text,
				applied_at TEXT
			)
		`;
		yield* session.run(migrationTableCreate);
	}
	const dbMigrations = yield* session.objects(sql`SELECT id, hash, created_at, name FROM ${sql.identifier(migrationsTable)};`);
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return yield* new MigratorInitError({ exitCode: "databaseMigrations" });
		if (migrations.length > 1) return yield* new MigratorInitError({ exitCode: "localMigrations" });
		const [migration] = migrations;
		if (!migration) return;
		yield* session.run(sql`insert into ${sql.identifier(migrationsTable)} ("hash", "created_at", "name", "applied_at") values(${migration.hash}, ${migration.folderMillis}, ${migration.name}, ${(/* @__PURE__ */ new Date()).toISOString()})`);
		return;
	}
	const migrationsToRun = getMigrationsToRun({
		localMigrations: migrations,
		dbMigrations
	});
	yield* session.transaction((tx) => Effect.gen(function* () {
		for (const migration of migrationsToRun) {
			for (const stmt of migration.sql) yield* tx.run(sql.raw(stmt));
			yield* tx.run(sql`insert into ${sql.identifier(migrationsTable)} ("hash", "created_at", "name", "applied_at") values(${migration.hash}, ${migration.folderMillis}, ${migration.name}, ${(/* @__PURE__ */ new Date()).toISOString()})`);
		}
	}));
});

//#endregion
export { SQLiteEffectPreparedQuery, SQLiteEffectSession, SQLiteEffectTransaction, migrate };
