// drizzle-orm@1.0.0-rc.4/sqlite-core/effect/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_sqlite_core_effect_db = require('./db.cjs');
let __entity_ts = require("../../entity.cjs");
let __utils_ts = require("../../utils.cjs");
let __sql_sql_ts = require("../../sql/sql.cjs");
let __migrator_utils_ts = require("../../migrator.utils.cjs");
let effect_Effect = require("effect/Effect");
effect_Effect = require_runtime.__toESM(effect_Effect);
let __cache_core_cache_effect_ts = require("../../cache/core/cache-effect.cjs");
let __cache_core_cache_ts = require("../../cache/core/cache.cjs");
let __sqlite_core_session_ts = require("../session.cjs");
let effect_Cause = require("effect/Cause");
effect_Cause = require_runtime.__toESM(effect_Cause);
let __effect_core_errors_ts = require("../../effect-core/errors.cjs");
let __up_migrations_effect_sqlite_ts = require("../../up-migrations/effect-sqlite.cjs");

//#region src/sqlite-core/effect/session.ts
var SQLiteEffectPreparedQuery = class extends __sqlite_core_session_ts.SQLitePreparedQuery {
	static [__entity_ts.entityKind] = "SQLiteEffectPreparedQuery";
	constructor(executeMethod = "all", executors, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(executeMethod, query, mapper, mode);
		this.executors = executors;
		this.logger = logger;
		this.cache = cache;
		this.queryMetadata = queryMetadata;
		this.cacheConfig = cacheConfig;
		if (cache && cache.strategy() === "all" && cacheConfig === void 0) this.cacheConfig = {
			enabled: true,
			autoInvalidate: true
		};
		if (!this.cacheConfig?.enabled) this.cacheConfig = void 0;
	}
	run(placeholderValues = {}) {
		return effect_Effect.gen({ self: this }, function* () {
			const { query, logger, executors } = this;
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues);
			yield* logger.logQuery(sql, params);
			return yield* this.queryWithCache(sql, params, "run", effect_Effect.suspend(() => executors.run(params)));
		});
	}
	all(placeholderValues = {}) {
		return effect_Effect.gen({ self: this }, function* () {
			const { query, logger, executors, mapper } = this;
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues);
			yield* logger.logQuery(sql, params);
			const rows = yield* this.queryWithCache(sql, params, "all", effect_Effect.suspend(() => executors.all(params)));
			return mapper ? mapper(rows) : rows;
		});
	}
	get(placeholderValues = {}) {
		return effect_Effect.gen({ self: this }, function* () {
			const { query, logger, executors, mapper } = this;
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues);
			yield* logger.logQuery(sql, params);
			const row = yield* this.queryWithCache(sql, params, "get", effect_Effect.suspend(() => executors.get(params)));
			if (!row) return;
			if (!mapper) return row;
			return mapper([row])[0];
		});
	}
	values(placeholderValues = {}) {
		return effect_Effect.gen({ self: this }, function* () {
			const { query, logger, executors } = this;
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues);
			yield* logger.logQuery(sql, params);
			return yield* this.queryWithCache(sql, params, "values", effect_Effect.suspend(() => executors.values(params)));
		});
	}
	execute(placeholderValues) {
		return this[this.executeMethod](placeholderValues);
	}
	/** @internal */
	queryWithCache(queryString, params, executeMethod, query) {
		return effect_Effect.gen({ self: this }, function* () {
			const { cacheConfig, queryMetadata } = this;
			const cache = yield* __cache_core_cache_effect_ts.EffectCache;
			const cacheStrat = cache && !(cache.cache && (0, __entity_ts.is)(cache.cache, __cache_core_cache_ts.NoopCache)) ? yield* effect_Effect.tryPromise(() => (0, __cache_core_cache_ts.strategyFor)(queryString, params, queryMetadata, cacheConfig)) : { type: "skip" };
			if (cacheStrat.type === "skip") return yield* query;
			if (cacheStrat.type === "invalidate") {
				const result = yield* query;
				yield* cache.onMutate({ tables: cacheStrat.tables });
				return result;
			}
			if (cacheStrat.type === "try") {
				const { tables, key: _key, isTag, autoInvalidate, config } = cacheStrat;
				const key = `${executeMethod}_${_key}`;
				const fromCache = yield* cache.get(key, tables, isTag, autoInvalidate);
				if (typeof fromCache !== "undefined") return fromCache;
				const result = yield* query;
				yield* cache.put(key, result, autoInvalidate ? tables : [], isTag, config);
				return result;
			}
			(0, __utils_ts.assertUnreachable)(cacheStrat);
		}).pipe(effect_Effect.provideService(__cache_core_cache_effect_ts.EffectCache, this.cache), effect_Effect.catch((e) => {
			return effect_Effect.fail(new __effect_core_errors_ts.EffectDrizzleQueryError({
				query: queryString,
				params,
				cause: effect_Cause.fail(e)
			}));
		}));
	}
};
var SQLiteEffectSession = class extends __sqlite_core_session_ts.SQLiteSession {
	static [__entity_ts.entityKind] = "SQLiteEffectSession";
	constructor(dialect) {
		super(dialect);
	}
	run(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "raw", false, "run").run();
	}
	objects(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "objects", false, "all").all();
	}
	object(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "objects", false, "get").get();
	}
	arrays(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays", false, "all").all();
	}
	array(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays", false, "get").get();
	}
};
var SQLiteEffectTransaction = class extends require_sqlite_core_effect_db.SQLiteEffectDatabase {
	static [__entity_ts.entityKind] = "SQLiteEffectTransaction";
	constructor(dialect, session, relations, nestedIndex = 0, forbidJsonb) {
		super(dialect, session, relations, forbidJsonb);
		this.nestedIndex = nestedIndex;
	}
	rollback() {
		return new __effect_core_errors_ts.EffectTransactionRollbackError();
	}
};
const migrate = effect_Effect.fn("migrate")(function* (migrations, session, config) {
	const migrationsTable = config === void 0 ? "__drizzle_migrations" : typeof config === "string" ? "__drizzle_migrations" : config.migrationsTable ?? "__drizzle_migrations";
	const { newDb } = yield* (0, __up_migrations_effect_sqlite_ts.upgradeIfNeeded)(migrationsTable, session, migrations);
	if (newDb) {
		const migrationTableCreate = __sql_sql_ts.sql`
			CREATE TABLE IF NOT EXISTS ${__sql_sql_ts.sql.identifier(migrationsTable)} (
				id INTEGER PRIMARY KEY,
				hash text NOT NULL,
				created_at numeric,
				name text,
				applied_at TEXT
			)
		`;
		yield* session.run(migrationTableCreate);
	}
	const dbMigrations = yield* session.objects(__sql_sql_ts.sql`SELECT id, hash, created_at, name FROM ${__sql_sql_ts.sql.identifier(migrationsTable)};`);
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return yield* new __effect_core_errors_ts.MigratorInitError({ exitCode: "databaseMigrations" });
		if (migrations.length > 1) return yield* new __effect_core_errors_ts.MigratorInitError({ exitCode: "localMigrations" });
		const [migration] = migrations;
		if (!migration) return;
		yield* session.run(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsTable)} ("hash", "created_at", "name", "applied_at") values(${migration.hash}, ${migration.folderMillis}, ${migration.name}, ${(/* @__PURE__ */ new Date()).toISOString()})`);
		return;
	}
	const migrationsToRun = (0, __migrator_utils_ts.getMigrationsToRun)({
		localMigrations: migrations,
		dbMigrations
	});
	yield* session.transaction((tx) => effect_Effect.gen(function* () {
		for (const migration of migrationsToRun) {
			for (const stmt of migration.sql) yield* tx.run(__sql_sql_ts.sql.raw(stmt));
			yield* tx.run(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsTable)} ("hash", "created_at", "name", "applied_at") values(${migration.hash}, ${migration.folderMillis}, ${migration.name}, ${(/* @__PURE__ */ new Date()).toISOString()})`);
		}
	}));
});

//#endregion
exports.SQLiteEffectPreparedQuery = SQLiteEffectPreparedQuery;
exports.SQLiteEffectSession = SQLiteEffectSession;
exports.SQLiteEffectTransaction = SQLiteEffectTransaction;
exports.migrate = migrate;
