// drizzle-orm@1.0.0-rc.4/effect-core/errors.js
import { entityKind } from "../entity.js";
import * as Schema$1 from "effect/Schema";

//#region src/effect-core/errors.ts
var EffectDrizzleError = class extends Schema$1.TaggedErrorClass()("EffectDrizzleError", {
	message: Schema$1.String,
	cause: Schema$1.Unknown
}) {
	static [entityKind] = "EffectDrizzleError";
};
var EffectDrizzleQueryError = class EffectDrizzleQueryError extends Schema$1.TaggedErrorClass()("EffectDrizzleQueryError", {
	query: Schema$1.String,
	params: Schema$1.Array(Schema$1.Any).pipe(Schema$1.mutable),
	cause: Schema$1.Unknown
}) {
	static [entityKind] = "EffectDrizzleQueryError";
	get message() {
		return `Failed query: ${this.query}\nparams: ${this.params}`;
	}
	constructor(params) {
		super(params);
		Error.captureStackTrace(this, EffectDrizzleQueryError);
	}
};
var EffectTransactionRollbackError = class extends Schema$1.TaggedErrorClass()("EffectTransactionRollbackError", {}) {
	static [entityKind] = "EffectTransactionRollbackError";
	message = "Rollback";
};
var MigratorInitError = class extends Schema$1.TaggedErrorClass()("MigratorInitError", { exitCode: Schema$1.Literals(["databaseMigrations", "localMigrations"]) }) {
	static [entityKind] = "MigratorInitError";
};

//#endregion
export { EffectDrizzleError, EffectDrizzleQueryError, EffectTransactionRollbackError, MigratorInitError };
