// drizzle-orm@1.0.0-rc.4/mysql-core/query-builders/query.js
import { entityKind } from "../../entity.js";

//#region src/mysql-core/query-builders/query.ts
var RelationalQueryBuilder = class {
	static [entityKind] = "MySqlRelationalQueryBuilderV2";
	constructor(schema, table, tableConfig, dialect, session, builder = MySqlRelationalQuery) {
		this.schema = schema;
		this.table = table;
		this.tableConfig = tableConfig;
		this.dialect = dialect;
		this.session = session;
		this.builder = builder;
	}
	findMany(config) {
		return new this.builder(this.schema, this.table, this.tableConfig, this.dialect, this.session, config ?? true, "many");
	}
	findFirst(config) {
		return new this.builder(this.schema, this.table, this.tableConfig, this.dialect, this.session, config ?? true, "first");
	}
};
var MySqlRelationalQuery = class {
	static [entityKind] = "MySqlRelationalQueryV2";
	constructor(schema, table, tableConfig, dialect, session, config, mode) {
		this.schema = schema;
		this.table = table;
		this.tableConfig = tableConfig;
		this.dialect = dialect;
		this.session = session;
		this.config = config;
		this.mode = mode;
	}
	_getQuery() {
		return this.dialect.buildRelationalQuery({
			schema: this.schema,
			table: this.table,
			tableConfig: this.tableConfig,
			queryConfig: this.config,
			mode: this.mode
		});
	}
	_toSQL() {
		const query = this._getQuery();
		return {
			builtQuery: this.dialect.sqlToQuery(query.sql),
			query
		};
	}
	getSQL() {
		return this._getQuery().sql;
	}
	toSQL() {
		return this._toSQL().builtQuery;
	}
};

//#endregion
export { MySqlRelationalQuery, RelationalQueryBuilder };
