// vike@0.4.260/dist/node/vite/plugins/pluginStripPointerImportAttribute.js
export { pluginStripPointerImportAttribute };
import { getImportStatements } from '../shared/parseEsModule.js';
import { getMagicString } from '../shared/getMagicString.js';
import '../assertEnvVite.js';
// Match `with { type: 'vike:pointer' }` (with optional whitespace variations)
const runtimeAttrRE = /\bwith\s*\{\s*type\s*:\s*['"]vike:pointer['"]\s*\}/g;
function pluginStripPointerImportAttribute() {
    return [
        {
            name: 'vike:pluginStripPointerImportAttribute',
            transform: {
                filter: {
                    code: {
                        include: 'vike:pointer',
                    },
                },
                handler(code, id) {
                    return stripPointerImportAttributes(code, id);
                },
            },
        },
    ];
}
async function stripPointerImportAttributes(code, id) {
    // Fast path: skip parsing when there is nothing to strip
    runtimeAttrRE.lastIndex = 0;
    if (!runtimeAttrRE.test(code))
        return;
    // Restrict stripping to genuine `import` statements, so that we don't touch occurrences
    // inside string literals/comments (e.g. documentation code examples).
    // https://github.com/brillout/docpress/issues/167
    let importStatements;
    try {
        importStatements = await getImportStatements(code);
    }
    catch {
        // Not parsable as an ES module (e.g. a raw Markdown file) => leave it untouched
        return;
    }
    const { magicString, getMagicStringResult } = getMagicString(code, id);
    for (const { ss, se } of importStatements)
        stripAttribute(magicString, code, ss, se);
    return getMagicStringResult();
}
function stripAttribute(magicString, code, start, end) {
    const region = code.slice(start, end);
    runtimeAttrRE.lastIndex = 0;
    let match;
    while ((match = runtimeAttrRE.exec(region)) !== null) {
        magicString.remove(start + match.index, start + match.index + match[0].length);
    }
}
