// drizzle-orm@1.0.0-rc.4/effect-sqlite-node/driver.js
import { EffectSQLiteNodeSession } from "./session.js";
import { entityKind } from "../entity.js";
import { jitCompatCheck } from "../utils.js";
import { SQLiteDialect } from "../sqlite-core/dialect.js";
import * as Effect from "effect/Effect";
import { EffectCache } from "../cache/core/cache-effect.js";
import { DefaultServices, DefaultServices as DefaultServices$1 } from "../effect-core/defaults.js";
import { EffectLogger } from "../effect-core/index.js";
import { SQLiteEffectDatabase } from "../sqlite-core/effect/db.js";
import { SqliteClient } from "@effect/sql-sqlite-node/SqliteClient";

//#region src/effect-sqlite-node/driver.ts
var EffectSQLiteNodeDatabase = class extends SQLiteEffectDatabase {
	static [entityKind] = "EffectSQLiteNodeDatabase";
};
/**
* Creates an EffectSQLiteNodeDatabase instance.
*
* Requires `SqliteClient`, `EffectLogger`, and `EffectCache` services to be provided.
* Use `DefaultServices` to provide default (no-op) logger and cache implementations.
*
* @example
* ```ts
* // With default services (no logging, no caching)
* const db = yield* SQLiteNodeDrizzle.make({ relations }).pipe(
*   Effect.provide(SQLiteNodeDrizzle.DefaultServices),
* );
*
* // With Effect-based logging
* const db = yield* SQLiteNodeDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layer),
*   Effect.provide(SQLiteNodeDrizzle.DefaultServices),
* );
*
* // With custom Drizzle logger
* const db = yield* SQLiteNodeDrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layerFromDrizzle(myLogger)),
*   Effect.provide(SQLiteNodeDrizzle.DefaultServices),
* );
* ```
*/
const make = Effect.fn("SQLiteNodeDrizzle.make")(function* (config = {}) {
	const client = yield* SqliteClient;
	const cache = yield* EffectCache;
	const logger = yield* EffectLogger;
	const dialect = new SQLiteDialect({ useJitMappers: jitCompatCheck(config.jit) });
	const relations = config.relations ?? {};
	const db = new EffectSQLiteNodeDatabase(dialect, new EffectSQLiteNodeSession(client, dialect, relations, {
		logger,
		cache
	}), relations);
	db.$client = client;
	db.$cache = cache;
	if (db.$cache) db.$cache["invalidate"] = cache.onMutate;
	return db;
});
/**
* Convenience function that creates an EffectSQLiteNodeDatabase with `DefaultServices` already provided.
*/
const makeWithDefaults = (config = {}) => make(config).pipe(Effect.provide(DefaultServices$1));

//#endregion
export { DefaultServices, EffectSQLiteNodeDatabase, make, makeWithDefaults };
