// drizzle-orm@1.0.0-rc.4/pg-core/codecs.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
const require_pg_core_array = require('./array.cjs');
const require_pg_core_columns_postgis_extension_utils = require('./columns/postgis_extension/utils.cjs');
let __codecs_ts = require("../codecs.cjs");
let __sql_sql_ts = require("../sql/sql.cjs");

//#region src/pg-core/codecs.ts
const PG_ALIAS_TO_TYPE_MAP = {
	int2: "smallint",
	integer: "int",
	int4: "int",
	int8: "bigint",
	decimal: "numeric",
	real: "float4",
	double: "float8",
	"double precision": "float8",
	serial2: "smallserial",
	serial4: "serial",
	serial8: "bigserial",
	character: "char",
	"character varying": "varchar",
	"time with time zone": "timetz",
	"time without time zone": "time",
	"timestamp with time zone": "timestamptz",
	"timestamp without time zone": "timestamp",
	boolean: "bool",
	"bit varying": "varbit"
};
function resolvePgTypeAlias(type) {
	return PG_ALIAS_TO_TYPE_MAP[type] ?? type;
}
const unionsTypeTable = {
	smallint: {
		smallint: "smallint",
		int: "int",
		bigint: "bigint:number",
		"bigint:number": "bigint:number",
		"bigint:string": "bigint:number",
		numeric: "numeric:number",
		"numeric:number": "numeric:number",
		"numeric:bigint": "numeric:number",
		float4: "float4",
		float8: "float8",
		smallserial: "smallint",
		serial: "int",
		bigserial: "bigint:number",
		"bigserial:number": "bigint:number",
		oid: "oid",
		regproc: "regproc",
		regprocedure: "regprocedure",
		regoper: "regoper",
		regoperator: "regoperator",
		regclass: "regclass",
		regtype: "regtype",
		regrole: "regrole",
		regnamespace: "regnamespace",
		regconfig: "regconfig",
		regdictionary: "regdictionary"
	},
	int: {
		smallint: "int",
		int: "int",
		bigint: "bigint:number",
		"bigint:number": "bigint:number",
		"bigint:string": "bigint:number",
		numeric: "numeric:number",
		"numeric:number": "numeric:number",
		"numeric:bigint": "numeric:number",
		float4: "float4",
		float8: "float8",
		smallserial: "int",
		serial: "int",
		bigserial: "bigint:number",
		"bigserial:number": "bigint:number",
		oid: "oid",
		regproc: "regproc",
		regprocedure: "regprocedure",
		regoper: "regoper",
		regoperator: "regoperator",
		regclass: "regclass",
		regtype: "regtype",
		regrole: "regrole",
		regnamespace: "regnamespace",
		regconfig: "regconfig",
		regdictionary: "regdictionary"
	},
	bigint: {
		smallint: "bigint",
		int: "bigint",
		bigint: "bigint",
		"bigint:number": "bigint",
		"bigint:string": "bigint",
		numeric: "numeric:bigint",
		"numeric:number": "numeric:bigint",
		"numeric:bigint": "numeric:bigint",
		float4: "float4",
		float8: "float8",
		smallserial: "bigint",
		serial: "bigint",
		bigserial: "bigint",
		"bigserial:number": "bigint",
		oid: "oid",
		regproc: "regproc",
		regprocedure: "regprocedure",
		regoper: "regoper",
		regoperator: "regoperator",
		regclass: "regclass",
		regtype: "regtype",
		regrole: "regrole",
		regnamespace: "regnamespace",
		regconfig: "regconfig",
		regdictionary: "regdictionary"
	},
	"bigint:number": {
		smallint: "bigint:number",
		int: "bigint:number",
		bigint: "bigint:number",
		"bigint:number": "bigint:number",
		"bigint:string": "bigint:number",
		numeric: "numeric:number",
		"numeric:number": "numeric:number",
		"numeric:bigint": "numeric:number",
		float4: "float4",
		float8: "float8",
		smallserial: "bigint:number",
		serial: "bigint:number",
		bigserial: "bigint:number",
		"bigserial:number": "bigint:number",
		oid: "oid",
		regproc: "regproc",
		regprocedure: "regprocedure",
		regoper: "regoper",
		regoperator: "regoperator",
		regclass: "regclass",
		regtype: "regtype",
		regrole: "regrole",
		regnamespace: "regnamespace",
		regconfig: "regconfig",
		regdictionary: "regdictionary"
	},
	"bigint:string": {
		smallint: "bigint:string",
		int: "bigint:string",
		bigint: "bigint:string",
		"bigint:number": "bigint:string",
		"bigint:string": "bigint:string",
		numeric: "numeric",
		"numeric:number": "numeric",
		"numeric:bigint": "numeric",
		float4: "float4",
		float8: "float8",
		smallserial: "bigint:string",
		serial: "bigint:string",
		bigserial: "bigint:string",
		"bigserial:number": "bigint:string",
		oid: "oid",
		regproc: "regproc",
		regprocedure: "regprocedure",
		regoper: "regoper",
		regoperator: "regoperator",
		regclass: "regclass",
		regtype: "regtype",
		regrole: "regrole",
		regnamespace: "regnamespace",
		regconfig: "regconfig",
		regdictionary: "regdictionary"
	},
	numeric: {
		smallint: "numeric",
		int: "numeric",
		bigint: "numeric",
		"bigint:number": "numeric",
		"bigint:string": "numeric",
		numeric: "numeric",
		"numeric:number": "numeric",
		"numeric:bigint": "numeric",
		float4: "float4",
		float8: "float8",
		smallserial: "numeric",
		serial: "numeric",
		bigserial: "numeric",
		"bigserial:number": "numeric"
	},
	"numeric:number": {
		smallint: "numeric:number",
		int: "numeric:number",
		bigint: "numeric:number",
		"bigint:number": "numeric:number",
		"bigint:string": "numeric:number",
		numeric: "numeric:number",
		"numeric:number": "numeric:number",
		"numeric:bigint": "numeric:number",
		float4: "float4",
		float8: "float8",
		smallserial: "numeric:number",
		serial: "numeric:number",
		bigserial: "numeric:number",
		"bigserial:number": "numeric:number"
	},
	"numeric:bigint": {
		smallint: "numeric:bigint",
		int: "numeric:bigint",
		bigint: "numeric:bigint",
		"bigint:number": "numeric:bigint",
		"bigint:string": "numeric:bigint",
		numeric: "numeric:bigint",
		"numeric:number": "numeric:bigint",
		"numeric:bigint": "numeric:bigint",
		float4: "float4",
		float8: "float8",
		smallserial: "numeric:bigint",
		serial: "numeric:bigint",
		bigserial: "numeric:bigint",
		"bigserial:number": "numeric:bigint"
	},
	float4: {
		smallint: "float4",
		int: "float4",
		bigint: "float4",
		"bigint:number": "float4",
		"bigint:string": "float4",
		numeric: "float4",
		"numeric:number": "float4",
		"numeric:bigint": "float4",
		float4: "float4",
		float8: "float8",
		smallserial: "float4",
		serial: "float4",
		bigserial: "float4",
		"bigserial:number": "float4"
	},
	float8: {
		smallint: "float8",
		int: "float8",
		bigint: "float8",
		"bigint:number": "float8",
		"bigint:string": "float8",
		numeric: "float8",
		"numeric:number": "float8",
		"numeric:bigint": "float8",
		float4: "float8",
		float8: "float8",
		smallserial: "float8",
		serial: "float8",
		bigserial: "float8",
		"bigserial:number": "float8"
	},
	money: { money: "money" },
	smallserial: {
		smallint: "smallint",
		int: "int",
		bigint: "bigint:number",
		"bigint:number": "bigint:number",
		"bigint:string": "bigint:number",
		numeric: "numeric:number",
		"numeric:number": "numeric:number",
		"numeric:bigint": "numeric:number",
		float4: "float4",
		float8: "float8",
		smallserial: "smallint",
		serial: "int",
		bigserial: "bigint:number",
		"bigserial:number": "bigint:number",
		oid: "oid",
		regproc: "regproc",
		regprocedure: "regprocedure",
		regoper: "regoper",
		regoperator: "regoperator",
		regclass: "regclass",
		regtype: "regtype",
		regrole: "regrole",
		regnamespace: "regnamespace",
		regconfig: "regconfig",
		regdictionary: "regdictionary"
	},
	serial: {
		smallint: "int",
		int: "int",
		bigint: "bigint:number",
		"bigint:number": "bigint:number",
		"bigint:string": "bigint:number",
		numeric: "numeric:number",
		"numeric:number": "numeric:number",
		"numeric:bigint": "numeric:number",
		float4: "float4",
		float8: "float8",
		smallserial: "int",
		serial: "int",
		bigserial: "bigint:number",
		"bigserial:number": "bigint:number",
		oid: "oid",
		regproc: "regproc",
		regprocedure: "regprocedure",
		regoper: "regoper",
		regoperator: "regoperator",
		regclass: "regclass",
		regtype: "regtype",
		regrole: "regrole",
		regnamespace: "regnamespace",
		regconfig: "regconfig",
		regdictionary: "regdictionary"
	},
	bigserial: {
		smallint: "bigint",
		int: "bigint",
		bigint: "bigint",
		"bigint:number": "bigint",
		"bigint:string": "bigint",
		numeric: "numeric:bigint",
		"numeric:number": "numeric:bigint",
		"numeric:bigint": "numeric:bigint",
		float4: "float4",
		float8: "float8",
		smallserial: "bigint",
		serial: "bigint",
		bigserial: "bigint",
		"bigserial:number": "bigint",
		oid: "oid",
		regproc: "regproc",
		regprocedure: "regprocedure",
		regoper: "regoper",
		regoperator: "regoperator",
		regclass: "regclass",
		regtype: "regtype",
		regrole: "regrole",
		regnamespace: "regnamespace",
		regconfig: "regconfig",
		regdictionary: "regdictionary"
	},
	"bigserial:number": {
		smallint: "bigint:number",
		int: "bigint:number",
		bigint: "bigint:number",
		"bigint:number": "bigint:number",
		"bigint:string": "bigint:number",
		numeric: "numeric:number",
		"numeric:number": "numeric:number",
		"numeric:bigint": "numeric:number",
		float4: "float4",
		float8: "float8",
		smallserial: "bigint:number",
		serial: "bigint:number",
		bigserial: "bigint:number",
		"bigserial:number": "bigint:number",
		oid: "oid",
		regproc: "regproc",
		regprocedure: "regprocedure",
		regoper: "regoper",
		regoperator: "regoperator",
		regclass: "regclass",
		regtype: "regtype",
		regrole: "regrole",
		regnamespace: "regnamespace",
		regconfig: "regconfig",
		regdictionary: "regdictionary"
	},
	char: {
		char: "char",
		varchar: "char",
		text: "char"
	},
	varchar: {
		char: "varchar",
		varchar: "varchar",
		text: "varchar"
	},
	text: {
		char: "text",
		varchar: "text",
		text: "text"
	},
	bytea: { bytea: "bytea" },
	date: {
		date: "date",
		"date:string": "date",
		timestamp: "timestamp",
		"timestamp:string": "timestamp",
		timestamptz: "timestamptz",
		"timestamptz:string": "timestamptz"
	},
	"date:string": {
		date: "date:string",
		"date:string": "date:string",
		timestamp: "timestamp:string",
		"timestamp:string": "timestamp:string",
		timestamptz: "timestamptz:string",
		"timestamptz:string": "timestamptz:string"
	},
	time: {
		time: "time",
		timetz: "timetz"
	},
	timetz: {
		time: "timetz",
		timetz: "timetz"
	},
	timestamp: {
		date: "timestamp",
		"date:string": "timestamp",
		timestamp: "timestamp",
		"timestamp:string": "timestamp",
		timestamptz: "timestamptz",
		"timestamptz:string": "timestamptz"
	},
	"timestamp:string": {
		date: "timestamp:string",
		"date:string": "timestamp:string",
		timestamp: "timestamp:string",
		"timestamp:string": "timestamp:string",
		timestamptz: "timestamptz:string",
		"timestamptz:string": "timestamptz:string"
	},
	timestamptz: {
		date: "timestamptz",
		"date:string": "timestamptz",
		timestamp: "timestamptz",
		"timestamp:string": "timestamptz",
		timestamptz: "timestamptz",
		"timestamptz:string": "timestamptz"
	},
	"timestamptz:string": {
		date: "timestamptz:string",
		"date:string": "timestamptz:string",
		timestamp: "timestamptz:string",
		"timestamp:string": "timestamptz:string",
		timestamptz: "timestamptz:string",
		"timestamptz:string": "timestamptz:string"
	},
	interval: {
		interval: "interval",
		"interval:tuple": "interval"
	},
	"interval:tuple": {
		interval: "interval:tuple",
		"interval:tuple": "interval:tuple"
	},
	bool: { bool: "bool" },
	enum: { enum: "enum" },
	point: {
		point: "point",
		"point:tuple": "point"
	},
	"point:tuple": {
		point: "point:tuple",
		"point:tuple": "point:tuple"
	},
	line: {
		line: "line",
		"line:tuple": "line"
	},
	"line:tuple": {
		line: "line:tuple",
		"line:tuple": "line:tuple"
	},
	lseg: { lseg: "lseg" },
	box: { box: "box" },
	path: { path: "path" },
	polygon: { polygon: "polygon" },
	circle: { circle: "circle" },
	cidr: {
		cidr: "cidr",
		inet: "inet"
	},
	inet: {
		cidr: "inet",
		inet: "inet"
	},
	macaddr: {
		macaddr: "macaddr",
		macaddr8: "macaddr"
	},
	macaddr8: {
		macaddr: "macaddr8",
		macaddr8: "macaddr8"
	},
	bit: {
		bit: "bit",
		varbit: "bit"
	},
	varbit: {
		bit: "varbit",
		varbit: "varbit"
	},
	tsvector: { tsvector: "tsvector" },
	tsquery: { tsquery: "tsquery" },
	uuid: { uuid: "uuid" },
	xml: { xml: "xml" },
	json: { json: "json" },
	jsonb: { jsonb: "jsonb" },
	int4range: { int4range: "int4range" },
	int8range: { int8range: "int8range" },
	numrange: { numrange: "numrange" },
	tsrange: { tsrange: "tsrange" },
	tstzrange: { tstzrange: "tstzrange" },
	daterange: { daterange: "daterange" },
	int4multirange: { int4multirange: "int4multirange" },
	int8multirange: { int8multirange: "int8multirange" },
	nummultirange: { nummultirange: "nummultirange" },
	tsmultirange: { tsmultirange: "tsmultirange" },
	tstzmultirange: { tstzmultirange: "tstzmultirange" },
	datemultirange: { datemultirange: "datemultirange" },
	oid: {
		smallint: "oid",
		int: "oid",
		bigint: "oid",
		"bigint:number": "oid",
		"bigint:string": "oid",
		smallserial: "oid",
		serial: "oid",
		bigserial: "oid",
		"bigserial:number": "oid",
		oid: "oid",
		regproc: "oid",
		regprocedure: "oid",
		regoper: "oid",
		regoperator: "oid",
		regclass: "oid",
		regtype: "oid",
		regrole: "oid",
		regnamespace: "oid",
		regconfig: "oid",
		regdictionary: "oid"
	},
	regproc: {
		smallint: "regproc",
		int: "regproc",
		bigint: "regproc",
		"bigint:number": "regproc",
		"bigint:string": "regproc",
		smallserial: "regproc",
		serial: "regproc",
		bigserial: "regproc",
		"bigserial:number": "regproc",
		oid: "regproc",
		regproc: "regproc",
		regprocedure: "regproc"
	},
	regprocedure: {
		smallint: "regprocedure",
		int: "regprocedure",
		bigint: "regprocedure",
		"bigint:number": "regprocedure",
		"bigint:string": "regprocedure",
		smallserial: "regprocedure",
		serial: "regprocedure",
		bigserial: "regprocedure",
		"bigserial:number": "regprocedure",
		oid: "regprocedure",
		regproc: "regprocedure",
		regprocedure: "regprocedure"
	},
	regoper: {
		smallint: "regoper",
		int: "regoper",
		bigint: "regoper",
		"bigint:number": "regoper",
		"bigint:string": "regoper",
		smallserial: "regoper",
		serial: "regoper",
		bigserial: "regoper",
		"bigserial:number": "regoper",
		oid: "regoper",
		regoper: "regoper",
		regoperator: "regoper"
	},
	regoperator: {
		smallint: "regoperator",
		int: "regoperator",
		bigint: "regoperator",
		"bigint:number": "regoperator",
		"bigint:string": "regoperator",
		smallserial: "regoperator",
		serial: "regoperator",
		bigserial: "regoperator",
		"bigserial:number": "regoperator",
		oid: "regoperator",
		regoper: "regoperator",
		regoperator: "regoperator"
	},
	regclass: {
		smallint: "regclass",
		int: "regclass",
		bigint: "regclass",
		"bigint:number": "regclass",
		"bigint:string": "regclass",
		smallserial: "regclass",
		serial: "regclass",
		bigserial: "regclass",
		"bigserial:number": "regclass",
		oid: "regclass",
		regclass: "regclass"
	},
	regtype: {
		smallint: "regtype",
		int: "regtype",
		bigint: "regtype",
		"bigint:number": "regtype",
		"bigint:string": "regtype",
		smallserial: "regtype",
		serial: "regtype",
		bigserial: "regtype",
		"bigserial:number": "regtype",
		oid: "regtype",
		regtype: "regtype"
	},
	regrole: {
		smallint: "regrole",
		int: "regrole",
		bigint: "regrole",
		"bigint:number": "regrole",
		"bigint:string": "regrole",
		smallserial: "regrole",
		serial: "regrole",
		bigserial: "regrole",
		"bigserial:number": "regrole",
		oid: "regrole",
		regrole: "regrole"
	},
	regnamespace: {
		smallint: "regnamespace",
		int: "regnamespace",
		bigint: "regnamespace",
		"bigint:number": "regnamespace",
		"bigint:string": "regnamespace",
		smallserial: "regnamespace",
		serial: "regnamespace",
		bigserial: "regnamespace",
		"bigserial:number": "regnamespace",
		oid: "regnamespace",
		regnamespace: "regnamespace"
	},
	regconfig: {
		smallint: "regconfig",
		int: "regconfig",
		bigint: "regconfig",
		"bigint:number": "regconfig",
		"bigint:string": "regconfig",
		smallserial: "regconfig",
		serial: "regconfig",
		bigserial: "regconfig",
		"bigserial:number": "regconfig",
		oid: "regconfig",
		regconfig: "regconfig"
	},
	regdictionary: {
		smallint: "regdictionary",
		int: "regdictionary",
		bigint: "regdictionary",
		"bigint:number": "regdictionary",
		"bigint:string": "regdictionary",
		smallserial: "regdictionary",
		serial: "regdictionary",
		bigserial: "regdictionary",
		"bigserial:number": "regdictionary",
		oid: "regdictionary",
		regdictionary: "regdictionary"
	}
};
const castToText = (name) => __sql_sql_ts.sql`${name}::text`;
const castToTextArr = (name, arrayDimensions) => __sql_sql_ts.sql`${name}::text${__sql_sql_ts.sql.raw("[]".repeat(arrayDimensions))}`;
/** Used for cases when casting requires to unwrap and rebuild arrays
*
* @example
* string_mtx::text[][] // can be casted to array directly
*
* encode(bytea_mtx, 'base64')[][] // invalid syntax, cast requires unwrapping and rebuilding array
*/
const arrayCompatCast = (cast) => (name, arrayDimensions) => {
	if (!arrayDimensions) return cast(name);
	const aliases = [];
	for (let i = 0; i < arrayDimensions; i++) aliases.push(__sql_sql_ts.sql.identifier(`s${i}`));
	let indexed = name;
	for (const alias of aliases) indexed = __sql_sql_ts.sql`${indexed}[${alias}]`;
	let expression = __sql_sql_ts.sql`array(\
select ${cast(indexed)} \
from generate_subscripts(${name}, ${__sql_sql_ts.sql.raw(arrayDimensions.toString())}) ${aliases[arrayDimensions - 1]} \
order by ${aliases[arrayDimensions - 1]})`;
	for (let dim = arrayDimensions - 1; dim > 0; dim--) expression = __sql_sql_ts.sql`array(\
select ${expression} \
from generate_subscripts(${name}, ${__sql_sql_ts.sql.raw(dim.toString())}) ${aliases[dim - 1]} \
order by ${aliases[dim - 1]})`;
	return __sql_sql_ts.sql`case when ${name} is null then null else ${expression} end`;
};
/** Used to recursively apply value normalizer to array of unknown dimensions */
const arrayCompatNormalize = (normalize) => {
	const loop = (value, arrayDimensions) => {
		const innerDimensions = arrayDimensions - 1;
		if (arrayDimensions > 1) for (let i = 0; i < value.length; ++i) loop(value[i], innerDimensions);
		else for (let i = 0; i < value.length; ++i) value[i] = normalize(value[i]);
		return value;
	};
	return loop;
};
/** Doesn't mutate original data - used for insertions */
const arrayCompatNormalizeInput = (normalize, transformToPgArray = false) => {
	const loop = (value, arrayDimensions) => {
		const innerDimensions = arrayDimensions - 1;
		const out = Array.from({ length: value.length });
		if (arrayDimensions > 1) for (let i = 0; i < value.length; ++i) out[i] = loop(value[i], innerDimensions);
		else for (let i = 0; i < value.length; ++i) out[i] = normalize(value[i]);
		return out;
	};
	return transformToPgArray ? (v, d) => require_pg_core_array.makePgArray(loop(v, d)) : loop;
};
/** Parses a raw PG array text representation, then applies a per-item normalizer */
const parsePgArrayAndNormalize = (normalize) => {
	const codec = arrayCompatNormalize(normalize);
	return (value, arrayDimensions) => codec(require_pg_core_array.parsePgArray(value), arrayDimensions);
};
const parseLineTuple = (v) => {
	const [a, b, c] = v.slice(1, -1).split(",");
	return [
		Number.parseFloat(a),
		Number.parseFloat(b),
		Number.parseFloat(c)
	];
};
const parseLineABC = (v) => {
	const [a, b, c] = v.slice(1, -1).split(",");
	return {
		a: Number.parseFloat(a),
		b: Number.parseFloat(b),
		c: Number.parseFloat(c)
	};
};
const parsePointTuple = (v) => {
	const [x, y] = v.slice(1, -1).split(",");
	return [Number.parseFloat(x), Number.parseFloat(y)];
};
const parsePointXY = (v) => {
	const [x, y] = v.slice(1, -1).split(",");
	return {
		x: Number.parseFloat(x),
		y: Number.parseFloat(y)
	};
};
const parseGeometryTuple = (v) => require_pg_core_columns_postgis_extension_utils.parseEWKB(v).point;
const parseGeometryXY = (v) => {
	const parsed = require_pg_core_columns_postgis_extension_utils.parseEWKB(v);
	return {
		x: parsed.point[0],
		y: parsed.point[1]
	};
};
const textToDate = (v) => new Date(v);
const textToDateWithTz = (v) => /* @__PURE__ */ new Date(v + "+0000");
const parsePgVector = (v) => {
	const body = v.slice(1, -1);
	if (body.length === 0) return [];
	return body.split(",").map(Number.parseFloat);
};
const genericPgCodecs = {
	bytea: {
		castInJson: (name) => __sql_sql_ts.sql`encode(${name}, 'base64')`,
		castArrayInJson: arrayCompatCast((name) => __sql_sql_ts.sql`encode(${name}, 'base64')`),
		normalizeInJson: (v) => Buffer.from(v, "base64"),
		normalizeArrayInJson: arrayCompatNormalize((v) => Buffer.from(v, "base64"))
	},
	bigint: {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalizeInJson: BigInt,
		normalizeArrayInJson: arrayCompatNormalize(BigInt)
	},
	"bigint:number": {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalize: Number,
		normalizeArray: arrayCompatNormalize(Number),
		normalizeInJson: Number,
		normalizeArrayInJson: arrayCompatNormalize(Number)
	},
	"bigint:string": {
		castInJson: castToText,
		castArrayInJson: castToTextArr
	},
	bigserial: {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalizeInJson: BigInt,
		normalizeArrayInJson: arrayCompatNormalize(BigInt),
		normalize: BigInt,
		normalizeArray: arrayCompatNormalize(BigInt)
	},
	"bigserial:number": {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalize: Number,
		normalizeArray: arrayCompatNormalize(Number),
		normalizeInJson: Number,
		normalizeArrayInJson: arrayCompatNormalize(Number)
	},
	date: {
		normalizeInJson: textToDate,
		normalizeArrayInJson: arrayCompatNormalize(textToDate)
	},
	"date:string": {},
	enum: {
		castArray: castToTextArr,
		normalizeParamArray: require_pg_core_array.makePgArray
	},
	"geometry(point)": {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalize: parseGeometryXY,
		normalizeArray: arrayCompatNormalize(parseGeometryXY),
		normalizeInJson: parseGeometryXY,
		normalizeArrayInJson: arrayCompatNormalize(parseGeometryXY)
	},
	"geometry(point):tuple": {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalize: parseGeometryTuple,
		normalizeArray: arrayCompatNormalize(parseGeometryTuple),
		normalizeInJson: parseGeometryTuple,
		normalizeArrayInJson: arrayCompatNormalize(parseGeometryTuple)
	},
	interval: { castArrayInJson: castToTextArr },
	json: { normalizeParamArray: arrayCompatNormalizeInput((v) => JSON.stringify(v), true) },
	jsonb: { normalizeParamArray: arrayCompatNormalizeInput((v) => JSON.stringify(v), true) },
	line: {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalize: parseLineABC,
		normalizeArray: arrayCompatNormalize(parseLineABC),
		normalizeInJson: parseLineABC,
		normalizeArrayInJson: arrayCompatNormalize(parseLineABC)
	},
	"line:tuple": {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalize: parseLineTuple,
		normalizeArray: arrayCompatNormalize(parseLineTuple),
		normalizeInJson: parseLineTuple,
		normalizeArrayInJson: arrayCompatNormalize(parseLineTuple)
	},
	numeric: {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		castArray: castToTextArr
	},
	"numeric:number": {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		castArray: castToTextArr,
		normalize: Number,
		normalizeArray: arrayCompatNormalize(Number),
		normalizeInJson: Number,
		normalizeArrayInJson: arrayCompatNormalize(Number)
	},
	"numeric:bigint": {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		castArray: castToTextArr,
		normalize: BigInt,
		normalizeArray: arrayCompatNormalize(BigInt),
		normalizeInJson: BigInt,
		normalizeArrayInJson: arrayCompatNormalize(BigInt)
	},
	point: {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalize: parsePointXY,
		normalizeArray: arrayCompatNormalize(parsePointXY),
		normalizeInJson: parsePointXY,
		normalizeArrayInJson: arrayCompatNormalize(parsePointXY)
	},
	"point:tuple": {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalize: parsePointTuple,
		normalizeArray: arrayCompatNormalize(parsePointTuple),
		normalizeInJson: parsePointTuple,
		normalizeArrayInJson: arrayCompatNormalize(parsePointTuple)
	},
	timestamp: {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalizeInJson: textToDateWithTz,
		normalizeArrayInJson: arrayCompatNormalize(textToDateWithTz)
	},
	timestamptz: {
		castInJson: castToText,
		castArrayInJson: castToTextArr,
		normalizeInJson: textToDate,
		normalizeArrayInJson: arrayCompatNormalize(textToDate)
	},
	"timestamp:string": {
		castInJson: castToText,
		castArrayInJson: castToTextArr
	},
	"timestamptz:string": {
		castInJson: castToText,
		castArrayInJson: castToTextArr
	},
	halfvec: {
		normalize: parsePgVector,
		normalizeArray: parsePgArrayAndNormalize(parsePgVector),
		normalizeInJson: parsePgVector,
		normalizeArrayInJson: arrayCompatNormalize(parsePgVector)
	},
	vector: {
		normalize: parsePgVector,
		normalizeArray: parsePgArrayAndNormalize(parsePgVector),
		normalizeInJson: parsePgVector,
		normalizeArrayInJson: arrayCompatNormalize(parsePgVector)
	}
};
const refineGenericPgCodecs = (extension) => (0, __codecs_ts.refineCodecs)(genericPgCodecs, extension);

//#endregion
exports.arrayCompatCast = arrayCompatCast;
exports.arrayCompatNormalize = arrayCompatNormalize;
exports.arrayCompatNormalizeInput = arrayCompatNormalizeInput;
exports.castToText = castToText;
exports.castToTextArr = castToTextArr;
exports.genericPgCodecs = genericPgCodecs;
exports.parseGeometryTuple = parseGeometryTuple;
exports.parseGeometryXY = parseGeometryXY;
exports.parseLineABC = parseLineABC;
exports.parseLineTuple = parseLineTuple;
exports.parsePgArrayAndNormalize = parsePgArrayAndNormalize;
exports.parsePgVector = parsePgVector;
exports.parsePointTuple = parsePointTuple;
exports.parsePointXY = parsePointXY;
exports.refineGenericPgCodecs = refineGenericPgCodecs;
exports.resolvePgTypeAlias = resolvePgTypeAlias;
exports.textToDate = textToDate;
exports.textToDateWithTz = textToDateWithTz;
exports.unionsTypeTable = unionsTypeTable;
