// @universal-middleware/core@0.4.18/dist/index.js
// src/runtime.ts
var knownUserAgents = {
  deno: "Deno",
  bun: "Bun",
  workerd: "Cloudflare-Workers",
  node: "Node.js"
};
var _getRuntimeKey = () => {
  const global = globalThis;
  const userAgentSupported = typeof navigator !== "undefined" && typeof navigator.userAgent === "string";
  if (userAgentSupported) {
    for (const [runtimeKey2, userAgent] of Object.entries(knownUserAgents)) {
      if (checkUserAgentEquals(userAgent)) {
        return runtimeKey2;
      }
    }
  }
  if (typeof global?.EdgeRuntime === "string") {
    return "edge-light";
  }
  if (global?.fastly !== void 0) {
    return "fastly";
  }
  if (global?.process?.release?.name === "node") {
    return "node";
  }
  return "other";
};
var runtimeKey;
var getRuntimeKey = () => {
  if (runtimeKey === void 0) {
    runtimeKey = _getRuntimeKey();
  }
  return runtimeKey;
};
var checkUserAgentEquals = (platform) => {
  const userAgent = navigator.userAgent;
  return userAgent.startsWith(platform);
};
function getRuntime(args) {
  const key = getRuntimeKey();
  return {
    runtime: key,
    ...args
  };
}

// src/adapter.ts
function getAdapter(key, args) {
  return {
    adapter: key,
    ...args
  };
}
function getAdapterRuntime(adapter, adapterArgs, runtimeArgs, request) {
  const a = getAdapter(adapter, adapterArgs);
  const r = getRuntime(runtimeArgs);
  const s = getSrvxNodeRuntime(request);
  return { ...r, ...a, ...s };
}
function getSrvxNodeRuntime(request) {
  const ret = {};
  if (request?.runtime?.node?.req) ret.req = request?.runtime.node.req;
  if (request?.runtime?.node?.res) ret.res = request?.runtime.node.res;
  return ret;
}

// src/const.ts
var universalSymbol = /* @__PURE__ */ Symbol.for("universal");
var unboundSymbol = /* @__PURE__ */ Symbol.for("unbound");
var contextSymbol = /* @__PURE__ */ Symbol.for("unContext");
var pathSymbol = /* @__PURE__ */ Symbol.for("unPath");
var methodSymbol = /* @__PURE__ */ Symbol.for("unMethod");
var orderSymbol = /* @__PURE__ */ Symbol.for("unOrder");
var nameSymbol = /* @__PURE__ */ Symbol.for("unName");
var urlSymbol = /* @__PURE__ */ Symbol.for("unUrl");
var optionsToSymbols = {
  name: nameSymbol,
  method: methodSymbol,
  path: pathSymbol,
  order: orderSymbol,
  context: contextSymbol
};
var MiddlewareOrder = /* @__PURE__ */ ((MiddlewareOrder2) => {
  MiddlewareOrder2[MiddlewareOrder2["RATE_LIMITING"] = -1e3] = "RATE_LIMITING";
  MiddlewareOrder2[MiddlewareOrder2["AUTHENTICATION"] = -900] = "AUTHENTICATION";
  MiddlewareOrder2[MiddlewareOrder2["AUTHORIZATION"] = -800] = "AUTHORIZATION";
  MiddlewareOrder2[MiddlewareOrder2["INPUT_VALIDATION"] = -700] = "INPUT_VALIDATION";
  MiddlewareOrder2[MiddlewareOrder2["CORS"] = -600] = "CORS";
  MiddlewareOrder2[MiddlewareOrder2["PARSING"] = -500] = "PARSING";
  MiddlewareOrder2[MiddlewareOrder2["CUSTOM_PRE_PROCESSING"] = -400] = "CUSTOM_PRE_PROCESSING";
  MiddlewareOrder2[MiddlewareOrder2["HANDLER"] = 0] = "HANDLER";
  MiddlewareOrder2[MiddlewareOrder2["RESPONSE_TRANSFORM"] = 100] = "RESPONSE_TRANSFORM";
  MiddlewareOrder2[MiddlewareOrder2["HEADER_MANAGEMENT"] = 200] = "HEADER_MANAGEMENT";
  MiddlewareOrder2[MiddlewareOrder2["RESPONSE_COMPRESSION"] = 300] = "RESPONSE_COMPRESSION";
  MiddlewareOrder2[MiddlewareOrder2["RESPONSE_CACHING"] = 400] = "RESPONSE_CACHING";
  MiddlewareOrder2[MiddlewareOrder2["LOGGING"] = 500] = "LOGGING";
  MiddlewareOrder2[MiddlewareOrder2["ERROR_HANDLING"] = 600] = "ERROR_HANDLING";
  MiddlewareOrder2[MiddlewareOrder2["CUSTOM_POST_PROCESSING"] = 700] = "CUSTOM_POST_PROCESSING";
  return MiddlewareOrder2;
})(MiddlewareOrder || {});

// src/compile.ts
function compileEnhance(middleware, options) {
  let code = "";
  for (const [key, value] of Object.entries(options)) {
    if (key in optionsToSymbols) {
      const symbolKey = Symbol.keyFor(optionsToSymbols[key]);
      code += `${middleware}[Symbol.for(${JSON.stringify(symbolKey)})]=${JSON.stringify(value)};`;
    }
  }
  return code;
}

// src/env.ts
function env(runtime) {
  switch (runtime.runtime) {
    case "bun":
    case "node":
    case "edge-light":
      return globalThis?.process?.env;
    case "deno":
      return Deno.env.toObject();
    case "workerd":
      return runtime.env ?? {};
    default:
      return {};
  }
}

// src/utils.ts
function isBodyInit(value) {
  return value === null || typeof value === "string" || value instanceof Blob || value instanceof ArrayBuffer || ArrayBuffer.isView(value) || value instanceof FormData || value instanceof URLSearchParams || value instanceof ReadableStream;
}
function mergeHeadersInto(first, ...sources) {
  for (const source of sources) {
    const headers = new Headers(source);
    for (const [key, value] of headers.entries()) {
      if (key === "set-cookie") {
        if (!first.getSetCookie().includes(value)) first.append(key, value);
      } else {
        if (first.get(key) !== value) first.set(key, value);
      }
    }
  }
  return first;
}
function nodeHeadersToWeb(nodeHeaders) {
  const headers = [];
  const keys = Object.keys(nodeHeaders);
  for (const key of keys) {
    const value = nodeHeaders[key];
    if (Array.isArray(value)) {
      for (const item of value) {
        headers.push([key, item]);
      }
    } else {
      headers.push([key, normalizeHttpHeader(value)]);
    }
  }
  return new Headers(headers);
}
function normalizeHttpHeader(value) {
  return value === void 0 ? "" : String(value);
}
function url(request) {
  if (request[urlSymbol]) {
    return request[urlSymbol];
  }
  if (Object.isFrozen(request) || Object.isSealed(request)) {
    return new URL(request.url);
  }
  request[urlSymbol] = new URL(request.url);
  return request[urlSymbol];
}
function cloneRequest(request, fields) {
  if (!fields) {
    return request.clone();
  }
  return new Request(fields?.url ?? request.url, {
    method: fields?.method ?? request.method,
    headers: fields?.headers ?? request.headers,
    body: fields?.body ?? request.body,
    mode: fields?.mode ?? request.mode,
    credentials: fields?.credentials ?? request.credentials,
    cache: fields?.cache ?? request.cache,
    redirect: fields?.redirect ?? request.redirect,
    referrer: fields?.referrer ?? request.referrer,
    integrity: fields?.integrity ?? request.integrity,
    keepalive: fields?.keepalive ?? request.keepalive,
    referrerPolicy: fields?.referrerPolicy ?? request.referrerPolicy,
    signal: fields?.signal ?? request.signal,
    // @ts-expect-error RequestInit: duplex option is required when sending a body
    duplex: "half"
  });
}
function getUniversal(subject) {
  return universalSymbol in subject ? subject[universalSymbol] : subject;
}
function getUniversalProp(subject, prop, defaultValue) {
  if (prop in subject) return subject[prop];
  if (universalSymbol in subject) return subject[universalSymbol][prop];
  return defaultValue;
}
function enhance(middleware, options) {
  const { immutable, ...rest } = options;
  const m = immutable === false ? middleware : cloneFunction(middleware);
  for (const [key, value] of Object.entries(rest)) {
    if (key in optionsToSymbols) {
      m[optionsToSymbols[key]] = value;
    }
  }
  return m;
}
function ordered(middlewares) {
  return Array.from(middlewares).sort(
    (a, b) => getUniversalProp(a, orderSymbol, 0) - getUniversalProp(b, orderSymbol, 0)
  );
}
function cloneFunction(originalFn) {
  const extendedFunction = function(...args) {
    return Reflect.apply(Object.getPrototypeOf(extendedFunction), this, args);
  };
  Object.setPrototypeOf(extendedFunction, originalFn);
  return extendedFunction;
}
function bindUniversal(universal, fn, wrapper) {
  const unboundFn = unboundSymbol in fn ? fn[unboundSymbol] : fn;
  const self = { [universalSymbol]: universal, [unboundSymbol]: unboundFn };
  const boundFn = unboundFn.bind(self);
  Object.assign(boundFn, self);
  return wrapper ? wrapper(boundFn) : boundFn;
}
function attachUniversal(universal, subject) {
  return Object.assign(subject, { [universalSymbol]: universal });
}
function isHandler(m) {
  const order = getUniversalProp(m, orderSymbol);
  const path = getUniversalProp(m, pathSymbol);
  if (typeof order === "number") {
    if (order !== 0 && path) {
      console.warn(
        `Found a Universal Middleware with "path" metadata. This will lead to unpredictable behaviour. Please open an issue at https://github.com/magne4000/universal-middleware and explain your use case with the expected behaviour.`
      );
    }
    return order === 0;
  }
  return Boolean(path);
}

// src/pipe.ts
function pipe(...a) {
  const ordererArgs = ordered(a);
  const fn = async function pipeInternal(request, context, runtime) {
    const pending = [];
    let _response;
    for (const m of ordererArgs) {
      if (isHandler(m) && _response) {
        continue;
      }
      const um = getUniversal(m);
      const response = await um(request, context ?? {}, runtime);
      if (typeof response === "function") {
        pending.push(response);
      } else if (response !== null && typeof response === "object") {
        if (response instanceof Response) {
          if (!_response) {
            _response = response;
          }
        } else {
          context = response;
        }
      }
    }
    if (!_response) {
      throw new Error("No Response found");
    }
    for (const m of pending) {
      const r = await m(_response);
      if (r) {
        _response = r;
      }
    }
    return _response;
  };
  if (!this?.noCast) {
    const lastMiddleware = a.at(-1);
    if (lastMiddleware && universalSymbol in lastMiddleware) {
      return bindUniversal(fn, lastMiddleware);
    }
  }
  return fn;
}

// src/route.ts
import { parse } from "regexparam";
function exec(path, result) {
  const out = {};
  const matches = result.pattern.exec(path);
  if (!matches) return null;
  for (let i = 0; i < result.keys.length; i++) {
    if (matches[i + 1]) {
      out[result.keys[i]] = matches[i + 1];
    }
  }
  return out;
}
function paramsFromRequest(request, path) {
  const url2 = new URL(request.url);
  return exec(url2.pathname, parse(path));
}
function params(request, runtime, path) {
  if (typeof path === "string") {
    return paramsFromRequest(request, path);
  }
  return runtime.params ?? null;
}

// ../../node_modules/.pnpm/rou3@0.8.1/node_modules/rou3/dist/index.mjs
var NullProtoObj = /* @__PURE__ */ (() => {
  const e = function() {
  };
  return e.prototype = /* @__PURE__ */ Object.create(null), Object.freeze(e.prototype), e;
})();
function createRouter() {
  return {
    root: { key: "" },
    static: new NullProtoObj()
  };
}
function expandGroupDelimiters(path) {
  let i = 0;
  let depth = 0;
  for (; i < path.length; i++) {
    const c = path.charCodeAt(i);
    if (c === 92) {
      i++;
      continue;
    }
    if (c === 40) {
      depth++;
      continue;
    }
    if (c === 41 && depth > 0) {
      depth--;
      continue;
    }
    if (c === 123 && depth === 0) break;
  }
  if (i >= path.length) return;
  let j = i + 1;
  depth = 0;
  for (; j < path.length; j++) {
    const c = path.charCodeAt(j);
    if (c === 92) {
      j++;
      continue;
    }
    if (c === 40) {
      depth++;
      continue;
    }
    if (c === 41 && depth > 0) {
      depth--;
      continue;
    }
    if (c === 125 && depth === 0) break;
  }
  if (j >= path.length) return;
  const mod = path[j + 1];
  const hasMod = mod === "?" || mod === "+" || mod === "*";
  const pre = path.slice(0, i);
  const body = path.slice(i + 1, j);
  const suf = path.slice(j + (hasMod ? 2 : 1));
  if (!hasMod) return [pre + body + suf];
  if (mod === "?") return [pre + body + suf, pre + suf];
  if (body.includes("/")) throw new Error("unsupported group repetition across segments");
  return [`${pre}(?:${body})${mod}${suf}`];
}
var UNNAMED_GROUP_PREFIX = "__rou3_unnamed_";
var _unnamedGroupPrefixLength = 15;
function hasSegmentWildcard(segment) {
  let depth = 0;
  for (let i = 0; i < segment.length; i++) {
    const ch = segment.charCodeAt(i);
    if (ch === 92) {
      i++;
      continue;
    }
    if (ch === 40) {
      depth++;
      continue;
    }
    if (ch === 41 && depth > 0) {
      depth--;
      continue;
    }
    if (ch === 42 && depth === 0) return true;
  }
  return false;
}
function replaceSegmentWildcards(segment, unnamedStart, toGroupKey = toUnnamedGroupKey) {
  let depth = 0;
  let nextIndex = unnamedStart;
  let replaced = "";
  for (let i = 0; i < segment.length; i++) {
    const ch = segment.charCodeAt(i);
    if (ch === 92) {
      replaced += segment[i];
      if (i + 1 < segment.length) replaced += segment[++i];
      continue;
    }
    if (ch === 40) {
      depth++;
      replaced += segment[i];
      continue;
    }
    if (ch === 41 && depth > 0) {
      depth--;
      replaced += segment[i];
      continue;
    }
    if (ch === 42 && depth === 0) {
      replaced += `(?<${toGroupKey(nextIndex++)}>[^/]*)`;
      continue;
    }
    replaced += segment[i];
  }
  return [replaced, nextIndex];
}
function toUnnamedGroupKey(index) {
  return `${UNNAMED_GROUP_PREFIX}${index}`;
}
function normalizeUnnamedGroupKey(key) {
  return key.startsWith("__rou3_unnamed_") ? key.slice(_unnamedGroupPrefixLength) : key;
}
function encodeEscapes(path) {
  return path.replace(/\\:/g, "\uFFFDA").replace(/\\\(/g, "\uFFFDB").replace(/\\\)/g, "\uFFFDC").replace(/\\\{/g, "\uFFFDD").replace(/\\\}/g, "\uFFFDE");
}
function decodeEscaped(segment) {
  return segment.replace(/\uFFFD([A-E])/g, (_, c) => c === "A" ? ":" : c === "B" ? "(" : c === "C" ? ")" : c === "D" ? "{" : "}");
}
function expandModifiers(segments) {
  for (let i = 0; i < segments.length; i++) {
    const m = segments[i].match(/^(.*:[\w-]+(?:\([^)]*\))?)([?+*])$/);
    if (!m) continue;
    const pre = segments.slice(0, i);
    const suf = segments.slice(i + 1);
    if (m[2] === "?") return ["/" + pre.concat(m[1]).concat(suf).join("/"), "/" + pre.concat(suf).join("/")];
    const name = m[1].match(/:([\w-]+)/)?.[1] || "_";
    const wc = "/" + [
      ...pre,
      `**:${name}`,
      ...suf
    ].join("/");
    const without = "/" + [...pre, ...suf].join("/");
    return m[2] === "+" ? [wc] : [wc, without];
  }
}
function normalizePath(path) {
  if (!path.includes("/.")) return path;
  const r = [];
  for (const s of path.split("/")) if (s === ".") continue;
  else if (s === ".." && r.length > 1) r.pop();
  else r.push(s);
  return r.join("/") || "/";
}
function splitPath(path) {
  const [_, ...s] = path.split("/");
  return s[s.length - 1] === "" ? s.slice(0, -1) : s;
}
function getMatchParams(segments, paramsMap) {
  const params2 = new NullProtoObj();
  for (const [index, name] of paramsMap) {
    const segment = index < 0 ? segments.slice(-(index + 1)).join("/") : segments[index];
    if (typeof name === "string") params2[name] = segment;
    else {
      const match = segment.match(name);
      if (match) for (const key in match.groups) params2[normalizeUnnamedGroupKey(key)] = match.groups[key];
    }
  }
  return params2;
}
function addRoute(ctx, method = "", path, data) {
  method = method.toUpperCase();
  if (path.charCodeAt(0) !== 47) path = `/${path}`;
  const groupExpanded = expandGroupDelimiters(path);
  if (groupExpanded) {
    for (const expandedPath of groupExpanded) addRoute(ctx, method, expandedPath, data);
    return;
  }
  path = encodeEscapes(path);
  const segments = splitPath(path);
  const expanded = expandModifiers(segments);
  if (expanded) {
    for (const p of expanded) addRoute(ctx, method, p, data);
    return;
  }
  let node = ctx.root;
  let _unnamedParamIndex = 0;
  const paramsMap = [];
  const paramsRegexp = [];
  for (let i = 0; i < segments.length; i++) {
    let segment = segments[i];
    if (segment.startsWith("**")) {
      if (!node.wildcard) node.wildcard = { key: "**" };
      node = node.wildcard;
      paramsMap.push([
        -(i + 1),
        segment.split(":")[1] || "_",
        segment.length === 2
      ]);
      break;
    }
    if (segment === "*" || segment.includes(":") || segment.includes("(") || hasSegmentWildcard(segment)) {
      if (!node.param) node.param = { key: "*" };
      node = node.param;
      if (segment === "*") paramsMap.push([
        i,
        String(_unnamedParamIndex++),
        true
      ]);
      else if (segment.includes(":", 1) || segment.includes("(") || hasSegmentWildcard(segment) || !/^:[\w-]+$/.test(segment)) {
        const [regexp, nextIndex] = getParamRegexp(segment, _unnamedParamIndex);
        _unnamedParamIndex = nextIndex;
        paramsRegexp[i] = regexp;
        node.hasRegexParam = true;
        paramsMap.push([
          i,
          regexp,
          false
        ]);
      } else paramsMap.push([
        i,
        segment.slice(1),
        false
      ]);
      continue;
    }
    if (segment === "\\*") segment = segments[i] = "*";
    else if (segment === "\\*\\*") segment = segments[i] = "**";
    segment = segments[i] = decodeEscaped(segment);
    const child = node.static?.[segment];
    if (child) node = child;
    else {
      const staticNode = { key: segment };
      if (!node.static) node.static = new NullProtoObj();
      node.static[segment] = staticNode;
      node = staticNode;
    }
  }
  const hasParams = paramsMap.length > 0;
  if (!node.methods) node.methods = new NullProtoObj();
  node.methods[method] ??= [];
  node.methods[method].push({
    data: data || null,
    paramsRegexp,
    paramsMap: hasParams ? paramsMap : void 0
  });
  if (!hasParams) ctx.static["/" + segments.join("/")] = node;
}
function getParamRegexp(segment, unnamedStart = 0) {
  let _i = unnamedStart;
  let _s = "", _d = 0;
  for (let j = 0; j < segment.length; j++) {
    const c = segment.charCodeAt(j);
    if (c === 40) _d++;
    else if (c === 41 && _d > 0) _d--;
    else if (c === 92 && _d === 0 && j + 1 < segment.length) {
      const n = segment[j + 1];
      if (n !== ":" && n !== "(" && n !== "*" && n !== "\\") {
        _s += "\uFFFE" + n;
        j++;
        continue;
      }
    }
    _s += segment[j];
  }
  [_s, _i] = replaceSegmentWildcards(_s, _i);
  const regex = _s.replace(/:([\w-]+)(?:\(([^)]*)\))?/g, (_, id, p) => `(?<${id}>${p || "[^/]+"})`).replace(/\((?![?<])/g, () => `(?<${toUnnamedGroupKey(_i++)}>`).replace(/\./g, "\\.").replace(/\uFFFE(.)/g, (_, c) => /[.*+?^${}()|[\]\\]/.test(c) ? `\\${c}` : c);
  return [new RegExp(`^${regex}$`), _i];
}
function findRoute(ctx, method = "", path, opts) {
  if (opts?.normalize) path = normalizePath(path);
  if (path.charCodeAt(path.length - 1) === 47) path = path.slice(0, -1);
  const staticNode = ctx.static[path];
  if (staticNode && staticNode.methods) {
    const staticMatch = staticNode.methods[method] || staticNode.methods[""];
    if (staticMatch !== void 0) return staticMatch[0];
  }
  const segments = splitPath(path);
  const match = _lookupTree(ctx.root, method, segments, 0)?.[0];
  if (match === void 0) return;
  if (opts?.params === false) return match;
  return {
    data: match.data,
    params: match.paramsMap ? getMatchParams(segments, match.paramsMap) : void 0
  };
}
function _lookupTree(node, method, segments, index) {
  if (index === segments.length) {
    if (node.methods) {
      const match = node.methods[method] || node.methods[""];
      if (match) return match;
    }
    if (node.param && node.param.methods) {
      const match = node.param.methods[method] || node.param.methods[""];
      if (match) {
        const pMap = match[0].paramsMap;
        if (pMap?.[pMap?.length - 1]?.[2]) return match;
      }
    }
    if (node.wildcard && node.wildcard.methods) {
      const match = node.wildcard.methods[method] || node.wildcard.methods[""];
      if (match) {
        const pMap = match[0].paramsMap;
        if (pMap?.[pMap?.length - 1]?.[2]) return match;
      }
    }
    return;
  }
  const segment = segments[index];
  if (node.static) {
    const staticChild = node.static[segment];
    if (staticChild) {
      const match = _lookupTree(staticChild, method, segments, index + 1);
      if (match) return match;
    }
  }
  if (node.param) {
    const match = _lookupTree(node.param, method, segments, index + 1);
    if (match) {
      if (node.param.hasRegexParam) {
        const exactMatch = match.find((m) => m.paramsRegexp[index]?.test(segment)) || match.find((m) => !m.paramsRegexp[index]);
        return exactMatch ? [exactMatch] : void 0;
      }
      return match;
    }
  }
  if (node.wildcard && node.wildcard.methods) return node.wildcard.methods[method] || node.wildcard.methods[""];
}

// src/router.ts
var UniversalRouter = class {
  router;
  #middlewares;
  #pipeMiddlewaresInUniversalRoute;
  #handle404;
  constructor(pipeMiddlewaresInUniversalRoute = true, handle404 = false) {
    this.router = createRouter();
    this.#middlewares = [];
    this.#pipeMiddlewaresInUniversalRoute = pipeMiddlewaresInUniversalRoute;
    this.#handle404 = handle404;
  }
  use(middleware) {
    this.#middlewares.push(middleware);
    return this;
  }
  route(handler) {
    const { path, method } = assertRoute(handler);
    const umHandler = getUniversal(handler);
    if (Array.isArray(method)) {
      for (const m of method) {
        addRoute(this.router, m, path, umHandler);
      }
    } else {
      addRoute(this.router, method, path, umHandler);
    }
    return this;
  }
  applyCatchAll() {
    if (this.#handle404) {
      for (const method of ["GET", "POST", "PATCH"]) {
        addRoute(this.router, method, "/**", () => {
          return new Response("NOT FOUND", {
            status: 404
          });
        });
      }
    }
    return this;
  }
  get [universalSymbol]() {
    const noCastPipe = pipe.bind({ noCast: true });
    return (request, ctx, runtime) => {
      const router = findRoute(this.router, request.method, url(request).pathname);
      if (router) {
        const routerCtx = getUniversalProp(router.data, contextSymbol);
        if (routerCtx) {
          Object.assign(ctx, routerCtx);
        }
        const handler = this.#pipeMiddlewaresInUniversalRoute && this.#middlewares.length > 0 ? (
          // biome-ignore lint/suspicious/noExplicitAny: ignored
          noCastPipe(...this.#middlewares, router.data)
        ) : router.data;
        if (router.params) {
          runtime.params ??= {};
          Object.assign(runtime.params, router.params);
        }
        return handler(request, ctx, runtime);
      }
      if (this.#pipeMiddlewaresInUniversalRoute && this.#middlewares.length > 0) {
        const middlewares = noCastPipe(...this.#middlewares);
        return middlewares(request, ctx, runtime);
      }
      if (this.#handle404) {
        return new Response("NOT FOUND", {
          status: 404
        });
      }
    };
  }
};
function apply(router, middlewares, defer) {
  const ms = ordered(middlewares);
  for (const m of ms) {
    if (isHandler(m)) {
      router.route(m);
    } else {
      router.use(m);
    }
  }
  if (!defer) {
    router.applyCatchAll();
  }
}
async function applyAsync(router, middlewares, defer) {
  const ms = ordered(middlewares);
  for (const m of ms) {
    if (isHandler(m)) {
      await router.route(m);
    } else {
      await router.use(m);
    }
  }
  if (!defer) {
    await router.applyCatchAll();
  }
}
function pipeRoute(middlewares, { pipeMiddlewaresInUniversalRoute = true, handle404 = false } = {}) {
  const router = new UniversalRouter(pipeMiddlewaresInUniversalRoute, handle404);
  apply(router, middlewares);
  return router[universalSymbol];
}
function assertRoute(middleware) {
  const path = getUniversalProp(middleware, pathSymbol);
  if (!path) {
    const name = getUniversalProp(middleware, nameSymbol);
    throw new TypeError(assertRouteErrorMessage("path", name));
  }
  const method = getUniversalProp(middleware, methodSymbol);
  if (!method) {
    const name = getUniversalProp(middleware, nameSymbol);
    throw new TypeError(assertRouteErrorMessage("method", name));
  }
  return { path, method };
}
function assertRouteErrorMessage(key, name) {
  if (name) {
    return `Route ${name} is defined without a "${key}". See https://universal-middleware.dev/helpers/enhance for details.`;
  }
  return `Unnamed route is defined without a "${key}". See https://universal-middleware.dev/helpers/enhance for details.`;
}
export {
  MiddlewareOrder,
  UniversalRouter,
  apply,
  applyAsync,
  attachUniversal,
  bindUniversal,
  cloneRequest,
  compileEnhance,
  contextSymbol,
  enhance,
  env,
  getAdapterRuntime,
  getRuntimeKey,
  getUniversal,
  getUniversalProp,
  isBodyInit,
  mergeHeadersInto,
  methodSymbol,
  nameSymbol,
  nodeHeadersToWeb,
  orderSymbol,
  params,
  pathSymbol,
  pipe,
  pipeRoute,
  universalSymbol,
  url
};
