// drizzle-orm@1.0.0-rc.4/effect-sqlite-do/driver.js
import { EffectSQLiteDOSession } from "./session.js";
import { entityKind } from "../entity.js";
import { SQLiteDialect } from "../sqlite-core/dialect.js";
import * as Effect from "effect/Effect";
import { EffectCache } from "../cache/core/cache-effect.js";
import { SqliteClient } from "@effect/sql-sqlite-do/SqliteClient";
import { DefaultServices, DefaultServices as DefaultServices$1 } from "../effect-core/defaults.js";
import { EffectLogger } from "../effect-core/index.js";
import { SQLiteEffectDatabase } from "../sqlite-core/effect/db.js";

//#region src/effect-sqlite-do/driver.ts
var EffectSQLiteDoDatabase = class extends SQLiteEffectDatabase {
	static [entityKind] = "EffectSQLiteDoDatabase";
};
/**
* Creates an EffectSQLiteDoDatabase instance.
*
* Requires `SqliteClient`, `EffectLogger`, and `EffectCache` services to be provided.
* Use `DefaultServices` to provide default (no-op) logger and cache implementations.
*
* @example
* ```ts
* // With default services (no logging, no caching)
* const db = yield* SQLiteDODrizzle.make({ relations }).pipe(
*   Effect.provide(SQLiteDODrizzle.DefaultServices),
* );
*
* // With Effect-based logging
* const db = yield* SQLiteDODrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layer),
*   Effect.provide(SQLiteDODrizzle.DefaultServices),
* );
*
* // With custom Drizzle logger
* const db = yield* SQLiteDODrizzle.make({ relations }).pipe(
*   Effect.provide(EffectLogger.layerFromDrizzle(myLogger)),
*   Effect.provide(SQLiteDODrizzle.DefaultServices),
* );
* ```
*/
const make = Effect.fn("SQLiteDODrizzle.make")(function* (config) {
	const client = yield* SqliteClient;
	const cache = yield* EffectCache;
	const logger = yield* EffectLogger;
	const dialect = new SQLiteDialect();
	const relations = config.relations ?? {};
	const db = new EffectSQLiteDoDatabase(dialect, new EffectSQLiteDOSession(client, dialect, relations, {
		logger,
		cache,
		storage: config.storage
	}), relations);
	db.$client = client;
	db.$cache = cache;
	if (db.$cache) db.$cache["invalidate"] = cache.onMutate;
	return db;
});
/**
* Convenience function that creates an EffectSQLiteDoDatabase with `DefaultServices` already provided.
*/
const makeWithDefaults = (config) => make(config).pipe(Effect.provide(DefaultServices$1));

//#endregion
export { DefaultServices, EffectSQLiteDoDatabase, make, makeWithDefaults };
