// drizzle-orm@1.0.0-rc.4/mysql-core/effect/query.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_mysql_core_query_builders_query = require('../query-builders/query.cjs');
let __entity_ts = require("../../entity.cjs");
let __effect_core_query_effect_ts = require("../../effect-core/query-effect.cjs");

//#region src/mysql-core/effect/query.ts
var MySqlEffectRelationalQuery = class extends require_mysql_core_query_builders_query.MySqlRelationalQuery {
	static [__entity_ts.entityKind] = "MySqlEffectRelationalQueryV2";
	prepare() {
		const { query, builtQuery } = this._toSQL();
		const mapper = this.dialect.mapperGenerators.relationalRows({
			isFirst: this.mode === "first",
			parseJson: false,
			parseJsonIfString: false,
			rootJsonMappers: false,
			arrayModeRoot: true,
			selection: query.selection
		});
		return this.session.prepareQuery(builtQuery, "arrays", mapper);
	}
	execute() {
		return this.prepare().execute();
	}
};
(0, __effect_core_query_effect_ts.applyEffectWrapper)(MySqlEffectRelationalQuery);

//#endregion
exports.MySqlEffectRelationalQuery = MySqlEffectRelationalQuery;
