// drizzle-orm@1.0.0-rc.4/mysql-core/effect/session.js
import { MySqlBasePreparedQuery, MySqlSession } from "../session.js";
import { MySqlEffectDatabase } from "./db.js";
import { entityKind, is } from "../../entity.js";
import { assertUnreachable } from "../../utils.js";
import { fillPlaceholders, sql } from "../../sql/sql.js";
import { getMigrationsToRun } from "../../migrator.utils.js";
import * as Effect from "effect/Effect";
import { EffectCache } from "../../cache/core/cache-effect.js";
import { NoopCache, strategyFor } from "../../cache/core/cache.js";
import * as Cause from "effect/Cause";
import { EffectDrizzleQueryError, EffectTransactionRollbackError, MigratorInitError } from "../../effect-core/errors.js";
import { upgradeIfNeeded } from "../../up-migrations/effect-mysql.js";

//#region src/mysql-core/effect/session.ts
var MySqlEffectPreparedQuery = class extends MySqlBasePreparedQuery {
	static [entityKind] = "MySqlEffectPreparedQuery";
	/** @internal */
	mapper;
	constructor(executor, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(query);
		this.executor = executor;
		this.mode = mode;
		this.logger = logger;
		this.cache = cache;
		this.queryMetadata = queryMetadata;
		this.cacheConfig = cacheConfig;
		this.mapper = mapper;
		if (cache && cache.strategy() === "all" && cacheConfig === void 0) this.cacheConfig = {
			enabled: true,
			autoInvalidate: true
		};
		if (!this.cacheConfig?.enabled) this.cacheConfig = void 0;
	}
	execute(placeholderValues = {}) {
		return Effect.gen({ self: this }, function* () {
			const { query, mapper, logger } = this;
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : fillPlaceholders(query.params, placeholderValues);
			yield* logger.logQuery(sql, params);
			const result = this.queryWithCache(sql, params, Effect.suspend(() => this.executor(params)));
			if (!mapper) return yield* result;
			return yield* result.pipe(Effect.map((rows) => mapper(rows)));
		});
	}
	/** @internal */
	queryWithCache(queryString, params, query) {
		return Effect.gen({ self: this }, function* () {
			const { cacheConfig, queryMetadata } = this;
			const cache = yield* EffectCache;
			const cacheStrat = cache && !(cache.cache && is(cache.cache, NoopCache)) ? yield* Effect.tryPromise(() => strategyFor(queryString, params, queryMetadata, cacheConfig)) : { type: "skip" };
			if (cacheStrat.type === "skip") return yield* query;
			if (cacheStrat.type === "invalidate") {
				const result = yield* query;
				yield* cache.onMutate({ tables: cacheStrat.tables });
				return result;
			}
			if (cacheStrat.type === "try") {
				const { tables, key, isTag, autoInvalidate, config } = cacheStrat;
				const fromCache = yield* cache.get(key, tables, isTag, autoInvalidate);
				if (typeof fromCache !== "undefined") return fromCache;
				const result = yield* query;
				yield* cache.put(key, result, autoInvalidate ? tables : [], isTag, config);
				return result;
			}
			assertUnreachable(cacheStrat);
		}).pipe(Effect.provideService(EffectCache, this.cache), Effect.catch((e) => {
			return Effect.fail(new EffectDrizzleQueryError({
				query: queryString,
				params,
				cause: Cause.fail(e)
			}));
		}));
	}
};
var MySqlEffectSession = class extends MySqlSession {
	static [entityKind] = "MySqlEffectSession";
	constructor(dialect) {
		super(dialect);
	}
	execute(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "raw").execute();
	}
	arrays(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays").execute();
	}
	objects(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "objects").execute();
	}
};
var MySqlEffectTransaction = class extends MySqlEffectDatabase {
	static [entityKind] = "MySqlEffectTransaction";
	constructor(dialect, session, relations, nestedIndex = 0) {
		super(dialect, session, relations);
		this.relations = relations;
		this.nestedIndex = nestedIndex;
	}
	rollback() {
		return new EffectTransactionRollbackError();
	}
};
const migrate = Effect.fn("migrate")(function* (migrations, session, config) {
	const migrationsTable = config.migrationsTable ?? "__drizzle_migrations";
	const { newDb } = yield* upgradeIfNeeded(migrationsTable, session, migrations);
	if (newDb) {
		const migrationTableCreate = sql`
			CREATE TABLE IF NOT EXISTS ${sql.identifier(migrationsTable)} (
				id SERIAL PRIMARY KEY,
				hash TEXT NOT NULL,
				created_at BIGINT,
				name TEXT,
				applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
			)
		`;
		yield* session.execute(migrationTableCreate);
	}
	const dbMigrations = yield* session.objects(sql`select id, hash, created_at, name from ${sql.identifier(migrationsTable)}`);
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return yield* new MigratorInitError({ exitCode: "databaseMigrations" });
		if (migrations.length > 1) return yield* new MigratorInitError({ exitCode: "localMigrations" });
		const [migration] = migrations;
		if (!migration) return;
		yield* session.execute(sql`insert into ${sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		return;
	}
	const migrationsToRun = getMigrationsToRun({
		localMigrations: migrations,
		dbMigrations
	});
	yield* session.transaction((tx) => Effect.gen(function* () {
		for (const migration of migrationsToRun) {
			for (const stmt of migration.sql) yield* tx.execute(sql.raw(stmt));
			yield* tx.execute(sql`insert into ${sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		}
	}));
});

//#endregion
export { MySqlEffectPreparedQuery, MySqlEffectSession, MySqlEffectTransaction, migrate };
