// drizzle-orm@1.0.0-rc.4/mysql-core/effect/session.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
const require_mysql_core_session = require('../session.cjs');
const require_mysql_core_effect_db = require('./db.cjs');
let __entity_ts = require("../../entity.cjs");
let __utils_ts = require("../../utils.cjs");
let __sql_sql_ts = require("../../sql/sql.cjs");
let __migrator_utils_ts = require("../../migrator.utils.cjs");
let effect_Effect = require("effect/Effect");
effect_Effect = require_runtime.__toESM(effect_Effect);
let __cache_core_cache_effect_ts = require("../../cache/core/cache-effect.cjs");
let __cache_core_cache_ts = require("../../cache/core/cache.cjs");
let effect_Cause = require("effect/Cause");
effect_Cause = require_runtime.__toESM(effect_Cause);
let __effect_core_errors_ts = require("../../effect-core/errors.cjs");
let __up_migrations_effect_mysql_ts = require("../../up-migrations/effect-mysql.cjs");

//#region src/mysql-core/effect/session.ts
var MySqlEffectPreparedQuery = class extends require_mysql_core_session.MySqlBasePreparedQuery {
	static [__entity_ts.entityKind] = "MySqlEffectPreparedQuery";
	/** @internal */
	mapper;
	constructor(executor, query, mapper, mode, logger, cache, queryMetadata, cacheConfig) {
		super(query);
		this.executor = executor;
		this.mode = mode;
		this.logger = logger;
		this.cache = cache;
		this.queryMetadata = queryMetadata;
		this.cacheConfig = cacheConfig;
		this.mapper = mapper;
		if (cache && cache.strategy() === "all" && cacheConfig === void 0) this.cacheConfig = {
			enabled: true,
			autoInvalidate: true
		};
		if (!this.cacheConfig?.enabled) this.cacheConfig = void 0;
	}
	execute(placeholderValues = {}) {
		return effect_Effect.gen({ self: this }, function* () {
			const { query, mapper, logger } = this;
			const sql = query._sql ? query._sql.join(" ") : query.sql;
			const params = query.params.length === 0 ? query.params : (0, __sql_sql_ts.fillPlaceholders)(query.params, placeholderValues);
			yield* logger.logQuery(sql, params);
			const result = this.queryWithCache(sql, params, effect_Effect.suspend(() => this.executor(params)));
			if (!mapper) return yield* result;
			return yield* result.pipe(effect_Effect.map((rows) => mapper(rows)));
		});
	}
	/** @internal */
	queryWithCache(queryString, params, query) {
		return effect_Effect.gen({ self: this }, function* () {
			const { cacheConfig, queryMetadata } = this;
			const cache = yield* __cache_core_cache_effect_ts.EffectCache;
			const cacheStrat = cache && !(cache.cache && (0, __entity_ts.is)(cache.cache, __cache_core_cache_ts.NoopCache)) ? yield* effect_Effect.tryPromise(() => (0, __cache_core_cache_ts.strategyFor)(queryString, params, queryMetadata, cacheConfig)) : { type: "skip" };
			if (cacheStrat.type === "skip") return yield* query;
			if (cacheStrat.type === "invalidate") {
				const result = yield* query;
				yield* cache.onMutate({ tables: cacheStrat.tables });
				return result;
			}
			if (cacheStrat.type === "try") {
				const { tables, key, isTag, autoInvalidate, config } = cacheStrat;
				const fromCache = yield* cache.get(key, tables, isTag, autoInvalidate);
				if (typeof fromCache !== "undefined") return fromCache;
				const result = yield* query;
				yield* cache.put(key, result, autoInvalidate ? tables : [], isTag, config);
				return result;
			}
			(0, __utils_ts.assertUnreachable)(cacheStrat);
		}).pipe(effect_Effect.provideService(__cache_core_cache_effect_ts.EffectCache, this.cache), effect_Effect.catch((e) => {
			return effect_Effect.fail(new __effect_core_errors_ts.EffectDrizzleQueryError({
				query: queryString,
				params,
				cause: effect_Cause.fail(e)
			}));
		}));
	}
};
var MySqlEffectSession = class extends require_mysql_core_session.MySqlSession {
	static [__entity_ts.entityKind] = "MySqlEffectSession";
	constructor(dialect) {
		super(dialect);
	}
	execute(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "raw").execute();
	}
	arrays(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "arrays").execute();
	}
	objects(query) {
		return this.prepareQuery(this.dialect.sqlToQuery(query), "objects").execute();
	}
};
var MySqlEffectTransaction = class extends require_mysql_core_effect_db.MySqlEffectDatabase {
	static [__entity_ts.entityKind] = "MySqlEffectTransaction";
	constructor(dialect, session, relations, nestedIndex = 0) {
		super(dialect, session, relations);
		this.relations = relations;
		this.nestedIndex = nestedIndex;
	}
	rollback() {
		return new __effect_core_errors_ts.EffectTransactionRollbackError();
	}
};
const migrate = effect_Effect.fn("migrate")(function* (migrations, session, config) {
	const migrationsTable = config.migrationsTable ?? "__drizzle_migrations";
	const { newDb } = yield* (0, __up_migrations_effect_mysql_ts.upgradeIfNeeded)(migrationsTable, session, migrations);
	if (newDb) {
		const migrationTableCreate = __sql_sql_ts.sql`
			CREATE TABLE IF NOT EXISTS ${__sql_sql_ts.sql.identifier(migrationsTable)} (
				id SERIAL PRIMARY KEY,
				hash TEXT NOT NULL,
				created_at BIGINT,
				name TEXT,
				applied_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
			)
		`;
		yield* session.execute(migrationTableCreate);
	}
	const dbMigrations = yield* session.objects(__sql_sql_ts.sql`select id, hash, created_at, name from ${__sql_sql_ts.sql.identifier(migrationsTable)}`);
	if (typeof config === "object" && config.init) {
		if (dbMigrations.length) return yield* new __effect_core_errors_ts.MigratorInitError({ exitCode: "databaseMigrations" });
		if (migrations.length > 1) return yield* new __effect_core_errors_ts.MigratorInitError({ exitCode: "localMigrations" });
		const [migration] = migrations;
		if (!migration) return;
		yield* session.execute(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		return;
	}
	const migrationsToRun = (0, __migrator_utils_ts.getMigrationsToRun)({
		localMigrations: migrations,
		dbMigrations
	});
	yield* session.transaction((tx) => effect_Effect.gen(function* () {
		for (const migration of migrationsToRun) {
			for (const stmt of migration.sql) yield* tx.execute(__sql_sql_ts.sql.raw(stmt));
			yield* tx.execute(__sql_sql_ts.sql`insert into ${__sql_sql_ts.sql.identifier(migrationsTable)} (\`hash\`, \`created_at\`, \`name\`) values(${migration.hash}, ${migration.folderMillis}, ${migration.name})`);
		}
	}));
});

//#endregion
exports.MySqlEffectPreparedQuery = MySqlEffectPreparedQuery;
exports.MySqlEffectSession = MySqlEffectSession;
exports.MySqlEffectTransaction = MySqlEffectTransaction;
exports.migrate = migrate;
