// @universal-deploy/vite@0.1.10/dist/index.js
import { node, node as node$1 } from "@universal-deploy/node/vite";
import { netlify } from "@universal-deploy/netlify/vite";
import { addEntry, catchAllEntry, getAllEntries } from "@universal-deploy/store";
import { addRoute, createRouter } from "rou3";
import { compileRouterToString } from "rou3/compiler";
import { createMiddleware } from "@universal-middleware/express";
import MagicString from "magic-string";

//#region src/const.ts
const catchAllId = "ud:catch-all";
const INSTANCE = Symbol("auto-instance");

//#endregion
//#region src/utils.ts
const pluginsUsage = {
	[catchAllId]: "@universal-deploy/vite",
	"ud:resolver": "@universal-deploy/vite"
};
function dependsOn(pluginName) {
	return { configResolved(config) {
		if (!config.plugins.some(({ name }) => name === pluginName)) this.error(`"${pluginName}" Vite plugin is missing. Install it from "${pluginsUsage[pluginName]}".`);
	} };
}
function assertFetchable(mod, id) {
	if (!mod || typeof mod !== "object") throw new Error(`Missing default export from ${id}`);
	if ("default" in mod && mod.default) mod = mod.default;
	if (!mod || typeof mod !== "object" || !("fetch" in mod) || typeof mod.fetch !== "function") throw new Error(`Default export from ${id} must include a { fetch() } function`);
	return mod;
}
async function asyncFlatten(arr) {
	const flattened = [];
	for (const item of arr) if (Array.isArray(item)) flattened.push(...await asyncFlatten(item));
	else if (item instanceof Promise) flattened.push(...await asyncFlatten([await item]));
	else if (item) flattened.push(item);
	return flattened;
}
/**
* Enables a plugin based on a specified condition callback which will be executed in the `config` hook.
*/
function enablePluginIf(condition, originalPlugin) {
	const originalConfig = originalPlugin.config;
	originalPlugin.config = {
		order: originalConfig && "order" in originalConfig ? originalConfig.order : "pre",
		async handler(c, e) {
			if (!await condition.call(this, c, e)) {
				const keysToDelete = Object.keys(originalPlugin).filter((k) => k !== "name");
				originalPlugin.name += ":disabled";
				for (const key of keysToDelete) delete originalPlugin[key];
			} else if (originalConfig) {
				if (typeof originalConfig === "function") return originalConfig.call(this, c, e);
				return originalConfig.handler.call(this, c, e);
			}
		}
	};
	return originalPlugin;
}
function shortenId(id, root) {
	const normalized = id.replaceAll("\\", "/");
	const normalizedRoot = root.replaceAll("\\", "/");
	const nmIdx = normalized.lastIndexOf("/node_modules/");
	if (nmIdx !== -1) {
		const [pkg, scope] = normalized.slice(nmIdx + 14).split("/");
		if (!pkg) return normalized;
		return pkg.startsWith("@") && scope ? `${pkg}/${scope}` : pkg;
	}
	return normalized.startsWith(normalizedRoot) ? normalized.slice(normalizedRoot.length).replace(/^\//, "") : normalized;
}

//#endregion
//#region src/plugins/glue.ts
/**
* Enable `@universal-deploy/netlify` only if `@netlify/vite-plugin` was found
*/
function netlifyGlue() {
	return netlify().map((p) => enablePluginIf(isNetlifyPluginInUse, p));
}
const isNetlifyPluginInUse = async (userConfig) => {
	return !!(await asyncFlatten(userConfig.plugins ?? [])).filter((p) => Boolean(p)).find((p) => p.name.match(/^vite-plugin-netlify/));
};

//#endregion
//#region src/plugins/supported.ts
/**
* Resolves an array of supported target plugins.
*
* @param callback
*/
function resolveTargets(callback) {
	const plugin = {
		name: "universal-deploy:resolve-targets-callback",
		config: {
			order: "pre",
			async handler(userConfig) {
				callback(await findSupportedDeploymentTargets(plugin, userConfig));
			}
		}
	};
	return plugin;
}
async function noDeploymentTargetFound(thisNodePlugin, c) {
	const targets = await findSupportedDeploymentTargets(thisNodePlugin, c);
	return Object.values(targets).every((target) => !target);
}
async function findSupportedDeploymentTargets(thisNodePlugin, c) {
	const plugins = (await asyncFlatten(c.plugins ?? [])).filter((p) => Boolean(p));
	const found = [];
	if (plugins.some((p) => p.name.match(/^vite-plugin-vercel/))) found.push("vite-plugin-vercel");
	if (plugins.some((p) => p.name.match(/^vite-plugin-cloudflare/))) found.push("@cloudflare/vite-plugin");
	if (plugins.some((p) => p.name.match(/^vite-plugin-netlify/))) found.push("@netlify/vite-plugin");
	if (plugins.some((p) => p.name.startsWith("ud:node:emit") && p[INSTANCE] !== thisNodePlugin?.[INSTANCE])) found.push("@universal-deploy/node");
	return found;
}

//#endregion
//#region src/plugins/target.ts
/**
* A generic target plugin that overrides the server entry with a custom wrapper.
*/
function target(entry) {
	let resolvedEntry;
	return {
		name: "ud:target:emit",
		apply: "build",
		config: {
			order: "post",
			handler() {
				const buildEnvOptions = {};
				if (this.meta?.rolldownVersion) buildEnvOptions.rolldownOptions = { input: { index: entry } };
				else buildEnvOptions.rollupOptions = { input: { index: entry } };
				return { environments: { ssr: { build: { ...buildEnvOptions } } } };
			}
		},
		async buildStart() {
			const resolved = await this.resolve(entry);
			if (resolved) resolvedEntry = resolved.id;
		},
		transform(code, id) {
			if (resolvedEntry && id === resolvedEntry) {
				if (!code.includes("virtual:ud:")) this.warn(`{ entry: "${entry}" } is missing "virtual:ud:catch-all" import.`);
			}
		}
	};
}

//#endregion
//#region src/plugins/auto.ts
/**
* Automatically enables the node adapter if no other deployment target (Vercel, Cloudflare, Netlify) is detected.
*/
function auto(options) {
	const instance = Symbol("instance");
	return [
		catchAll(),
		devServer(),
		...options?.entry ? [target(options.entry)] : [...node$1(options?.node).map((p) => {
			p[INSTANCE] = instance;
			return enablePluginIf((config) => noDeploymentTargetFound(p, config), p);
		}), ...netlifyGlue()]
	];
}

//#endregion
//#region src/plugins/catch-all.ts
const re_catchAll = /^virtual:ud:catch-all$/;
const re_catchAllDefault = /^virtual:ud:catch-all\?default$/;
function generateCode(entries, compiledFindRoute) {
	const eagerEntries = entries.filter((e) => e.eager);
	const staticImports = eagerEntries.map((e) => `import __eager_${e.key} from ${JSON.stringify(e.resolvedId)};`).join("\n");
	const mapEntries = entries.map((e) => e.eager ? `  "${e.key}": () => Promise.resolve({ default: __eager_${e.key} })` : `  "${e.key}": () => import(${JSON.stringify(e.resolvedId)})`).join(",\n");
	const idEntries = entries.map((e) => `  "${e.key}": ${JSON.stringify(e.label)}`).join(",\n");
	const reExports = eagerEntries.map((e) => `export * from ${JSON.stringify(e.resolvedId)};`).join("\n");
	const spreads = eagerEntries.map((e) => `  ...__eager_${e.key},`).join("\n");
	return `
${staticImports}
const __map = {
${mapEntries}
};
const __ids = {
${idEntries}
};

${compiledFindRoute}

${assertFetchable.toString()}

${reExports}
export default {
${spreads}
  async fetch(request, ...args) {
    const url = new URL(request.url);
    const key = findRoute(request.method, url.pathname);
    if (!key || !key.data) return;
    const mod = await __map[key.data]();
    return assertFetchable(mod, __ids[key.data]).fetch(request, ...args);
  }
}`;
}
function catchAll() {
	let root = "";
	return {
		name: catchAllId,
		configResolved(config) {
			root = config.root;
		},
		resolveId: {
			filter: { id: [re_catchAll, re_catchAllDefault] },
			handler(id) {
				return id;
			}
		},
		load: {
			filter: { id: [re_catchAll, re_catchAllDefault] },
			async handler() {
				const router = createRouter();
				const entries = [];
				const seen = /* @__PURE__ */ new Map();
				const duplicates = /* @__PURE__ */ new Set();
				for (const meta of getAllEntries()) {
					const resolved = await this.resolve(meta.id);
					if (!resolved) throw new Error(`Failed to resolve ${meta.id}`);
					const routes = new Set(Array.isArray(meta.route) ? meta.route : [meta.route]);
					const methods = Array.isArray(meta.method) ? meta.method : [meta.method ?? ""];
					const existing = seen.get(resolved.id);
					if (existing) {
						let added = false;
						for (const route of routes) if (!existing.routes.has(route)) {
							added = true;
							existing.routes.add(route);
							for (const method of methods) addRoute(router, method, route, existing.key);
						}
						if (!added) duplicates.add(resolved.id);
						if (routes.has("/**")) existing.eager = true;
					} else {
						const entry = {
							key: `m${entries.length}`,
							resolvedId: resolved.id,
							label: shortenId(resolved.id, root),
							eager: routes.has("/**"),
							routes
						};
						entries.push(entry);
						seen.set(resolved.id, entry);
						for (const route of routes) for (const method of methods) addRoute(router, method, route, entry.key);
					}
				}
				if (duplicates.size > 0) this.warn(`\nDuplicate entries detected in virtual:ud:catch-all. \nDuplicates:\n - ${Array.from(duplicates.values()).join("\n - ")}`);
				return generateCode(entries, compileRouterToString(router, "findRoute"));
			}
		}
	};
}

//#endregion
//#region src/plugins/dev-server.ts
function isRunnableDevEnvironment(environment) {
	return "runner" in environment;
}
function isFetchableDevEnvironment(environment) {
	return "dispatchFetch" in environment;
}
/**
* Resolves catch-all entry and forwards incoming requests to its `fetch()` handler.
*
* @param environment which server-side environment to use. Defaults to `ssr`
*/
function devServer({ environment } = {}) {
	return {
		name: "universal-deploy:dev-server",
		apply(_config, { command, mode }) {
			return command === "serve" && mode !== "test";
		},
		configureServer(server) {
			return () => {
				const ssr = server.environments[environment ?? "ssr"];
				if (!ssr) return this.error(`Cannot find environment "${environment}"`);
				if (ssr.config.consumer !== "server") return this.error(`Cannot use environment "${environment}" (requires a server environment)`);
				let resolvedId;
				async function devMiddleware(request) {
					if (isRunnableDevEnvironment(ssr)) {
						if (!resolvedId) {
							const resolved = await ssr.pluginContainer.resolveId(catchAllEntry);
							if (!resolved?.id) throw new Error(`Could not resolve server entry ${catchAllEntry}`);
							resolvedId = resolved.id;
						}
						return (await envImportFetchable(ssr, resolvedId)).fetch(request);
					} else if (isFetchableDevEnvironment(ssr)) return ssr.dispatchFetch(request);
				}
				server.middlewares.use(createMiddleware(() => devMiddleware)());
			};
		}
	};
}
async function envImportFetchable(env, resolvedId) {
	return assertFetchable(await env.runner.import(resolvedId), resolvedId);
}

//#endregion
//#region src/plugins/resolver.ts
const pluginName = "ud:resolver";
/**
* Keep track of resolved server entries
*/
function resolver() {
	const stateId = /* @__PURE__ */ new WeakMap();
	const apiState = /* @__PURE__ */ new WeakMap();
	function getApi(env) {
		if (!apiState.has(env)) apiState.set(env, /* @__PURE__ */ new Map());
		const entriesMap = apiState.get(env);
		return {
			findEntries: entriesMap,
			addEntry(resolvedId, ...meta) {
				const entries = entriesMap.get(resolvedId) ?? [];
				entries.push(...meta);
				entriesMap.set(resolvedId, entries);
			}
		};
	}
	return {
		name: pluginName,
		enforce: "pre",
		perEnvironmentStartEndDuringDev: true,
		sharedDuringBuild: false,
		buildStart() {
			const entriesId = /* @__PURE__ */ new Set();
			for (const e of getAllEntries()) entriesId.add(e.id);
			stateId.set(this.environment, entriesId);
		},
		api(env) {
			const { findEntries } = getApi(env);
			return { findEntries };
		},
		resolveId: {
			order: "pre",
			async handler(id, importer, options) {
				if (stateId.get(this.environment)?.has(id)) {
					const resolved = await this.resolve(id, importer, {
						skipSelf: true,
						...options
					});
					if (resolved) getApi(this.environment).addEntry(resolved.id, ...getAllEntries().filter((e) => e.id === id));
					return resolved;
				}
			}
		}
	};
}
const resolverPluginWm = /* @__PURE__ */ new WeakMap();
function isServerEntry(env, id) {
	let resolverPlugin = resolverPluginWm.get(env);
	if (!resolverPlugin) {
		resolverPlugin = env.plugins.find((p) => p.name === pluginName);
		if (resolverPlugin) resolverPluginWm.set(env, resolverPlugin);
	}
	if (!resolverPlugin || !resolverPlugin.api) throw new Error(`Missing ${pluginName} plugin`);
	return (resolverPlugin.api(env).findEntries.get(id)?.length ?? 0) > 0;
}

//#endregion
//#region src/plugins/hmr.ts
function hmr() {
	return {
		name: "ud:hmr",
		apply: "serve",
		applyToEnvironment(env) {
			return env.config.consumer === "server";
		},
		async transform(code, id) {
			if (!isServerEntry(this.environment, id)) return;
			if (code.includes("import.meta.hot.accept(")) return;
			const s = new MagicString(code);
			s.append(`
if (import.meta.hot) {
  import.meta.hot.accept();
}
`);
			return {
				code: s.toString(),
				map: s.generateMap({ hires: true })
			};
		}
	};
}

//#endregion
//#region src/plugins/rollup-entries-compat.ts
/**
* Vite plugin that provides compatibility for resolving the SSR Rollup entry
* and registers it in the shared `store.entries`.
*
* Notes:
* - The effect is applied only once per process (guarded by an internal flag).
* - It currently registers a catch-all route pattern.
*/
function compat(config) {
	return {
		name: "ud:rollup-ssr-entry-compat",
		config: { handler(userConfig) {
			const buildOptions = userConfig.environments?.ssr?.build;
			const input = buildOptions?.rolldownOptions?.input ?? buildOptions?.rollupOptions?.input;
			const inputStr = typeof config?.entry === "string" ? config.entry : typeof input === "string" ? input : Array.isArray(input) && input.length > 0 ? input[0] : input && "index" in input ? input.index : void 0;
			if (!inputStr) return;
			addEntry({
				id: inputStr,
				route: "/**"
			});
		} },
		...dependsOn("ud:catch-all")
	};
}

//#endregion
//#region src/index.ts
var src_default = auto;

//#endregion
export { INSTANCE, auto, catchAll, catchAllId, compat, src_default as default, devServer, hmr, isServerEntry, node, resolveTargets, resolver };
