// drizzle-orm@1.0.0-rc.4/mysql2/codecs.js
import { sql } from "../sql/sql.js";
import { floatFromDouble, refineGenericMySqlCodecs } from "../mysql-core/codecs.js";

//#region src/mysql2/codecs.ts
const mysql2Codecs = refineGenericMySqlCodecs({
	bigint: { normalize: BigInt },
	"bigint:number": { normalize: Number },
	boolean: { normalize: (value) => value === 1 },
	date: { normalize: (value) => new Date(value) },
	datetime: { normalize: (value) => /* @__PURE__ */ new Date(value.replace(" ", "T") + "Z") },
	timestamp: { normalize: (value) => /* @__PURE__ */ new Date(value + "+0000") },
	"decimal:number": { normalize: Number },
	"decimal:bigint": { normalize: BigInt },
	float: {
		cast: (name) => sql`cast(${name} as double)`,
		normalize: floatFromDouble
	},
	binary: { normalize: (value) => value.toString() },
	varbinary: { normalize: (value) => value.toString() },
	blob: { normalize: (value) => value.toString("utf8") },
	tinyblob: { normalize: (value) => value.toString("utf8") },
	mediumblob: { normalize: (value) => value.toString("utf8") },
	longblob: { normalize: (value) => value.toString("utf8") }
});

//#endregion
export { mysql2Codecs };
