// drizzle-orm@1.0.0-rc.4/mysql2/codecs.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../_virtual/_rolldown/runtime.cjs');
let __sql_sql_ts = require("../sql/sql.cjs");
let __mysql_core_codecs_ts = require("../mysql-core/codecs.cjs");

//#region src/mysql2/codecs.ts
const mysql2Codecs = (0, __mysql_core_codecs_ts.refineGenericMySqlCodecs)({
	bigint: { normalize: BigInt },
	"bigint:number": { normalize: Number },
	boolean: { normalize: (value) => value === 1 },
	date: { normalize: (value) => new Date(value) },
	datetime: { normalize: (value) => /* @__PURE__ */ new Date(value.replace(" ", "T") + "Z") },
	timestamp: { normalize: (value) => /* @__PURE__ */ new Date(value + "+0000") },
	"decimal:number": { normalize: Number },
	"decimal:bigint": { normalize: BigInt },
	float: {
		cast: (name) => __sql_sql_ts.sql`cast(${name} as double)`,
		normalize: __mysql_core_codecs_ts.floatFromDouble
	},
	binary: { normalize: (value) => value.toString() },
	varbinary: { normalize: (value) => value.toString() },
	blob: { normalize: (value) => value.toString("utf8") },
	tinyblob: { normalize: (value) => value.toString("utf8") },
	mediumblob: { normalize: (value) => value.toString("utf8") },
	longblob: { normalize: (value) => value.toString("utf8") }
});

//#endregion
exports.mysql2Codecs = mysql2Codecs;
