// drizzle-orm@1.0.0-rc.4/bun-sql/mysql/codecs.js
import { sql } from "../../sql/sql.js";
import { castToText, floatFromDouble, refineGenericMySqlCodecs } from "../../mysql-core/codecs.js";

//#region src/bun-sql/mysql/codecs.ts
const bunSqlMySqlCodecs = refineGenericMySqlCodecs({
	"bigint:number": { normalize: Number },
	"bigint:string": { normalize: String },
	boolean: { normalize: (value) => value === 1 },
	year: {
		cast: (name) => sql`cast(${name} as unsigned)`,
		normalize: Number
	},
	"date:string": { normalize: (value) => value.toISOString().slice(0, -14) },
	"datetime:string": { normalize: (value) => value.toISOString().slice(0, -1).replace("T", " ") },
	"timestamp:string": { cast: castToText },
	"decimal:number": { normalize: Number },
	"decimal:bigint": { normalize: BigInt },
	float: {
		cast: (name) => sql`cast(${name} as double)`,
		normalize: floatFromDouble
	},
	binary: { normalize: (value) => value.toString() },
	varbinary: { normalize: (value) => value.toString() },
	blob: { normalize: (value) => value.toString("utf8") },
	tinyblob: { normalize: (value) => value.toString("utf8") },
	mediumblob: { normalize: (value) => value.toString("utf8") },
	longblob: { normalize: (value) => value.toString("utf8") }
});

//#endregion
export { bunSqlMySqlCodecs };
