// drizzle-orm@1.0.0-rc.4/bun-sql/mysql/codecs.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtime = require('../../_virtual/_rolldown/runtime.cjs');
let __sql_sql_ts = require("../../sql/sql.cjs");
let __mysql_core_codecs_ts = require("../../mysql-core/codecs.cjs");

//#region src/bun-sql/mysql/codecs.ts
const bunSqlMySqlCodecs = (0, __mysql_core_codecs_ts.refineGenericMySqlCodecs)({
	"bigint:number": { normalize: Number },
	"bigint:string": { normalize: String },
	boolean: { normalize: (value) => value === 1 },
	year: {
		cast: (name) => __sql_sql_ts.sql`cast(${name} as unsigned)`,
		normalize: Number
	},
	"date:string": { normalize: (value) => value.toISOString().slice(0, -14) },
	"datetime:string": { normalize: (value) => value.toISOString().slice(0, -1).replace("T", " ") },
	"timestamp:string": { cast: __mysql_core_codecs_ts.castToText },
	"decimal:number": { normalize: Number },
	"decimal:bigint": { normalize: BigInt },
	float: {
		cast: (name) => __sql_sql_ts.sql`cast(${name} as double)`,
		normalize: __mysql_core_codecs_ts.floatFromDouble
	},
	binary: { normalize: (value) => value.toString() },
	varbinary: { normalize: (value) => value.toString() },
	blob: { normalize: (value) => value.toString("utf8") },
	tinyblob: { normalize: (value) => value.toString("utf8") },
	mediumblob: { normalize: (value) => value.toString("utf8") },
	longblob: { normalize: (value) => value.toString("utf8") }
});

//#endregion
exports.bunSqlMySqlCodecs = bunSqlMySqlCodecs;
